{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lys2 Version 2 - sent to Bruce on Jan 27 \n",
    "## try to minimize! the codon pair score in the cost function\n",
    "## no longer care about keeping the rest of the gene wt\n",
    "The minimum possible codon pair score for for [120:-60], without caring for any other sequences is -293.61"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGCTCCCACAACTTGATGTCCCGCATGATAGCTTCTCTAATAAGTATGCAGTTGCACTTAGTGTGTGGGCAGCACTTATCTATCGGGTTACAGGGGATGATGATATAGTCTTGTATATAGCTAATAATAAGATACTTAGATTTAATATACAACCTACTTGGAGCTTTAATGAATTGTATAGTACTATTAATAATGAACTTAATAAGCTTAATTCTATAGAAGCTAACTTCTCTTTTGATGAATTAGCTGAGAAGATACAATCCTGTCAGGATCTAGAACGTACACCCCAATTGTTTCGTCTGGCTTTTCTTGAGAATCAAGACTTTAAGTTAGATGAATTTAAACACCACCTAGTTGACTTTGCACTTAATCTAGATACGTCTAATAATGCCCATGTACTTAATCTTATCTATAATAGCTTGTTGTATAGTAATGAGCGGGTTACTATAGTTGCGGATCAATTTACACAATACCTTACAGCAGCCTTGTCGGACCCGTCTAATTGTATTACTAAGATTAGCCTTATTACAGCCTCGTCTAAAGATAGCCTCCCAGACCCTACTAAGAATCTAGGTTGGTGTGACTTTGTAGGTTGTATACATGATATCTTTCAGGATAATGCTGAGGCTTTTCCAGAACGTACTTGTGTAGTTGAGACCCCTACACTTAATTCAGATAAGTCTAGATCTTTTACATATCGGGATATTAATCGTACGTCTAATATAGTAGCCCACTACCTTATAAAAACGGGTATTAAGCGGGGGGATGTAGTTATGATCTATAGCTCCCGGGGGGTTGATCTTATGGTGTGTGTTATGGGAGTCCTTAAAGCGGGGGCTACATTCTCAGTTATAGATCCAGCCTACCCCCCGGCCCGACAGACTATCTATCTAGGAGTAGCTAAGCCTAGGGGACTTATAGTTATTCGGGCAGCGGGTCAGTTAGATCAACTAGTTGAGGATTATATTAATGATGAACTTGAGATAGTTAGTCGTATTAATTCTATAGCTATACAAGAGAATGGTACTATAGAAGGGGGTAAGTTAGATAATGGTGAGGATGTGTTAGCCCCGTATGATCATTATAAAGATACGCGTACTGGAGTAGTTGTAGGTCCAGACTCTAACCCTACACTTAGCTTTACATCGGGTTCGGAGGGTATCCCTAAGGGGGTCCTAGGCCGACACTTTAGTCTAGCCTATTACTTTAATTGGATGTCTAAACGCTTTAATCTTACAGAGAATGATAAGTTTACTATGTTGTCGGGTATAGCCCATGATCCTATACAGAGGGATATGTTTACACCCCTGTTTCTAGGAGCCCAGTTGTATGTACCTACTCAGGATGATATAGGTACCCCGGGTAGGTTAGCTGAGTGGATGTCTAAGTATGGGTGTACAGTTACACACCTTACCCCAGCTATGGGTCAGTTGCTTACAGCCCAGGCTACTACCCCTTTTCCTAAGTTACATCATGCTTTTTTTGTAGGGGATATACTTACTAAGCGGGATTGCCTTAGACTTCAGACCCTAGCTGAGAATTGCCGTATAGTTAATATGTATGGTACTACGGAGACCCAGAGAGCAGTTAGTTACTTTGAGGTTAAGTCTAAGAATGATGATCCTAACTTCCTTAAAAAACTTAAGGATGTTATGCCGGCGGGTAAGGGTATGTTAAATGTACAACTCCTTGTAGTTAATCGTAATGATCGTACTCAGATCTGTGGTATAGGTGAGATAGGTGAGATCTATGTCCGAGCGGGGGGTCTGGCTGAGGGTTATCGGGGTCTGCCAGAACTTAATAAAGAAAAATTTGTGAATAATTGGTTTGTAGAAAAAGATCATTGGAATTACTTAGATAAGGATAATGGAGAACCCTGGCGTCAATTCTGGTTGGGTCCTAGGGATAGGTTGTATCGTACTGGGGATCTAGGTAGGTACCTCCCGAATGGGGATTGTGAATGTTGTGGTCGGGCAGATGATCAGGTTAAGATTCGGGGTTTTCGTATAGAATTAGGGGAGATAGATACCCATATTAGTCAGCACCCTCTAGTTAGGGAGAATATTACCCTAGTCCGAAAGAATGCAGATAATGAACCTACACTTATTACCTTTATGGTCCCGCGTTTTGATAAGCCAGATGATCTTAGTAAGTTTCAGTCTGATGTCCCGAAAGAAGTTGAGACTGACCCTATAGTTAAAGGACTTATAGGTTACCACCTCCTTAGTAAGGATATACGTACATTCCTTAAAAAACGCCTAGCTAGCTATGCTATGCCTAGCCTTATAGTAGTTATGGATAAGTTGCCACTTAACCCGAATGGTAAGGTAGATAAGCCTAAGTTGCAATTTCCTACCCCTAAACAACTTAATCTTGTAGCTGAGAATACAGTCTCAGAGACTGATGATAGTCAATTTACTAATGTAGAACGTGAGGTTAGGGATCTGTGGTTGTCTATACTCCCTACTAAGCCAGCCTCAGTTAGTCCAGATGATAGCTTTTTTGATCTAGGGGGTCATAGTATCTTAGCTACTAAGATGATCTTTACACTTAAAAAAAAACTTCAGGTTGATCTCCCTCTAGGTACTATCTTTAAGTACCCTACTATTAAAGCTTTTGCAGCTGAGATAGATCGTATTAAGTCCTCGGGGGGTTCGAGTCAGGGGGAAGTAGTTGAGAATGTTACAGCTAATTATGCAGAGGATGCAAAAAAGTTAGTTGAGACCCTCCCGTCTAGCTACCCGTCTAGAGAATACTTTGTAGAACCTAACTCGGCTGAGGGTAAGACTACTATTAATGTGTTTGTTACAGGGGTTACGGGTTTTCTAGGGAGTTATATCTTAGCTGACTTGTTAGGCCGATCCCCTAAGAATTATAGCTTTAAGGTGTTTGCACATGTCCGAGCTAAGGATGAAGAAGCAGCTTTTGCGCGACTTCAGAAAGCGGGTATTACCTATGGTACGTGGAATGAAAAATTTGCGAGTAATATTAAGGTAGTCCTAGGGGATCTTAGTAAGTCCCAATTTGGCTTGTCGGATGAGAAGTGGATGGATCTAGCTAATACAGTTGATATTATTATACATAATGGAGCCTTAGTACACTGGGTGTACCCCTATGCTAAGCTTAGAGACCCGAATGTTATAAGTACTATTAATGTTATGAGTCTAGCAGCGGTTGGCAAACCTAAGTTCTTTGACTTTGTGTCTAGTACTAGTACCCTAGATACGGAATATTACTTTAATCTTAGTGATAAGTTAGTCTCAGAGGGTAAGCCGGGTATACTTGAGTCGGATGATCTTATGAACTCAGCCTCGGGTCTTACAGGGGGTTATGGTCAGTCTAAGTGGGCAGCGGAATATATTATTCGACGAGCGGGGGAGAGGGGACTTAGGGGGTGTATAGTTAGACCGGGTTATGTTACGGGAGCCTCGGCTAATGGTTCGTCTAATACTGATGACTTCTTGCTTAGATTCCTTAAGGGGTCAGTCCAACTAGGTAAGATCCCAGATATAGAAAACTCAGTTAATATGGTCCCGGTTGATCATGTAGCCCGAGTTGTAGTAGCTACGAGCCTTAATCCCCCGAAAGAGAATGAATTAGCTGTAGCCCAGGTTACGGGTCACCCGCGTATCTTGTTTAAGGATTACTTGTATACTCTCCATGATTATGGGTATGATGTTGAGATAGAATCCTATAGTAAGTGGAAAAAGTCCCTTGAAGCCTCAGTTATAGATAGGAATGAAGAGAATGCCTTGTACCCCCTGTTACATATGGTGTTAGATAACCTCCCGGAATCTACTAAAGCCCCGGAATTAGATGATAGGAATGCAGTTGCGAGCCTTAAAAAAGATACAGCCTGGACTGGAGTTGACTGGTCTAATGGTATAGGAGTTACCCCTGAGGAAGTAGGTATCTATATAGCTTTTCTTAATAAGGTAGGTTTTCTACCCCCCCCTACTCATAATGATAAGTTGCCACTCCCGTCTATAGAACTTACACAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "total number of codon diferences between original and recoded sequences 853\n",
      "AA sequence remains the same, good\n",
      "4179\n",
      "num desired codon pairs in original sequence: 66\n",
      "num desired codon pairs in recoded sequence: 155\n",
      "codon pair score of the original sequence: 34.78041769633767\n",
      "codon pair score of the recoded sequence: -211.20682461324952\n",
      "CAI of the original sequence is: 0.714988264144\n",
      "CAI of the recoded sequence is: 0.710975563707\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "\n",
    "# number of desired codon paris in sequence of any length\n",
    "def num_costly_codon_pairs(seq, count_negatives):\n",
    "    num = 0\n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        '''if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "            num += 1\n",
    "        if (count_negatives):\n",
    "            if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "                num -= 1'''\n",
    "        '''if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "            num += 1\n",
    "        if (count_negatives):\n",
    "            if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "                num -= 1'''\n",
    "        if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"T\" or seq[i+4] == \"C\"))):\n",
    "            num += 1\n",
    "        if (count_negatives):\n",
    "            if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "                num -= 1   \n",
    "    return(num)\n",
    "\n",
    "# codon pair score of any length sequence\n",
    "def get_cps(seq):\n",
    "    cps = 0\n",
    "    if (len(seq) == 6):\n",
    "        return(cp_scores[seq])  \n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    return(cps)\n",
    "\n",
    "# codon adaptation index of any length sequence\n",
    "def get_cai(seq):\n",
    "    codon_occs = dict()\n",
    "    # set up the dict\n",
    "    for codon in (codon_to_aa_dict.keys()):\n",
    "        codon_occs[codon] = 0\n",
    "    # get the frequency of each codon\n",
    "    for i in range(0,len(seq),3):\n",
    "        codon_occs[seq[i:(i+3)]] += 1\n",
    "    # get the CAI of each codon\n",
    "    codon_CAI = dict()\n",
    "    for aa in (aa_to_syn_codons_dict.keys()):\n",
    "        highest_frequency = 0\n",
    "        for codon in (aa_to_syn_codons_dict[aa]):\n",
    "            if (codon_occs[codon] > highest_frequency):\n",
    "                highest_frequency = codon_occs[codon]\n",
    "        for codon in (aa_to_syn_codons_dict[aa]):\n",
    "            if (highest_frequency == 0):\n",
    "                codon_CAI[codon] = 0\n",
    "            else:\n",
    "                codon_CAI[codon] = codon_occs[codon]/highest_frequency\n",
    "    # get the CAI as the geometric mean of all the codon's individual CAI values\n",
    "    cai_list = []\n",
    "    for i in range(0,len(seq),3):\n",
    "        cai_list.append(codon_CAI[seq[i:(i+3)]])\n",
    "    return(scistats.gmean(cai_list))\n",
    "\n",
    "# read in synonymous codons\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "\n",
    "# read in the codon pair scores\n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"yeast_cps.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[0]] = float(line[1])\n",
    "\n",
    "# read in the gene's codon sequence\n",
    "codons_file = open(\"S288C_YBR115C_LYS2_genomic.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "# read in the gene's amino acid sequence\n",
    "aa_file = open(\"S288C_YBR115C_LYS2_protein.fsa\")\n",
    "original_AA = \"\"\n",
    "for l in aa_file:\n",
    "    l = l.rstrip()\n",
    "    if l[0] != \">\":\n",
    "        original_AA = original_AA + l\n",
    "aa_file.close()\n",
    "\n",
    "if (len(original_AA)*3 != len(original_codons)):\n",
    "    print(\"PROBLEM: codons and AA don't match\")\n",
    "\n",
    "    \n",
    "# store the sequences, and then work with the ones that have \"chopped off\" start and end\n",
    "long_original_AA = original_AA\n",
    "original_AA = original_AA[40:-20]\n",
    "long_original_codons = original_codons\n",
    "original_codons = original_codons[120:-60]\n",
    "   \n",
    "# store_best will contain, at each AA position, and at each of the syn codon choices,\n",
    "# the best cost of the suffix that starts with this codon, the current codon pair score,\n",
    "# and the next codon that corresponds to this cost\n",
    "store_best = [None] * len(original_AA)\n",
    "for i in range(0,len(original_AA)):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = 0\n",
    "                \n",
    "\n",
    "# initialize the dynamic program by setting store_best for the last amino acid position\n",
    "for codon in store_best[-1]:\n",
    "    store_best[i][codon] = (0,0,\"\")\n",
    "\n",
    "\n",
    "# do the dynamic program, stating from the one before the end of the list, until the beginning\n",
    "for pos in reversed(range(0, len(store_best)-1)):\n",
    "    for codon in store_best[pos]:\n",
    "        # consider as cost the cost of adding this codon to one of the best sequences\n",
    "        # from the previous suffix\n",
    "        best_prev = \"\"\n",
    "        best_cps = sys.maxsize # keep this\n",
    "        best_cost = -1 #sys.maxsize #-1\n",
    "        for prev_codon in store_best[pos+1]:\n",
    "            cost = num_costly_codon_pairs(codon+prev_codon,1) + store_best[pos+1][prev_codon][0]\n",
    "            cp_score = get_cps(codon+prev_codon) + store_best[pos+1][prev_codon][1]\n",
    "            #print(cp_score)\n",
    "            # maximizer the cost\n",
    "            #if (cp_score < best_cps):\n",
    "            if ((cost > best_cost) or # > <\n",
    "                # break the tie with lowest codon pair score\n",
    "                ((cost == best_cost) and (cp_score < best_cps))):\n",
    "                best_cost = cost\n",
    "                best_cps = cp_score\n",
    "                best_prev = prev_codon\n",
    "        store_best[pos][codon] = (best_cost, best_cps, best_prev)\n",
    "        #print(\"storing for \" + str(pos) + \" \" + codon + \": \" + str(best_cps) + \" \" + best_prev)\n",
    "\n",
    "# going from the beginning to the end, read off the recoded sequences from the store_best\n",
    "# re-doing the computation for the first position, to pick the best of the syn codons\n",
    "# the rest of the seuqnces is just read out\n",
    "recoded_sequence = \"\"\n",
    "next_codon = \"\"\n",
    "for pos in range(0, len(store_best)):\n",
    "    if pos == 0:\n",
    "        # maximum\n",
    "        best_first = \"\"\n",
    "        best_first_cps = sys.maxsize # keep this\n",
    "        best_first_cost = -1 # sys.maxsize #-1\n",
    "        for codon in store_best[0]:\n",
    "            # maximize the cost\n",
    "            #if (store_best[pos][codon][1] < best_first_cps):\n",
    "            if ((store_best[pos][codon][0] > best_first_cost) or  #> <\n",
    "               ((store_best[pos][codon][0] == best_first_cost) and (store_best[pos][codon][1] < best_first_cps))):\n",
    "                best_first = codon\n",
    "                best_first_cost = store_best[pos][codon][0]\n",
    "                best_first_cps = store_best[pos][codon][1]\n",
    "                next_codon = store_best[pos][codon][2]\n",
    "        recoded_sequence += best_first\n",
    "    else:\n",
    "        recoded_sequence += next_codon\n",
    "        next_codon = store_best[pos][next_codon][2]\n",
    "\n",
    "# now restore the original sequences and patch up the recoded sequence with the beginning and end\n",
    "original_AA = long_original_AA\n",
    "original_codons = long_original_codons\n",
    "recoded_sequence = original_codons[0:120] + recoded_sequence + original_codons[-60:]\n",
    "\n",
    "#print(original_codons)\n",
    "print(recoded_sequence)    \n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "print(len(original_codons))\n",
    "print(\"num desired codon pairs in original sequence: \" + str(num_costly_codon_pairs(original_codons, 0)))\n",
    "print(\"num desired codon pairs in recoded sequence: \" + str(num_costly_codon_pairs(recoded_sequence, 0)))\n",
    "print(\"codon pair score of the original sequence: \" + str(get_cps(original_codons)))\n",
    "print(\"codon pair score of the recoded sequence: \" + str(get_cps(recoded_sequence)))\n",
    "print(\"CAI of the original sequence is: \" + str(get_cai(original_codons)))\n",
    "print(\"CAI of the recoded sequence is: \" + str(get_cai(recoded_sequence)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Version 1\n",
    "vienna package - something like mfold - look into with justin - Bruce not worried about this\n",
    "vienna package verbose\n",
    "\n",
    "splice sites - Justin,\n",
    "https://users.soe.ucsc.edu/~leslie/runIntronSearch.html\n",
    "http://harlequin.jax.org/drupal/?q=yeastRNApred\n",
    "\n",
    "transcriptional terminators - Justin did not do anything;  they are not well defined....\n",
    "\n",
    "restriction sites - Bruce (look at what genscript uses)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGTTACCTCAGCTCGACGTGCCTCACGATAGTTTTTCTAACAAATACGCCGTCGCTTTGAGCGTATGGGCCGCATTGATATATCGCGTAACCGGCGACGACGATATCGTTCTTTATATCGCGAATAACAAAATCTTAAGATTCAATATTCAACCAACGTGGTCATTTAACGAGCTGTATTCTACAATTAACAACGAGTTGAACAAGCTCAATTCTATCGAGGCCAATTTTTCCTTCGACGAGCTCGCCGAAAAAATTCAAAGTTGCCAAGATCTCGAAAGGACCCCTCAGTTGTTCCGTCTCGCCTTTCTCGAAAACCAAGATTTCAAACTCGACGAGTTCAAGCATCATCTCGTCGACTTCGCTTTGAATCTCGATACCAGTAATAACGCGCACGTTTTGAACTTAATTTATAACAGCTTACTGTATTCGAACGAACGCGTAACCATCGTCGCCGACCAATTTACTCAATATTTGACCGCCGCGCTAAGCGATCCATCCAATTGCATAACTAAAATCTCTCTGATCACCGCATCATCCAAGGATAGTTTACCCGATCCAACTAAGAACCTCGGCTGGTGCGATTTCGTCGGGTGTATTCACGACATTTTCCAGGACAACGCCGAAGCCTTCCCCGAGAGAACCTGCGTCGTCGAGACTCCAACACTAAATTCCGACAAGTCCCGTTCTTTCACTTATCGCGACATCAACCGCACTTCTAACATCGTCGCCCATTATTTGATTAAAACCGGTATCAAACGCGGCGACGTCGTGATGATCTATTCTTCTCGCGGCGTCGATTTGATGGTATGCGTGATGGGCGTCTTGAAAGCCGGCGCAACCTTTAGCGTTATCGACCCCGCATATCCCCCCGCCAGACAAACCATTTACCTCGGCGTCGCTAAACCACGCGGGTTGATCGTTATTCGCGCCGCCGGACAACTCGATCAACTCGTCGAAGATTACATCAACGACGAACTCGAGATCGTTTCAAGAATCAATTCCATCGCTATTCAAGAAAACGGTACCATCGAAGGCGGCAAACTCGACAACGGCGAGGACGTTCTCGCTCCATACGATCACTACAAAGACACCAGAACCGGCGTCGTCGTCGGACCCGATTCCAACCCAACCCTATCTTTCACAAGCGGTTCCGAAGGTATTCCTAAGGGCGTTCTCGGTAGACATTTTTCCCTCGCTTATTATTTCAATTGGATGTCCAAAAGGTTCAACTTAACCGAAAACGATAAATTCACAATGCTGAGCGGTATCGCACACGATCCAATTCAACGCGATATGTTTACACCATTATTTCTCGGCGCCCAATTGTACGTCCCTACTCAAGACGATATCGGTACACCCGGCCGTCTCGCCGAATGGATGAGTAAGTACGGTTGCACCGTTACCCATTTAACACCCGCCATGGGTCAATTACTTACCGCCCAAGCTACTACACCATTCCCTAAGTTACATCACGCGTTCTTCGTCGGCGACATTTTAACAAAACGCGATTGTCTGAGGTTACAAACCCTCGCCGAAAATTGCCGTATCGTTAATATGTACGGTACCACCGAAACACAGCGCGCCGTTTCTTATTTCGAAGTTAAATCAAAAAACGACGATCCAAACTTTTTGAAAAAATTGAAAGACGTCATGCCCGCCGGTAAAGGTATGTTGAACGTTCAGCTACTCGTCGTTAACAGGAACGATCGTACTCAAATATGCGGTATCGGCGAAATCGGCGAGATTTACGTTCGCGCCGGCGGTCTCGCCGAAGGTTATCGCGGATTACCCGAATTGAATAAAGAAAAATTCGTGAACAACTGGTTCGTCGAAAAAGATCACTGGAATTATCTCGATAAGGATAACGGCGAACCTTGGAGACAATTCTGGCTCGGTCCACGCGATAGATTGTACAGAACCGGCGATCTCGGTCGTTATCTACCAAACGGCGACTGCGAATGTTGCGGTCGCGCCGACGATCAAGTTAAAATTCGCGGGTTCAGAATCGAACTCGGCGAAATCGATACGCACATTTCCCAACATCCACTCGTACGCGAAAACATTACTCTCGTTCGCAAAAACGCCGACAACGAGCCAACATTGATCACATTTATGGTCCCAAGATTCGACAAGCCCGACGACTTGTCTAAGTTCCAAAGCGACGTTCCAAAGGAGGTCGAAACCGACCCTATCGTTAAGGGCTTAATCGGTTACCATCTTTTATCCAAGGACATCAGGACTTTCTTAAAGAAAAGACTCGCTAGCTACGCTATGCCTTCCTTGATCGTCGTTATGGATAAACTACCATTGAATCCAAACGGTAAAGTCGATAAGCCTAAACTTCAATTCCCAACTCCCAAGCAATTAAATCTCGTCGCCGAAAATACCGTTAGCGAAACCGACGACTCTCAGTTTACCAACGTCGAGCGCGAGGTTCGCGACTTATGGTTAAGTATATTACCTACCAAGCCCGCAAGCGTATCACCCGACGATTCGTTTTTCGATCTCGGCGGTCATTCTATCCTCGCTACCAAAATGATTTTTACCTTAAAGAAAAAGCTGCAAGTCGATTTACCACTCGGCACAATTTTCAAGTATCCAACGATAAAGGCCTTCGCCGCCGAAATCGACAGAATTAAATCAAGCGGCGGATCATCTCAAGGCGAGGTCGTCGAAAACGTCACCGCAAATTACGCCGAAGACGCCAAGAAACTCGTCGAGACGCTACCAAGTTCGTACCCCTCTCGCGAATATTTCGTCGAACCTAATAGCGCCGAAGGAAAAACAACAATTAACGTGTTCGTTACCGGCGTCACCGGATTTCTCGGCTCCTACATCCTCGCCGATTTGCTCGGACGTTCTCCAAAGAACTACAGTTTCAAAGTGTTCGCCCACGTCCGCGCCAAGGACGAAGAAGCCGCATTCGCAAGATTACAAAAGGCCGGTATCACCTACGGTACTTGGAACGAAAAATTCGCCTCAAATATTAAAGTCGTACTCGGCGATTTATCTAAAAGCCAATTCGGTCTTAGCGACGAGAAGTGGATGGATCTCGCAAACACCGTCGATATAATTATCCATAACGGCGCGCTCGTTCACTGGGTTTATCCATACGCCAAATTGCGCGATCCAAACGTTATTTCAACTATCAACGTTATGAGCCTCGCCGCCGTCGGCAAGCCAAAGTTCTTCGACTTCGTTTCCTCCACTTCTACTCTCGACACCGAATACTACTTTAATTTGAGCGATAAACTCGTTAGCGAAGGGAAGCCCGGCATTCTCGAAAGCGACGATTTAATGAACAGCGCAAGCGGGCTCACCGGCGGATACGGTCAGTCCAAATGGGCCGCCGAGTACATCATTAGACGCGCCGGCGAACGCGGCCTACGCGGGTGTATCGTCAGACCCGGTTACGTAACCGGCGCCAGCGCCAACGGTTCTTCAAACACCGACGATTTCTTATTGAGATTTTTGAAAGGTAGCGTCCAACTCGGTAAGATTCCCGATATCGAAAATTCCGTGAATATGGTTCCCGTCGATCACGTCGCTCGCGTCGTCGTCGCTACGTCTTTGAATCCTCCCAAAGAAAACGAACTCGCCGTCGCTCAAGTAACCGGTCACCCAAGAATATTATTCAAAGACTACTTGTATACTTTACACGATTACGGTTACGACGTCGAAATCGAAAGCTATTCTAAATGGAAGAAATCACTCGAGGCGAGCGTTATCGACAGGAACGAAGAAAACGCGTTGTATCCTTTGCTACACATGGTCCTCGACAACTTACCCGAAAGTACCAAAGCTCCCGAACTCGACGATAGGAACGCCGTCGCATCTTTAAAGAAAGACACCGCATGGACCGGCGTCGATTGGTCTAACGGAATCGGCGTTACTCCCGAAGAGGTCGGTATATATATCGCATTTTTAAACAAGGTCGGATTTTTACCTCCACCAACTCATAACGACAAACTTCCACTGCCAAGTATCGAACTAACTCAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "num costly codon pairs: 372\n",
      "total number of codon diferences between original and recoded sequences 308\n",
      "AA sequence remains the same, good\n",
      "4179\n",
      "num costly codon pairs in original sequences: 65\n",
      "codon pair score of the original sequence: 34.78041769633767\n",
      "codon pair score of the recoded sequence: -44.26762541921643\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "        \n",
    "def get_cps(seq):\n",
    "    cps = 0\n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    return(cps)        \n",
    "\n",
    "def num_costly_codon_pairs(seq, do_print):\n",
    "    num = 0\n",
    "    to_print = \"\"\n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "        #if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "            num += 1\n",
    "            if (do_print == 1):\n",
    "                to_print += str(i) + \" \" #seq[i:i+6] + \" \" + \n",
    "    if (do_print):\n",
    "        print(to_print)\n",
    "    return(num)\n",
    "\n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"yeast_cps.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[0]] = float(line[1])\n",
    "    \n",
    "codons_file = open(\"S288C_YBR115C_LYS2_genomic.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "aa_file = open(\"S288C_YBR115C_LYS2_protein.fsa\")\n",
    "original_AA = \"\"\n",
    "for l in aa_file:\n",
    "    l = l.rstrip()\n",
    "    if l[0] != \">\":\n",
    "        original_AA = original_AA + l\n",
    "aa_file.close()\n",
    "\n",
    "if (len(original_AA)*3 != len(original_codons)):\n",
    "    print(\"PROBLEM: codons and AA don't match\")\n",
    "\n",
    "    \n",
    "# store the sequences, and then work with the ones that have \"chopped off\" start and end\n",
    "long_original_AA = original_AA\n",
    "original_AA = original_AA[40:-20]\n",
    "long_original_codons = original_codons\n",
    "original_codons = original_codons[120:-60]\n",
    "    \n",
    "store_best = [None] * len(original_AA)\n",
    "for i in range(0,len(original_AA)):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = 0\n",
    "                \n",
    "\n",
    "# do the last AA\n",
    "for codon in store_best[-1]:\n",
    "    store_best[i][codon] = (0,\"\")\n",
    "\n",
    "\n",
    "# do the dynamic program, stating from the one before the end of the list, until the end\n",
    "for pos in reversed(range(0, len(store_best)-1)):\n",
    "    for codon in store_best[pos]:\n",
    "        # consider as cost the cost of adding this codon to one of the best sequences\n",
    "        # from the previous suffix\n",
    "        # maximizer the dimers\n",
    "        best_prev = \"\"\n",
    "        best_cost = -1 #sys.maxsize #-1\n",
    "        for prev_codon in store_best[pos+1]:\n",
    "            cost = num_costly_codon_pairs(codon+prev_codon,0) + store_best[pos+1][prev_codon][0]\n",
    "            # maximizer the dimers\n",
    "            if ((cost > best_cost) or # > <\n",
    "                ((cost == best_cost) and (prev_codon == original_codons[(pos+1)*3:(pos+1)*3+3]))): #try to keep same seq\n",
    "                best_cost = cost\n",
    "                best_prev = prev_codon\n",
    "        store_best[pos][codon] = (best_cost, best_prev)\n",
    "        #print(\"storing for \" + str(pos) + \" \" + codon + \": \" + str(best_cost) + \" \" + best_prev)\n",
    "\n",
    "recoded_sequence = \"\"\n",
    "next_codon = \"\"\n",
    "for pos in range(0, len(store_best)):\n",
    "    if pos == 0:\n",
    "        # maximum\n",
    "        best_first = \"\"\n",
    "        best_first_cost = -1 # sys.maxsize #-1\n",
    "        for codon in store_best[0]:\n",
    "            # maximum\n",
    "            if ((store_best[pos][codon][0] > best_first_cost) or  #> <\n",
    "                ((cost == best_cost) and (codon == original_codons[(pos+1)*3:(pos+1)*3+3]))): #try to keep same seq\n",
    "                best_first = codon\n",
    "                best_first_cost = store_best[pos][codon][0]\n",
    "                next_codon = store_best[pos][codon][1]\n",
    "        recoded_sequence += best_first\n",
    "    else:\n",
    "        recoded_sequence += next_codon\n",
    "        next_codon = store_best[pos][next_codon][1]\n",
    "\n",
    "# now restore the original sequences and patch up the recoded sequence with the beginning and end\n",
    "original_AA = long_original_AA\n",
    "original_codons = long_original_codons\n",
    "recoded_sequence = original_codons[0:120] + recoded_sequence + original_codons[-60:]\n",
    "\n",
    "#print(original_codons)\n",
    "print(recoded_sequence)    \n",
    "print(\"num costly codon pairs: \" + str(num_costly_codon_pairs(recoded_sequence, 0)))\n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "print(len(original_codons))\n",
    "print(\"num costly codon pairs in original sequences: \" + str(num_costly_codon_pairs(original_codons, 0)))\n",
    "print(\"codon pair score of the original sequence: \" + str(get_cps(original_codons)))\n",
    "print(\"codon pair score of the recoded sequence: \" + str(get_cps(recoded_sequence)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
