{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "original codons\n",
      "GGCGCACATGAAAACTCAAATAGAGCGTATGGTAGTTCTACCATTAATTACACCACCATTAATTATTATAGAGATTCAGCTAGTAACGCGGCTTCGAAACAGGACTTCTCTCAAGACCCTTCCAAGTTCACCGAGCCCATCAAGGATGTCCTGATAAAAACAGCCCCAATGCTAAACTCGCCAAACATAGAGGCTTGCGGGTATAGCGATAGAGTACTGCAATTAACACTGGGAAACTCCACTATAACCACACAGGAGGCGGCTAATTCAGTAGTCGCTTATGGGCGTTGGCCTGAATATCTGAGGGACAGCGAAGCCAATCCAGTGGACCAGCCGACAGAACCAGACGTCGCTGCATGCAGGTTTTATACGCTAGACACCGTGTCTTGGACGAAAGAGTCGCGAGGGTGGTGGTGGAAGTTGCCTGATGCACTGAGGGACATGGGACTCTTTGGGCAAAATATGTACTACCACTACCTAGGTAGGTCCGGGTACACCGTGCATGTACAGTGTAACGCCTCCAAATTCCACCAGGGGGCACTAGGGGTATTCGCCGTACCAGAGATGTGTCTGGCCGGGGATAGCAACACCACTACCATGCACACCAGCTATCAAAATGCCAATCCTGGCGAGAAAGGAGGCACTTTCACGGGTACGTTCACTCCTGACAACAACCAGACATCACCTGCCCGCAGGTTCTGCCCGGTGGATTACCTCCTTGGAAATGGCACGTTGTTGGGGAATGCCTTTGTGTTCCCGCACCAGATAATAAACCTACGGACCAACAACTGTGCTACACTGGTACTCCCTTACGTGAACTCCCTCTCGATAGATAGTATGGTAAAGCACAATAATTGGGGAATTGCAATATTACCATTGGCCCCATTAAATTTTGCTAGTGAGTCCTCCCCAGAGATTCCAATCACCTTGACCATAGCCCCTATGTGCTGTGAGTTCAATGGATTAAGAAACATCACCCTGCCACGCTTACAGGGCCTGCCGGTCATGAACACCCCTGGTAGCAATCAATATCTTACTGCAGACAACTTCCAGTCACCGTGTGCGCTGCCTGAATTTGATGTGACCCCACCTATTGACATACCCGGTGAAGTAAAGAACATGATGGAATTGGCAGAAATCGACACCATGATTCCCTTTGACTTAAGTGCCACAAAAAAGAACACCATGGAAATGTATAGGGTTCGGTTAAGTGACAAACCACATACAGACGATCCCATACTCTGCCTGTCACTCTCTCCAGCCTCAGATCCTAGGTTGTCACATACTATGCTTGGAGAAATCCTAAATTACTACACACACTGGGCAGGATCCCTGAAGTTCACGTTTCTGTTCTGTGGATCCATGATGGCAACTGGCAAACTGTTGGTGTCATACGCGCCTCCTGGAGCCGACCCACCAAAGAAGCGTAAGGAGGCGATGTTGGGAACACATGTGATCTGGGACATAGGACTGCAGTCCTCATGTACTATGGTAGTGCCATGGATTAGCAACACCACGTATCGGCAAACCATAGATGATAGTTTCACCGAAGGCGGATACATCAGCGTCTTCTACCAAACTAGAATAGTCGTCCCTCTTTCGACACCCAGAGAGATGGACATCCTTGGTTTTGTGTCAGCGTGTAATGACTTCAGCGTGCGCTTGTTGCGAGATACCACACATATAGAGCAAAAAGCGCTAGCACAGGGGTTAGGTCAGATGCTTGAAAGCATGATTGACAACACAGTCCGTGAAACGGTGGGGGCGGCAACATCTAGAGACGCTCTCCCAAACACTGAAGCCAGTGGACCAACACACTCCAAGGAAATTCCGGCACTCACCGCAGTGGAAACTGGGGCCACAAATCCACTAGTCCCTTCTGATACAGTGCAAACCAGACATGTTGTACAACATAGGTCAAGGTCAGAGTCTAGCATAGAGTCTTTCTTCGCGCGGGGTGCATGCGTGACCATTATGACCGTGGATAACCCAGCTTCCACCACGAATAAGGATAAGCTTTTTGCAGTGTGGAAGATCACTTATAAAGATACTGTCCAGTTACGGAGGAAATTGGAGTTCTTCACCTATTCTAGATTTGATATGGAACTTACCTTTGTGGTTACTGCAAATTTCACTGAGACTAACAATGGCCATGCATTAAATCAAGTGTACCAAATTATGTACGTACCACCAGGCGCTCCAGTGCCCGAAAAATGGGACGACTACACATGGCAAACCTCATCAAATCCATCAATCTTTTACACCTACGGGACAGCTCCAGCCCGGATCTCGGTACCGTATGTTGGTATTTCGAACGCCTATTCACACTTTTACGACGGTTTTTCCAAAGTACCACTGAAGGACCAGTCGGCAGCACTAGGTGACTCCCTTTATGGTGCAGCATCTCTAAATGACTTCGGTATTTTGGCTGTTAGAGTAGTCAATGATCACAACCCGACCAAGGTCACCTCCAAAATCAGAGTGTATCTAAAACCCAAACACATCAGAGTCTGGTGCCCGCGTCCACCGAGGGCAGTGGCGTACTACGGCCCTGGAGTGGATTACAAGGATGGTACGCTTACACCCCTCTCCACCAAGGATCTGACCACATAT\n",
      "recoded sequence\n",
      "GGCGCTCACGAAAACTCGAATCGCGCATACGGATCGTCGACGATCAATTATACGACGATCAATTACTATCGCGATTCCGCGTCGAACGCCGCGTCGAAACAGGATTTTTCGCAAGATCCGTCGAAATTTACCGAACCGATTAAAGACGTACTTATCAAAACCGCGCCGATGCTCAATTCGCCGAATATCGAAGCTTGCGGTTATTCCGATCGCGTATTGCAACTTACGCTCGGAAACTCGACGATTACGACGCAAGAAGCGGCGAATTCCGTCGTCGCGTACGGTCGATGGCCCGAATACTTACGCGATTCCGAAGCGAATCCGGTCGATCAACCGACCGAACCCGACGTCGCCGCTTGTCGATTCTATACGCTCGATACCGTATCGTGGACGAAAGAATCTCGCGGTTGGTGGTGGAAATTACCGGACGCGTTACGCGATATGGGATTATTCGGTCAAAATATGTACTATCACTATCTCGGTCGATCCGGTTATACCGTACACGTACAATGCAACGCGTCGAAATTTCATCAAGGCGCATTGGGCGTATTCGCGGTACCCGAAATGTGTCTCGCCGGCGATTCGAATACGACGACGATGCATACGTCGTATCAAAACGCGAATCCCGGCGAAAAGGGGGGTACGTTTACCGGTACGTTTACGCCGGATAACAATCAAACGTCGCCCGCTCGACGATTTTGTCCGGTCGATTATCTACTCGGAAACGGTACGTTACTCGGAAACGCATTCGTATTTCCGCATCAAATTATCAATCTACGAACGAACAATTGCGCAACGCTCGTATTACCGTACGTCAATTCGCTATCGATCGATTCGATGGTTAAACATAACAATTGGGGTATCGCGATATTACCGCTCGCGCCGTTAAATTTCGCAAGCGAATCGTCGCCCGAAATACCGATTACGCTTACGATCGCGCCGATGTGTTGCGAATTCAACGGATTACGCAATATTACGCTACCGCGATTACAGGGATTACCGGTTATGAATACGCCCGGATCGAATCAATATCTTACCGCCGACAATTTTCAATCGCCTTGCGCATTACCGGAATTCGACGTTACGCCGCCGATCGATATACCCGGCGAAGTAAAAAACATGATGGAACTCGCCGAAATCGATACGATGATACCGTTCGATCTATCCGCGACGAAAAAAAATACGATGGAAATGTATCGCGTACGATTAAGCGACAAACCGCATACCGACGATCCGATACTATGTCTATCGTTATCGCCCGCAAGCGATCCGCGATTATCGCATACGATGCTCGGCGAAATACTCAATTATTATACGCATTGGGCCGGATCGCTTAAATTTACGTTTCTTTTTTGCGGATCGATGATGGCGACGGGTAAATTACTCGTATCGTACGCCCCCCCCGGCGCCGATCCGCCGAAAAAGCGAAAAGAGGCGATGCTCGGTACGCACGTTATATGGGATATCGGGTTGCAATCGTCGTGTACGATGGTCGTACCGTGGATCTCGAATACGACGTATCGACAAACGATCGACGATTCGTTTACCGAAGGCGGATATATATCGGTTTTTTATCAAACGCGAATCGTCGTACCGTTATCGACGCCTCGCGAAATGGATATACTCGGGTTCGTATCCGCTTGCAACGATTTTTCGGTACGATTATTACGCGATACGACGCATATCGAACAAAAAGCGCTCGCGCAAGGTCTCGGTCAAATGCTCGAATCGATGATCGACAATACGGTACGCGAAACGGTCGGTGCCGCGACGTCGCGCGACGCGTTACCGAATACCGAAGCGAGCGGTCCTACGCACTCGAAAGAGATACCGGCGCTTACCGCGGTCGAAACCGGCGCGACGAATCCGCTCGTACCAAGCGATACCGTACAAACGCGACACGTCGTACAACACCGATCGCGAAGCGAATCGTCGATCGAAAGTTTTTTCGCGAGGGGCGCTTGCGTTACGATCATGACGGTCGACAATCCCGCGTCGACGACGAATAAGGATAAACTATTCGCGGTATGGAAGATTACGTATAAGGATACCGTACAATTACGACGAAAGCTCGAATTTTTTACGTATTCGCGATTCGATATGGAACTTACGTTCGTCGTTACCGCGAATTTTACCGAAACGAACAACGGTCACGCACTCAATCAGGTATATCAAATTATGTACGTACCGCCCGGCGCACCGGTACCCGAAAAATGGGACGATTATACGTGGCAAACGAGCTCGAATCCGTCGATTTTTTATACGTACGGTACCGCACCCGCTCGCATAAGCGTACCGTACGTCGGTATCTCGAACGCGTATTCGCATTTTTACGACGGATTCTCGAAGGTACCGCTTAAGGATCAATCCGCCGCACTCGGCGATTCGCTATACGGTGCCGCGAGTTTAAACGATTTCGGTATACTCGCGGTACGCGTCGTAAACGATCACAATCCGACGAAGGTTACGTCGAAAATTCGCGTATATCTTAAACCGAAACATATACGCGTATGGTGTCCGCGACCGCCTCGCGCGGTCGCGTATTACGGACCGGGGGTCGATTATAAAGACGGTACGCTTACGCCGTTATCGACGAAAGATCTTACGACGTAC\n",
      "total number of codon diferences between original and recoded sequences 613\n",
      "AA sequence remains the same, good\n",
      "num desired codon pairs in original sequence: 76\n",
      "num desired codon pairs in recoded sequence: 742\n",
      "869\n",
      "-30.52999999999996\n",
      "frame1 codon pair score of the original sequence: -0.03513233601841192\n",
      "869\n",
      "-409.1919999999999\n",
      "frame1 codon pair score of the recoded sequence: -0.47087686996547745\n",
      "frame2&3 indep codon pair score of the original sequence: -31.61352552200005\n",
      "frame2&3 indep codon pair score of the recoded sequence: -787.6334876509989\n",
      "CAI of the original sequence is: 0.735171695523\n",
      "CAI of the recoded sequence is: 0.563999485139\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "#import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "\n",
    "\n",
    "# number of desired codon paris in sequence of any length - in this case, only in frame 2 and 3\n",
    "def num_costly_codon_pairs(seq):\n",
    "    num_indep = 0\n",
    "    # now do the independent scoring, frame 2\n",
    "    for i in range(1,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] in pairs_to_maximize_indep):\n",
    "            num_indep += 1\n",
    "    # now do the independent scoring, frame 3\n",
    "    for i in range(2,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] in pairs_to_maximize_indep):\n",
    "            num_indep += 1\n",
    "    return(num_indep)\n",
    "\n",
    "# codon pair score of any length sequence\n",
    "def get_cps_indep(seq):\n",
    "    cps_indep = 0 \n",
    "    # for indep_dependent pairs scores - frame 2\n",
    "    for i in range(1,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_indep):\n",
    "            #print(\"Missing \" + seq[i:i+6] + \" from independent scores\")\n",
    "            continue\n",
    "        cps_indep += cp_scores_indep[seq[i:i+6]]\n",
    "    # for indep_dependent pairs scores - frame 3\n",
    "    for i in range(2,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_indep):\n",
    "            #print(\"Missing \" + seq[i:i+6] + \" from independent scores\")\n",
    "            continue\n",
    "        cps_indep += cp_scores_indep[seq[i:i+6]]\n",
    "    return(cps_indep)\n",
    "\n",
    "\n",
    "# codon adaptation index of any length sequence\n",
    "def get_cai(seq):\n",
    "    # get the CAI as the geometric mean of all the codon's individual CAI values\n",
    "    cai_list = []\n",
    "    for i in range(0,len(seq),3):\n",
    "        cai_list.append(codon_CAI[seq[i:(i+3)]])\n",
    "    return(scistats.gmean(cai_list))\n",
    "\n",
    "# read in synonymous codons\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "\n",
    "# read in the human orfs and make the CAI for all the codons\n",
    "codon_occs = dict()\n",
    "# set up the dict\n",
    "for codon in (codon_to_aa_dict.keys()):\n",
    "    codon_occs[codon] = 0\n",
    "genes_file = open(\"human_CCDS_nucleotide.current.fna\")\n",
    "for line in genes_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\": \n",
    "        if (len(line)%3 != 0):\n",
    "            #print(\"PROBLEM, genes file is not multiple of 3 per line.\")\n",
    "            continue\n",
    "        # get the frequency of each codon\n",
    "        for i in range(0,len(line),3):\n",
    "            codon_occs[line[i:(i+3)]] += 1\n",
    "\n",
    "# get the CAI of each codon\n",
    "codon_CAI = dict()\n",
    "for aa in (aa_to_syn_codons_dict.keys()):\n",
    "    highest_frequency = 0\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (codon_occs[codon] > highest_frequency):\n",
    "            highest_frequency = codon_occs[codon]\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (highest_frequency == 0):\n",
    "            codon_CAI[codon] = 0\n",
    "        else:\n",
    "            codon_CAI[codon] = codon_occs[codon]/highest_frequency\n",
    "            \n",
    "\n",
    "# read in the gene's codon sequence\n",
    "codons_file = open(\"polio_genomics.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "print(\"original codons\")\n",
    "print(original_codons)\n",
    "#original_codons = \"GGCGCACATGAA\"\n",
    "\n",
    "original_AA = \"\"\n",
    "for i in range(0,len(original_codons),3):\n",
    "    original_AA += codon_to_aa_dict[original_codons[i:i+3]]\n",
    "\n",
    "#print(original_AA)\n",
    "\n",
    "# store_best will contain, at each AA position, and at each of the syn codon choices, and the following AA's\n",
    "# syn codon choices, \n",
    "# the best cost of the prefix that starts with this codon, the current codon pair score,\n",
    "# and the next codon that corresponds to this cost\n",
    "store_best = [None] * (len(original_AA)-1)\n",
    "for i in range(0,len(original_AA)-1):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = dict()\n",
    "        for codon_next in aa_to_syn_codons_dict[original_AA[i+1]]:\n",
    "            store_best[i][codon][codon_next] = 0\n",
    "            \n",
    "# initialize the dynamic program by setting store_best for the first amino acid position\n",
    "for codon in store_best[0]:\n",
    "    for codon_next in store_best[1]:\n",
    "        store_best[0][codon][codon_next] = (0,0,\"\",codon)\n",
    "\n",
    "# do the dynamic program, starting from the beginning\n",
    "for pos in range(1, len(store_best)):\n",
    "    for codon in store_best[pos]:\n",
    "        for codon_next in store_best[pos][codon]:\n",
    "            # consider as cost the cost of adding this codon to one of the best sequences\n",
    "            # from the previous suffix\n",
    "            best_prev = \"\"\n",
    "            best_cps = sys.maxsize # keep this\n",
    "            best_cost = -1 #sys.maxsize #-1\n",
    "            best_seq = \"\"\n",
    "            for prev_codon in store_best[pos-1]:\n",
    "                cost = store_best[pos-1][prev_codon][codon][0] + num_costly_codon_pairs(prev_codon + codon + codon_next) \n",
    "                cp_score = get_cps_indep(store_best[pos-1][prev_codon][codon][3] + codon + codon_next)\n",
    "                #print(str(prev_codon))\n",
    "                if ((cost > best_cost) or # > <\n",
    "                    # break the tie with lowest codon pair score\n",
    "                    ((cost == best_cost) and (cp_score < best_cps))):\n",
    "                    best_cost = cost\n",
    "                    best_cps = cp_score\n",
    "                    best_prev = prev_codon\n",
    "                    best_seq = store_best[pos-1][prev_codon][codon][3] + codon\n",
    "            store_best[pos][codon][codon_next] = (best_cost, best_cps, best_prev, best_seq)\n",
    "            #print(\"storing for \" + str(pos) + \" \" + codon +  \" \" + codon_next + \" : \" + str(best_cps) + \" \" + best_prev + \" \" + best_seq)\n",
    "            #GGC GCA CAT GAA\n",
    "\n",
    "# do the final position\n",
    "best_cps = sys.maxsize # keep this\n",
    "best_cost = -1 #sys.maxsize #-1\n",
    "best_seq = \"\"\n",
    "for codon in aa_to_syn_codons_dict[original_AA[-1]]:\n",
    "    for prev_codon in aa_to_syn_codons_dict[original_AA[-2]]:\n",
    "        cost = store_best[-1][prev_codon][codon][0]\n",
    "        cp_score = store_best[-1][prev_codon][codon][1]\n",
    "        if ((cost > best_cost) or # > <\n",
    "            # break the tie with lowest codon pair score\n",
    "            ((cost == best_cost) and (cp_score < best_cps))):\n",
    "            best_cost = cost\n",
    "            best_cps = cp_score\n",
    "            best_seq = store_best[-1][prev_codon][codon][3] + codon\n",
    "#print(str(best_cps) + \" \" + str(best_cost) + \" \" + best_seq)\n",
    "\n",
    "recoded_sequence = best_seq\n",
    "print(\"recoded sequence\")\n",
    "print(recoded_sequence)\n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "#print(len(original_codons))\n",
    "print(\"num desired codon pairs in original sequence: \" + str(num_costly_codon_pairs(original_codons)))\n",
    "print(\"num desired codon pairs in recoded sequence: \" + str(num_costly_codon_pairs(recoded_sequence)))\n",
    "print(\"frame1 codon pair score of the original sequence: \" + str(get_cpb(original_codons)))\n",
    "print(\"frame1 codon pair score of the recoded sequence: \" + str(get_cpb(recoded_sequence)))\n",
    "print(\"frame2&3 indep codon pair score of the original sequence: \" + str(get_cps_indep(original_codons)))\n",
    "print(\"frame2&3 indep codon pair score of the recoded sequence: \" + str(get_cps_indep(recoded_sequence)))\n",
    "print(\"CAI of the original sequence is: \" + str(get_cai(original_codons)))\n",
    "print(\"CAI of the recoded sequence is: \" + str(get_cai(recoded_sequence)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGCGAA -1.45019984 -0.249640625\n",
      "TCGCGA -1.417669026 -0.31936197916666664\n",
      "CGATCG -1.167580764 -0.209578125\n",
      "CGAACG -1.123957937 -0.17597135416666665\n",
      "ACGCGA -1.119915209 -0.24942187500000002\n",
      "CGCGAT -1.099095148 -0.209578125\n",
      "GCGAAA -1.098748918 -0.25395833333333334\n",
      "GCGAAC -1.019916026 -0.2770130208333333\n",
      "CGGTCG -1.019057144 0.05541145833333333\n",
      "CGCGTA -1.011012033 -0.19433333333333333\n",
      "CGCAAT -1.000669928 -0.14427864583333333\n",
      "CGTCGA -0.980861058 -0.17250520833333333\n",
      "GTTGCG -0.97246442 -0.1947421875\n",
      "TCGATC -0.965938171 -0.2586822916666667\n",
      "GCGATC -0.964832724 -0.2769453125\n",
      "TCGCGT -0.962076223 -0.2776484375\n",
      "TTTCGC -0.961382801 -0.3334270833333334\n",
      "ACGATC -0.9359366 -0.22278645833333335\n",
      "TTGCGA -0.929889696 -0.2578697916666667\n",
      "CAATCG -0.921598731 -0.19277604166666668\n",
      "CGACGA -0.916926437 -0.22066145833333334\n",
      "GTCGAA -0.915712283 -0.12534635416666667\n",
      "CGCGAC -0.914552851 -0.2053046875\n",
      "CCGATC -0.911151013 -0.2724010416666667\n",
      "TCGCAA -0.905443572 -0.3279296875\n",
      "CGATCA -0.905085672 -0.19709114583333334\n",
      "TATACG -0.902349388 -0.09743229166666666\n",
      "CGTACG -0.900977206 -0.06487239583333333\n",
      "GGCGAA -0.892319621 -0.205375\n",
      "CGATTG -0.88175734 -0.2679142857142857\n",
      "TACGCG -0.880597524 -0.16180729166666666\n",
      "CCCCCC -0.878485479 -0.07998697916666667\n",
      "TTACGC -0.878110625 -0.34165364583333335\n",
      "GCGCGA -0.876586165 -0.25161197916666667\n",
      "CTCGCG -0.871737206 -0.13719791666666667\n",
      "GTCGCG -0.871294886 -0.12752604166666667\n",
      "TTCGCG -0.870137249 -0.1713125\n",
      "TTTTCG -0.867468853 -0.09694010416666667\n",
      "GCGCAA -0.866754702 -0.297171875\n",
      "CGAAAA -0.866649963 -0.08483854166666667\n",
      "GTCGAT -0.866254896 -0.15670052083333333\n",
      "CGCATA -0.865676932 -0.18130555555555558\n",
      "CGATCC -0.86039943 -0.19570572916666665\n",
      "CGATCT -0.859830773 -0.191953125\n",
      "CGCAAC -0.859393431 -0.15087760416666665\n",
      "CGGTAT -0.856793082 -0.06505208333333334\n",
      "ATACGC -0.854391445 -0.29779947916666666\n",
      "ATTCGC -0.853178998 -0.28773177083333334\n",
      "TCGAAC -0.849952808 -0.2868984375\n",
      "ATTGCG -0.843370589 -0.171984375\n",
      "AACGCG -0.837058204 -0.045744047619047615\n",
      "TCTCGA -0.836762457 -0.21866927083333332\n",
      "ACGAAC -0.836238349 -0.21863281250000002\n",
      "ACGATA -0.834683674 -0.28795833333333337\n",
      "CGATAC -0.830396789 -0.17048177083333335\n",
      "ACCGGT -0.828823641 -0.0293984375\n",
      "CGATTA -0.828171432 -0.20744444444444443\n",
      "GGTCGA -0.827730444 -0.2930052083333333\n",
      "GCGTAC -0.82443875 -0.2961380208333333\n",
      "GACGAA -0.820176075 -0.1356845238095238\n",
      "GGGGGG -0.819427113 -0.06570052083333335\n",
      "CGAACC -0.817142536 -0.183140625\n",
      "GTCGTA -0.813618267 -0.1408888888888889\n",
      "ATACCG -0.811260311 -0.05411197916666666\n",
      "CGGTAC -0.810668082 -0.08481249999999999\n",
      "GGCGTA -0.808751453 -0.1637222222222222\n",
      "GCGATT -0.808044258 -0.23370572916666668\n",
      "ATCGCG -0.806931822 -0.1600390625\n",
      "CGATAT -0.799124163 -0.1322473958333333\n",
      "CGAACT -0.798648468 -0.12486979166666667\n",
      "ACGCAA -0.797187115 -0.24211458333333336\n",
      "TTTACG -0.794600675 -0.09968229166666667\n",
      "TTGCGC -0.791201358 -0.22744270833333333\n",
      "TCGACG -0.791135063 -0.17369010416666666\n",
      "ATTTCG -0.790350313 -0.08655989583333333\n",
      "AGCGTA -0.789005827 -0.1412\n",
      "GCGTAT -0.788354339 -0.2926380208333333\n",
      "CCGAAC -0.782894062 -0.26316406249999996\n",
      "ACGAAA -0.781519231 -0.195578125\n",
      "GTCGAC -0.778657438 -0.1061328125\n",
      "ATTACG -0.778379119 -0.0753359375\n",
      "CATACG -0.772762113 -0.145859375\n",
      "CGAAAC -0.768311205 -0.11464583333333334\n",
      "CGAACA -0.76763744 -0.11897916666666666\n",
      "CGTATA -0.766771987 -0.1072638888888889\n",
      "ACGCGT -0.766560048 -0.20770833333333333\n",
      "GACGTA -0.766162954 -0.08425\n",
      "CTATCG -0.764763829 -0.15597916666666667\n",
      "TTGCGT -0.763052618 -0.2587552083333333\n",
      "ACGATT -0.761669233 -0.179546875\n",
      "TCGATT -0.758802233 -0.21544270833333334\n",
      "CCGCGA -0.758529716 -0.3027213541666667\n",
      "GGTACG -0.758423112 -0.0011822916666666683\n",
      "CGCATT -0.75747816 -0.11828124999999999\n",
      "GCTCGA -0.75112043 -0.20154687500000001\n",
      "TATCGA -0.747178844 -0.23260416666666667\n",
      "TCGAAA -0.746394083 -0.26384375000000004\n",
      "TGCGAA -0.743694054 -0.22522916666666665\n",
      "GCGAAT -0.743501787 -0.24476041666666665\n",
      "CGCAAA -0.742603249 -0.09176302083333332\n",
      "TATCCG -0.737983743 0.01030208333333333\n",
      "ATACGA -0.73656592 -0.29129166666666667\n",
      "CTCGAA -0.736331362 -0.086046875\n",
      "CCCGTA -0.733049906 -0.07783333333333334\n",
      "CGAATT -0.731446611 -0.0867578125\n",
      "TACGTA -0.724940115 -0.0605\n",
      "ATATCG -0.724269398 -0.11784375\n",
      "TCGCAT -0.720440475 -0.27670833333333333\n",
      "CGATTT -0.720072322 -0.14861979166666667\n",
      "CGTACT -0.712670238 -0.10007812499999999\n",
      "TATGCG -0.712033816 -0.1265078125\n",
      "CTCGAT -0.711580173 -0.11740104166666666\n",
      "GATACG -0.711483676 -0.06140773809523809\n",
      "CTCGTA -0.707994817 -0.12158333333333333\n",
      "CGGTTA -0.707709403 -0.046152777777777786\n",
      "GTCGAG -0.706568061 -0.1234765625\n",
      "CGGATC -0.702341167 -0.052734375\n",
      "AACGGT -0.701976497 -0.06570535714285715\n",
      "CGAAAG -0.700170064 -0.07491927083333336\n",
      "AATTCG -0.700015247 0.014053571428571431\n",
      "CTCGAC -0.699851926 -0.06683333333333333\n",
      "TATCGG -0.696509757 -0.21630208333333334\n",
      "GATTCG -0.695323256 0.009910714285714285\n",
      "CGAAAT -0.694644091 -0.0934140625\n",
      "ACTCGA -0.694437644 -0.1757734375\n",
      "GCGAAG -0.694077907 -0.2179739583333333\n",
      "GTCGCA -0.692763542 -0.18024479166666668\n",
      "TCACGA -0.692298844 -0.08117187499999999\n",
      "TACGAT -0.691654058 -0.0754921875\n",
      "TCCGGT -0.691535641 -0.01089583333333334\n",
      "TTCGCA -0.691404434 -0.22403125000000002\n",
      "GATCCG -0.690445004 0.04369047619047618\n",
      "CGACCG -0.689764123 -0.040000000000000015\n",
      "TACGAA -0.682745812 -0.0956796875\n",
      "ATGCGA -0.681125274 -0.2186640625\n",
      "CGAATC -0.680357732 -0.12329427083333334\n",
      "CGGATT -0.679313307 0.020263020833333333\n",
      "GTATCG -0.678083349 -0.12627604166666664\n",
      "AATCCG -0.677727174 0.04665476190476191\n",
      "TCGCGC -0.674797025 -0.24816927083333334\n",
      "CTTGCG -0.674598312 -0.1547734375\n",
      "TATCGC -0.673265729 -0.21477864583333334\n",
      "GCCGAT -0.67253484 -0.07879166666666666\n",
      "GGCGAT -0.671616692 -0.1653125\n",
      "TATCGT -0.670632502 -0.22755468750000002\n",
      "GACGAT -0.670112287 -0.11261309523809525\n",
      "CAACCG -0.669701249 -0.06785416666666666\n",
      "GGTCGC -0.667676129 -0.26690364583333337\n",
      "GTTACG -0.667635995 -0.07588020833333334\n",
      "CGCAAG -0.667324662 -0.15377864583333334\n",
      "TTACCG -0.667316479 -0.05859114583333333\n",
      "CGCGTT -0.667257624 -0.16143229166666667\n",
      "CCGATT -0.666922471 -0.22916145833333332\n",
      "ATCGTA -0.663021306 -0.15243055555555554\n",
      "CGTTCG -0.662395341 -0.04714062499999999\n",
      "CGGTTG -0.661536239 -0.10694285714285715\n",
      "CGGATA -0.659907794 -0.011444444444444445\n",
      "CGCGAG -0.659863933 -0.22220052083333333\n",
      "TGCGTA -0.659040716 -0.1825138888888889\n",
      "TTTGCG -0.657816071 -0.1639296875\n",
      "AAGCGT -0.654665651 -0.2216845238095238\n",
      "ACGAAT -0.649477177 -0.18638020833333335\n",
      "GCGCAT -0.647550953 -0.24595052083333335\n",
      "AATCGC -0.647277773 -0.1728720238095238\n",
      "GATCGC -0.64568929 -0.17522321428571427\n",
      "AACGAA -0.645247612 -0.13542559523809522\n",
      "CGAATA -0.645210026 -0.1275\n",
      "AAGCGC -0.643881616 -0.21297619047619046\n",
      "AGCGAA -0.642829503 -0.1899583333333333\n",
      "AACGAT -0.641444952 -0.11235416666666666\n",
      "CGTATC -0.640416883 -0.07087760416666666\n",
      "GCCGAA -0.638175334 -0.08869270833333334\n",
      "TTCGTA -0.637430132 -0.1126111111111111\n",
      "GCGACG -0.637031913 -0.1523151041666667\n",
      "CGATTC -0.636906415 -0.1484661458333333\n",
      "GCCGTA -0.636309734 -0.021569444444444447\n",
      "GTACCG -0.635563643 -0.03542708333333333\n",
      "ACCGAT -0.632950723 -0.06959375\n",
      "AATCGT -0.631775805 -0.18747321428571428\n",
      "CGCTTA -0.631222575 -0.11076388888888888\n",
      "GTTTCG -0.627707007 -0.0909375\n",
      "ACGGAT -0.627413463 -0.24017968750000002\n",
      "AAGCGA -0.627131387 -0.21854761904761905\n",
      "TGCGAT -0.626607604 -0.18516666666666667\n",
      "AATGCG -0.62621993 -0.04480357142857143\n",
      "TATTCG -0.626206651 -0.06710416666666666\n",
      "TTATCG -0.624635132 -0.13723697916666666\n",
      "ATCTCG -0.61906613 -0.08194791666666666\n",
      "AATCGA -0.61868103 -0.19324404761904762\n",
      "ACGTAT -0.616777189 -0.2194505208333333\n",
      "AACGTA -0.612838583 -0.08168333333333333\n",
      "TACGAC -0.610775425 -0.06023437499999999\n",
      "ACGCAT -0.609831156 -0.19089322916666668\n",
      "CATTCG -0.60868362 -0.07549739583333333\n",
      "TCCGAT -0.605501203 -0.040872395833333325\n",
      "AATCGG -0.605085131 -0.17461309523809523\n",
      "CAAGCG -0.604947313 -0.13703645833333333\n",
      "TACGCA -0.604571005 -0.20394791666666665\n",
      "CGTACA -0.604129436 -0.040291666666666656\n",
      "CGTATT -0.603756409 -0.054049479166666664\n",
      "CCGAAA -0.602828887 -0.24010937499999999\n",
      "CGTTAT -0.60169003 -0.105890625\n",
      "CTTTCG -0.601143038 -0.08147916666666667\n",
      "TTGCGG -0.600073435 -0.23298697916666666\n",
      "TCAATC -0.599221955 -0.07465104166666667\n",
      "CTACCG -0.598610362 -0.059408854166666664\n",
      "GATCGG -0.597636135 -0.1769642857142857\n",
      "CGCCGA -0.59662908 -0.09093489583333333\n",
      "CGACGG -0.594949701 -0.198625\n",
      "ACGGGT -0.594374865 -0.13532291666666668\n",
      "ACCGTA -0.593964725 -0.030680555555555558\n",
      "ACGTAC -0.593557281 -0.2229505208333333\n",
      "ACGACG -0.593256231 -0.13084895833333332\n",
      "ATGCGC -0.592068068 -0.18823697916666665\n",
      "CACGAA -0.590640037 -0.18306510416666666\n",
      "TACGGT -0.590506184 -0.12508854166666666\n",
      "CCCGAT -0.589476139 -0.08244791666666666\n",
      "TCTCGC -0.589072432 -0.19313020833333333\n",
      "CCGTCG -0.589001469 -0.11215104166666667\n",
      "TCCGTA -0.587805836 -0.001902777777777781\n",
      "TACTCG -0.585813065 -0.08129166666666668\n",
      "GATCGA -0.585184947 -0.1955952380952381\n",
      "GAATCG -0.584193603 -0.045601190476190476\n",
      "CGCTCG -0.584112592 0.02601302083333333\n",
      "TTGTCG -0.583719564 0.011614583333333331\n",
      "CGCTTG -0.581416694 -0.12626666666666667\n",
      "GGATCG -0.579672519 -0.0420390625\n",
      "CGTACC -0.579550598 -0.07738541666666666\n",
      "CTCGCA -0.576896489 -0.18991666666666668\n",
      "AACGGA -0.574265866 -0.022110119047619052\n",
      "AAACGA -0.573292545 -0.21076190476190476\n",
      "CGGCGA -0.571720223 -0.020611979166666683\n",
      "CTTACG -0.571492464 -0.07298177083333335\n",
      "GCCGTT -0.569394807 -0.05676302083333333\n",
      "AAACGC -0.568858516 -0.19854761904761906\n",
      "GCAATC -0.567771146 -0.08362760416666667\n",
      "AAAGCG -0.567714199 -0.025601190476190475\n",
      "ATCGAA -0.563733121 -0.12376302083333333\n",
      "TTTTTT -0.563294564 -0.07588541666666666\n",
      "CGTTAC -0.562467046 -0.14003125\n",
      "AAACGG -0.561955326 -0.1924880952380952\n",
      "CGCTAT -0.559937099 -0.10969270833333333\n",
      "CGGAAT -0.55926254 0.014989583333333334\n",
      "GTCGGT -0.559166331 -0.12873177083333334\n",
      "ACGTCG -0.55903623 -0.067765625\n",
      "CGGTCT -0.558942808 0.06666666666666667\n",
      "ACTCGT -0.557256348 -0.13168229166666667\n",
      "CGCGGT -0.557017127 -0.018255208333333335\n",
      "ACGAAG -0.556973808 -0.15959375\n",
      "TACGGA -0.555718145 -0.08694270833333334\n",
      "GTCGTT -0.554525372 -0.1067421875\n",
      "TCTCGT -0.55289452 -0.174578125\n",
      "GCGACA -0.551205685 -0.1205234375\n",
      "TACGCT -0.550370651 -0.1749296875\n",
      "GTGCGA -0.549288936 -0.21018489583333333\n",
      "ACCGGA -0.547418443 0.010213541666666673\n",
      "AGGTCG -0.546936969 0.007428571428571428\n",
      "ATTCCG -0.546524701 0.017437499999999998\n",
      "AGACGA -0.545568212 -0.19163690476190479\n",
      "CTCGAG -0.54454239 -0.08417708333333335\n",
      "ACCGAA -0.544365031 -0.07949479166666668\n",
      "ATCGAT -0.54432928 -0.1551171875\n",
      "TTCGAA -0.544117806 -0.08430989583333333\n",
      "CGACTA -0.542262899 -0.14704166666666665\n",
      "GCCGGT -0.541843656 -0.015739583333333335\n",
      "TCGGGT -0.540238237 -0.15357291666666667\n",
      "CGTATG -0.538673722 -0.1332761904761905\n",
      "GGTCCG -0.538414308 0.011570312499999999\n",
      "CACGTA -0.538412547 -0.12318055555555556\n",
      "TCCGAA -0.538011158 -0.050773437500000004\n",
      "TTTCCG -0.536413889 0.011723958333333333\n",
      "TTCGAT -0.535435664 -0.1156640625\n",
      "CGGGTT -0.534055619 -0.08259114583333334\n",
      "GAACGC -0.533955195 -0.20368452380952382\n",
      "CGTCGC -0.533846267 -0.14640364583333337\n",
      "AAAACG -0.533298607 -0.02416964285714286\n",
      "AGCGGT -0.532615161 -0.06406250000000001\n",
      "GACGGT -0.531291565 -0.06475892857142858\n",
      "CGGTTT -0.530997321 -0.03157552083333334\n",
      "TCGACC -0.530988184 -0.18186458333333333\n",
      "TCGTCG -0.530663057 -0.09528645833333332\n",
      "AGTACG -0.530522099 0.010011904761904762\n",
      "ATCACG -0.53016358 -0.11221354166666667\n",
      "GGTTCG -0.529211484 0.03554166666666667\n",
      "TACGCC -0.528982273 -0.15784114583333334\n",
      "GGGTCG -0.52833583 0.022479166666666665\n",
      "AAAAAA -0.527626688 0.01432738095238095\n",
      "CGCACT -0.526984773 -0.02221875\n",
      "GGGGTA -0.526550852 -0.12102777777777778\n",
      "AACGCA -0.525989329 -0.09390476190476191\n",
      "CGACGT -0.52551861 -0.21014322916666667\n",
      "AATACG -0.524865591 -0.0550327380952381\n",
      "GCGACC -0.522678795 -0.16048958333333332\n",
      "CGACAA -0.521904742 -0.1065\n",
      "CGAGCG -0.520824676 -0.11147135416666668\n",
      "CGAAGT -0.51805124 -0.04194791666666666\n",
      "CACGAT -0.516461362 -0.16287760416666666\n",
      "AAACCG -0.515602184 0.025375\n",
      "AAGCGG -0.514972406 -0.2004375\n",
      "CGGGTA -0.514085068 -0.13095833333333332\n",
      "CAATCT -0.510309309 -0.16516406250000001\n",
      "TACGTT -0.509253251 -0.045924479166666664\n",
      "AGTCCG -0.508981178 0.03849107142857143\n",
      "CGAAGG -0.508142261 -0.05625520833333332\n",
      "TCGAGC -0.505444506 -0.10757291666666667\n",
      "GCATAC -0.504369722 -0.0448828125\n",
      "306\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "#import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "\n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"human_cps_coleman.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter='\\t')\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[1]] = float(line[5])\n",
    "\n",
    "\n",
    "cp_scores_indep = dict()\n",
    "cp_scores_file = open(\"human_FIndep_Scores.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "counter = 0\n",
    "pairs_to_maximize_indep = []\n",
    "for line in cp_reader:\n",
    "    cp_scores_indep[line[0]] = float(line[1])\n",
    "    counter += 1\n",
    "    if (counter <= 372):\n",
    "        if (line[0][2] == \"T\" and line[0][3] == \"A\" and ((line[0][4] == \"A\" or line[0][4] == \"G\"))):\n",
    "            continue\n",
    "        indep = line[0]\n",
    "        cp_score = 0\n",
    "        cp_num = 0\n",
    "        for i in [\"A\", \"C\", \"G\", \"T\"]:\n",
    "            if ((i + indep[:-1]) in cp_scores):\n",
    "                cp_score += cp_scores[i + indep[:-1]]\n",
    "                cp_num += 1\n",
    "            if ((indep[1:] + i) in cp_scores):\n",
    "                cp_score += cp_scores[indep[1:] + i]\n",
    "                cp_num += 1\n",
    "            cp_score /= cp_num\n",
    "        if ((cp_score < 0) and (cp_score/cp_scores_indep[indep] > 0.4)):\n",
    "            continue\n",
    "        print(indep + \" \" + str(cp_scores_indep[indep]) + \" \" + str(cp_score))\n",
    "        pairs_to_maximize_indep.append(line[0])\n",
    "print(len(pairs_to_maximize_indep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "869\n",
      "-405.281\n",
      "-0.4663762945914845\n",
      "869\n",
      "-30.52999999999996\n",
      "-0.03513233601841192\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "recoded_sequence = \"GGCGCTCACGAAAACTCGAATCGCGCATACGGATCGTCGACGATCAATTATACGACGATCAATTACTACCGCGATTCCGCGTCGAACGCCGCGTCGAAACAGGATTTTTCGCAAGATCCGTCGAAATTTACCGAACCGATTAAAGACGTACTTATCAAAACCGCGCCGATGCTCAATTCGCCGAATATCGAAGCTTGCGGTTATTCCGATCGCGTATTGCAACTTACGCTCGGTAACTCGACGATTACGACGCAAGAAGCGGCGAATTCGGTCGTCGCGTACGGTCGATGGCCCGAATACTTACGCGATTCCGAAGCGAATCCGGTCGATCAACCGACCGAACCGGACGTCGCCGCTTGTCGATTCTATACGCTCGATACCGTATCGTGGACGAAAGAATCGCGAGGTTGGTGGTGGAAATTACCGGACGCGTTACGCGATATGGGATTATTCGGTCAAAATATGTACTATCACTATCTCGGTCGATCCGGTTATACCGTACACGTACAATGCAACGCGTCGAAATTTCATCAAGGCGCATTGGGCGTATTCGCGGTACCCGAAATGTGTCTCGCCGGCGATTCGAATACGACGACGATGCATACGTCGTATCAAAACGCGAATCCCGGCGAAAAAGGCGGTACGTTTACCGGTACGTTTACGCCGGATAACAATCAAACGTCGCCCGCTCGACGATTTTGTCCGGTCGATTATCTACTCGGAAACGGTACGTTACTCGGGAATGCCTTCGTATTTCCGCATCAAATTATCAATTTGCGTACGAACAATTGCGCAACGCTCGTATTACCGTACGTTAATTCGCTATCGATCGATTCGATGGTTAAACATAACAATTGGGGTATTGCGATATTACCGCTCGCACCGCTTAATTTCGCAAGCGAATCGTCGCCCGAAATACCGATTACGCTTACGATCGCGCCGATGTGTTGCGAATTCAACGGATTACGCAATATTACGTTACCGCGATTACAGGGATTACCGGTTATGAATACGCCCGGATCGAATCAATATCTTACCGCCGACAATTTTCAATCGCCTTGCGCATTACCGGAATTCGACGTTACGCCGCCGATCGATATACCGGGCGAAGTAAAAAACATGATGGAACTCGCCGAAATCGATACGATGATACCGTTCGATCTATCCGCGACGAAAAAAAATACGATGGAAATGTATCGCGTACGATTATCCGACAAACCGCATACCGACGATCCGATACTATGTCTATCGTTATCGCCCGCAAGCGATCCGCGATTATCGCATACGATGCTCGGCGAAATACTCAATTATTATACGCATTGGGCCGGATCGCTTAAATTTACGTTTCTTTTTTGCGGATCGATGATGGCGACCGGTAAATTACTCGTATCGTACGCCCCCCCCGGCGCCGATCCGCCGAAAAAGCGAAAAGAGGCGATGCTCGGTACGCACGTTATATGGGATATCGGATTGCAATCGTCGTGTACGATGGTCGTACCGTGGATCTCGAATACGACGTATCGACAAACGATCGACGATTCGTTTACCGAAGGCGGATATATATCGGTTTTTTATCAAACGCGAATCGTCGTACCGCTTAGTACGCCGCGCGAAATGGATATACTCGGATTCGTATCCGCTTGCAACGATTTTTCGGTACGATTATTACGCGATACGACGCATATCGAACAAAAAGCGCTAGCACAAGGTCTCGGTCAAATGCTCGAATCGATGATCGACAATACCGTACGCGAAACGGTCGGTGCCGCGACGTCGCGCGACGCGTTACCGAATACCGAAGCGAGCGGTCCTACGCACTCGAAAGAGATACCGGCGCTTACCGCGGTCGAAACCGGCGCGACGAATCCGCTCGTACCGAGCGATACCGTACAAACGCGACACGTCGTACAACACCGGTCTCGATCCGAATCGTCGATCGAATCGTTTTTCGCTCGCGGCGCTTGCGTTACGATCATGACGGTCGACAATCCGGCGTCGACGACGAATAAGGATAAACTATTCGCGGTATGGAAGATCACGTATAAGGATACCGTACAATTACGACGAAAGCTCGAATTTTTCACGTATTCTCGATTCGATATGGAACTTACGTTCGTCGTTACCGCGAATTTTACCGAAACGAACAACGGTCACGCACTCAATCAGGTATATCAAATTATGTACGTACCGCCCGGCGCACCGGTACCCGAAAAATGGGACGATTATACGTGGCAAACGAGCTCGAATCCGTCGATTTTTTACACGTACGGTACCGCACCCGCTCGTATATCCGTACCGTACGTCGGTATCTCGAACGCGTATTCGCATTTTTACGACGGATTCTCGAAGGTACCGCTTAAGGATCAATCCGCCGCACTCGGCGATTCGCTATACGGTGCCGCGAGTTTAAACGATTTCGGTATACTCGCGGTACGCGTCGTTAACGATCACAATCCGACGAAGGTTACGTCGAAAATTCGCGTATATCTTAAACCGAAACATATACGCGTATGGTGTCCGCGACCGCCTCGCGCGGTCGCGTATTACGGACCGGGGGTCGATTATAAAGACGGTACGCTTACGCCGTTATCGACGAAAGATCTTACGACGTAC\"\n",
    "original_codons = \"GGCGCACATGAAAACTCAAATAGAGCGTATGGTAGTTCTACCATTAATTACACCACCATTAATTATTATAGAGATTCAGCTAGTAACGCGGCTTCGAAACAGGACTTCTCTCAAGACCCTTCCAAGTTCACCGAGCCCATCAAGGATGTCCTGATAAAAACAGCCCCAATGCTAAACTCGCCAAACATAGAGGCTTGCGGGTATAGCGATAGAGTACTGCAATTAACACTGGGAAACTCCACTATAACCACACAGGAGGCGGCTAATTCAGTAGTCGCTTATGGGCGTTGGCCTGAATATCTGAGGGACAGCGAAGCCAATCCAGTGGACCAGCCGACAGAACCAGACGTCGCTGCATGCAGGTTTTATACGCTAGACACCGTGTCTTGGACGAAAGAGTCGCGAGGGTGGTGGTGGAAGTTGCCTGATGCACTGAGGGACATGGGACTCTTTGGGCAAAATATGTACTACCACTACCTAGGTAGGTCCGGGTACACCGTGCATGTACAGTGTAACGCCTCCAAATTCCACCAGGGGGCACTAGGGGTATTCGCCGTACCAGAGATGTGTCTGGCCGGGGATAGCAACACCACTACCATGCACACCAGCTATCAAAATGCCAATCCTGGCGAGAAAGGAGGCACTTTCACGGGTACGTTCACTCCTGACAACAACCAGACATCACCTGCCCGCAGGTTCTGCCCGGTGGATTACCTCCTTGGAAATGGCACGTTGTTGGGGAATGCCTTTGTGTTCCCGCACCAGATAATAAACCTACGGACCAACAACTGTGCTACACTGGTACTCCCTTACGTGAACTCCCTCTCGATAGATAGTATGGTAAAGCACAATAATTGGGGAATTGCAATATTACCATTGGCCCCATTAAATTTTGCTAGTGAGTCCTCCCCAGAGATTCCAATCACCTTGACCATAGCCCCTATGTGCTGTGAGTTCAATGGATTAAGAAACATCACCCTGCCACGCTTACAGGGCCTGCCGGTCATGAACACCCCTGGTAGCAATCAATATCTTACTGCAGACAACTTCCAGTCACCGTGTGCGCTGCCTGAATTTGATGTGACCCCACCTATTGACATACCCGGTGAAGTAAAGAACATGATGGAATTGGCAGAAATCGACACCATGATTCCCTTTGACTTAAGTGCCACAAAAAAGAACACCATGGAAATGTATAGGGTTCGGTTAAGTGACAAACCACATACAGACGATCCCATACTCTGCCTGTCACTCTCTCCAGCCTCAGATCCTAGGTTGTCACATACTATGCTTGGAGAAATCCTAAATTACTACACACACTGGGCAGGATCCCTGAAGTTCACGTTTCTGTTCTGTGGATCCATGATGGCAACTGGCAAACTGTTGGTGTCATACGCGCCTCCTGGAGCCGACCCACCAAAGAAGCGTAAGGAGGCGATGTTGGGAACACATGTGATCTGGGACATAGGACTGCAGTCCTCATGTACTATGGTAGTGCCATGGATTAGCAACACCACGTATCGGCAAACCATAGATGATAGTTTCACCGAAGGCGGATACATCAGCGTCTTCTACCAAACTAGAATAGTCGTCCCTCTTTCGACACCCAGAGAGATGGACATCCTTGGTTTTGTGTCAGCGTGTAATGACTTCAGCGTGCGCTTGTTGCGAGATACCACACATATAGAGCAAAAAGCGCTAGCACAGGGGTTAGGTCAGATGCTTGAAAGCATGATTGACAACACAGTCCGTGAAACGGTGGGGGCGGCAACATCTAGAGACGCTCTCCCAAACACTGAAGCCAGTGGACCAACACACTCCAAGGAAATTCCGGCACTCACCGCAGTGGAAACTGGGGCCACAAATCCACTAGTCCCTTCTGATACAGTGCAAACCAGACATGTTGTACAACATAGGTCAAGGTCAGAGTCTAGCATAGAGTCTTTCTTCGCGCGGGGTGCATGCGTGACCATTATGACCGTGGATAACCCAGCTTCCACCACGAATAAGGATAAGCTTTTTGCAGTGTGGAAGATCACTTATAAAGATACTGTCCAGTTACGGAGGAAATTGGAGTTCTTCACCTATTCTAGATTTGATATGGAACTTACCTTTGTGGTTACTGCAAATTTCACTGAGACTAACAATGGCCATGCATTAAATCAAGTGTACCAAATTATGTACGTACCACCAGGCGCTCCAGTGCCCGAAAAATGGGACGACTACACATGGCAAACCTCATCAAATCCATCAATCTTTTACACCTACGGGACAGCTCCAGCCCGGATCTCGGTACCGTATGTTGGTATTTCGAACGCCTATTCACACTTTTACGACGGTTTTTCCAAAGTACCACTGAAGGACCAGTCGGCAGCACTAGGTGACTCCCTTTATGGTGCAGCATCTCTAAATGACTTCGGTATTTTGGCTGTTAGAGTAGTCAATGATCACAACCCGACCAAGGTCACCTCCAAAATCAGAGTGTATCTAAAACCCAAACACATCAGAGTCTGGTGCCCGCGTCCACCGAGGGCAGTGGCGTACTACGGCCCTGGAGTGGATTACAAGGATGGTACGCTTACACCCCTCTCCACCAAGGATCTGACCACATAT\"\n",
    "def get_cpb(seq):\n",
    "    cps = 0\n",
    "    num_pairs = 0\n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        num_pairs += 1\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    print(num_pairs)\n",
    "    print(cps)\n",
    "    return((cps/num_pairs) )\n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"human_cps_coleman.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter='\\t')\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[1]] = float(line[5])\n",
    "print(get_cpb(recoded_sequence))\n",
    "print(get_cpb(original_codons))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get human orf coding fasta file for CAI - human_CCDS_nucleotide.current.fna\n",
    "# read in human frame dep and indep scores\n",
    "polio_genomics.fsa human_FD_Scores.csv human_FIndep_Scores.csv\n",
    "\n",
    "# do a dynamic program to score the orfs for frame_independent over two non-coding frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NruiI [399]\n",
      "NheI [1697]\n",
      "SnaBI  [2181]\n",
      "BsmI  [740]\n",
      "Making a new iteration of the loop!!!!\n",
      "Adding original NruiI site\n",
      "68 TAT ['TAC', 'TAT']\n",
      "862 ATC ['ATA', 'ATC', 'ATT']\n",
      "1607 CCT ['CCA', 'CCC', 'CCG', 'CCT']\n",
      "1917 TCG ['AGC', 'AGT', 'TCA', 'TCC', 'TCG', 'TCT']\n",
      "2088 TCG ['AGC', 'AGT', 'TCA', 'TCC', 'TCG', 'TCT']\n",
      "Adding original NheI site\n",
      "775 CTA ['CTA', 'CTC', 'CTG', 'CTT', 'TTA', 'TTG']\n",
      "2036 ATT ['ATA', 'ATC', 'ATT']\n",
      "2081 TTT ['TTC', 'TTT']\n",
      "2261 TAT ['TAC', 'TAT']\n",
      "Adding original BsmI site\n",
      "Making a new iteration of the loop!!!!\n",
      "total number of codon diferences between original and recoded sequences 605\n",
      "AA sequence remains the same, good\n",
      "GGCGCTCACGAAAACTCGAATCGCGCATACGGATCGTCGACGATCAATTATACGACGATCAATTACTACCGCGATTCCGCGTCGAACGCCGCGTCGAAACAGGATTTTTCGCAAGATCCGTCGAAATTTACCGAACCGATTAAAGACGTACTTATCAAAACCGCGCCGATGCTCAATTCGCCGAATATCGAAGCTTGCGGTTATTCCGATCGCGTATTGCAACTTACGCTCGGTAACTCGACGATTACGACGCAAGAAGCGGCGAATTCGGTCGTCGCGTACGGTCGATGGCCCGAATACTTACGCGATTCCGAAGCGAATCCGGTCGATCAACCGACCGAACCGGACGTCGCCGCTTGTCGATTCTATACGCTCGATACCGTATCGTGGACGAAAGAATCGCGAGGTTGGTGGTGGAAATTACCGGACGCGTTACGCGATATGGGATTATTCGGTCAAAATATGTACTATCACTATCTCGGTCGATCCGGTTATACCGTACACGTACAATGCAACGCGTCGAAATTTCATCAAGGCGCATTGGGCGTATTCGCGGTACCCGAAATGTGTCTCGCCGGCGATTCGAATACGACGACGATGCATACGTCGTATCAAAACGCGAATCCCGGCGAAAAAGGCGGTACGTTTACCGGTACGTTTACGCCGGATAACAATCAAACGTCGCCCGCTCGACGATTTTGTCCGGTCGATTATCTACTCGGAAACGGTACGTTACTCGGGAATGCCTTCGTATTTCCGCATCAAATTATCAATTTGCGTACGAACAATTGCGCAACGCTCGTATTACCGTACGTTAATTCGCTATCGATCGATTCGATGGTTAAACATAACAATTGGGGTATTGCGATATTACCGCTCGCACCGCTTAATTTCGCAAGCGAATCGTCGCCCGAAATACCGATTACGCTTACGATCGCGCCGATGTGTTGCGAATTCAACGGATTACGCAATATTACGTTACCGCGATTACAGGGATTACCGGTTATGAATACGCCCGGATCGAATCAATATCTTACCGCCGACAATTTTCAATCGCCTTGCGCATTACCGGAATTCGACGTTACGCCGCCGATCGATATACCGGGCGAAGTAAAAAACATGATGGAACTCGCCGAAATCGATACGATGATACCGTTCGATCTATCCGCGACGAAAAAAAATACGATGGAAATGTATCGCGTACGATTATCCGACAAACCGCATACCGACGATCCGATACTATGTCTATCGTTATCGCCCGCAAGCGATCCGCGATTATCGCATACGATGCTCGGCGAAATACTCAATTATTATACGCATTGGGCCGGATCGCTTAAATTTACGTTTCTTTTTTGCGGATCGATGATGGCGACCGGTAAATTACTCGTATCGTACGCCCCCCCCGGCGCCGATCCGCCGAAAAAGCGAAAAGAGGCGATGCTCGGTACGCACGTTATATGGGATATCGGATTGCAATCGTCGTGTACGATGGTCGTACCGTGGATCTCGAATACGACGTATCGACAAACGATCGACGATTCGTTTACCGAAGGCGGATATATATCGGTTTTTTATCAAACGCGAATCGTCGTACCGCTTAGTACGCCGCGCGAAATGGATATACTCGGATTCGTATCCGCTTGCAACGATTTTTCGGTACGATTATTACGCGATACGACGCATATCGAACAAAAAGCGCTAGCACAAGGTCTCGGTCAAATGCTCGAATCGATGATCGACAATACCGTACGCGAAACGGTCGGTGCCGCGACGTCGCGCGACGCGTTACCGAATACCGAAGCGAGCGGTCCTACGCACTCGAAAGAGATACCGGCGCTTACCGCGGTCGAAACCGGCGCGACGAATCCGCTCGTACCGAGCGATACCGTACAAACGCGACACGTCGTACAACACCGGTCTCGATCCGAATCGTCGATCGAATCGTTTTTCGCTCGCGGCGCTTGCGTTACGATCATGACGGTCGACAATCCGGCGTCGACGACGAATAAGGATAAACTATTCGCGGTATGGAAGATCACGTATAAGGATACCGTACAATTACGACGAAAGCTCGAATTTTTCACGTATTCTCGATTCGATATGGAACTTACGTTCGTCGTTACCGCGAATTTTACCGAAACGAACAACGGTCACGCACTCAATCAGGTATATCAAATTATGTACGTACCGCCCGGCGCACCGGTACCCGAAAAATGGGACGATTATACGTGGCAAACGAGCTCGAATCCGTCGATTTTTTACACGTACGGTACCGCACCCGCTCGTATATCCGTACCGTACGTCGGTATCTCGAACGCGTATTCGCATTTTTACGACGGATTCTCGAAGGTACCGCTTAAGGATCAATCCGCCGCACTCGGCGATTCGCTATACGGTGCCGCGAGTTTAAACGATTTCGGTATACTCGCGGTACGCGTCGTTAACGATCACAATCCGACGAAGGTTACGTCGAAAATTCGCGTATATCTTAAACCGAAACATATACGCGTATGGTGTCCGCGACCGCCTCGCGCGGTCGCGTATTACGGACCGGGGGTCGATTATAAAGACGGTACGCTTACGCCGTTATCGACGAAAGATCTTACGACGTAC\n",
      "num desired codon pairs in original sequence: 92\n",
      "num desired codon pairs in recoded sequence: 866\n",
      "frame1 codon pair score of the original sequence: 13.719914612\n",
      "frame1 codon pair score of the recoded sequence: 3.9833412090000095\n",
      "frame2&3 indep codon pair score of the original sequence: -31.61352552200005\n",
      "frame2&3 indep codon pair score of the recoded sequence: -772.2770158389995\n",
      "CAI of the original sequence is: 0.735171695523\n",
      "CAI of the recoded sequence is: 0.563859583008\n"
     ]
    }
   ],
   "source": [
    "#print(original_codons)\n",
    "#print(recoded_sequence)\n",
    "import re\n",
    "\n",
    "fixed_sequence = recoded_sequence\n",
    "\n",
    "#NruI TCGCGA\n",
    "NruiI = [m.start() for m in re.finditer(\"TCGCGA\", original_codons)]\n",
    "print(\"NruiI \" + str(NruiI))\n",
    "\n",
    "#NheI GCTAGC\n",
    "NheI = [m.start() for m in re.finditer(\"GCTAGC\", original_codons)]\n",
    "print(\"NheI \" + str(NheI))\n",
    "\n",
    "#SnaBI TACGTA\n",
    "SnaBI = [m.start() for m in re.finditer(\"TACGTA\", original_codons)]\n",
    "print(\"SnaBI \", str(SnaBI))\n",
    "\n",
    "#PflMI CCANNNNNTGG\n",
    "for i in [m.start() for m in re.finditer(\"CCA\", original_codons)]:\n",
    "    if (original_codons[i+8:i+11] == \"TGG\"):\n",
    "        print(i)\n",
    "\n",
    "#BsmI GAATGC\n",
    "BsmI = [m.start() for m in re.finditer(\"GAATGC\", original_codons)]\n",
    "print(\"BsmI \", str(BsmI))\n",
    "\n",
    "made_changes = True\n",
    "\n",
    "while(made_changes):\n",
    "    made_changes = False\n",
    "    print(\"Making a new iteration of the loop!!!!\")\n",
    "    \n",
    "    #NruI TCGCGA\n",
    "    for loc in NruiI:\n",
    "        assert loc%3 == 0, \"NruiI is in frame0 in original\"\n",
    "        if (fixed_sequence[loc:loc+6] != \"TCGCGA\"):\n",
    "            made_changes = True\n",
    "            print(\"Adding original NruiI site\")\n",
    "            fixed_sequence = fixed_sequence[:loc] + \"TCGCGA\" + fixed_sequence[loc+6:]\n",
    "    # now try to find new instances\n",
    "    for loc in [m.start() for m in re.finditer(\"TCGCGA\", fixed_sequence)]:\n",
    "        if (loc in NruiI):\n",
    "            continue\n",
    "        made_changes = True\n",
    "        new_loc = loc - loc%3\n",
    "        old_codon = fixed_sequence[new_loc:new_loc+3]\n",
    "        syn_codons = aa_to_syn_codons_dict[codon_to_aa_dict[old_codon]]\n",
    "        new_codon = \"\"\n",
    "        print(str(loc) + \" \" + old_codon + \" \" + str(syn_codons))\n",
    "        for syn in syn_codons:\n",
    "            if (syn != old_codon):\n",
    "                new_codon = syn\n",
    "        assert new_codon != \"\", \"Can not find a substitution!\"\n",
    "        fixed_sequence = fixed_sequence[:new_loc] + new_codon + fixed_sequence[new_loc+3:]\n",
    "        \n",
    "    #NheI GCTAGC\n",
    "    for loc in NheI:\n",
    "        if (fixed_sequence[loc:loc+6] != \"GCTAGC\"):\n",
    "            made_changes = True\n",
    "            print(\"Adding original NheI site\")\n",
    "            if (loc%3 == 0):\n",
    "                fixed_sequence = fixed_sequence[:loc] + \"GCTAGC\" + fixed_sequence[loc+6:]\n",
    "            elif (loc%3 == 1):\n",
    "                fixed_sequence = fixed_sequence[:loc-1] + original_codons[loc-1:loc+8] + fixed_sequence[loc+8:]\n",
    "            elif (loc%3 == 2):\n",
    "                fixed_sequence = fixed_sequence[:loc-2] + original_codons[loc-2:loc+7] + fixed_sequence[loc+7:]\n",
    "    # now try to find new instances\n",
    "    for loc in [m.start() for m in re.finditer(\"GCTAGC\", fixed_sequence)]:\n",
    "        if (loc in NheI):\n",
    "            continue\n",
    "        made_changes = True\n",
    "        new_loc = loc - loc%3\n",
    "        old_codon = fixed_sequence[new_loc:new_loc+3]\n",
    "        syn_codons = aa_to_syn_codons_dict[codon_to_aa_dict[old_codon]]\n",
    "        new_codon = \"\"\n",
    "        print(str(loc) + \" \" + old_codon + \" \" + str(syn_codons))\n",
    "        for syn in syn_codons:\n",
    "            if (syn != old_codon):\n",
    "                new_codon = syn\n",
    "        assert new_codon != \"\", \"Can not find a substitution!\"\n",
    "        fixed_sequence = fixed_sequence[:new_loc] + new_codon + fixed_sequence[new_loc+3:]\n",
    "        \n",
    "    #SnaBI TACGTA\n",
    "    for loc in SnaBI:\n",
    "        assert loc%3 == 0, \"TACGTA is in frame0 in original\"\n",
    "        if (fixed_sequence[loc:loc+6] != \"TACGTA\"):\n",
    "            made_changes = True\n",
    "            print(\"Adding original SnaBI site\")\n",
    "            fixed_sequence = fixed_sequence[:loc] + \"TACGTA\" + fixed_sequence[loc+6:]\n",
    "    # now try to find new instances\n",
    "    for loc in [m.start() for m in re.finditer(\"TACGTA\", fixed_sequence)]:\n",
    "        if (loc in SnaBI):\n",
    "            continue\n",
    "        made_changes = True\n",
    "        new_loc = loc - loc%3\n",
    "        old_codon = fixed_sequence[new_loc:new_loc+3]\n",
    "        syn_codons = aa_to_syn_codons_dict[codon_to_aa_dict[old_codon]]\n",
    "        new_codon = \"\"\n",
    "        print(str(loc) + \" \" + old_codon + \" \" + str(syn_codons))\n",
    "        for syn in syn_codons:\n",
    "            if (syn != old_codon):\n",
    "                new_codon = syn\n",
    "        assert new_codon != \"\", \"Can not find a substitution!\"\n",
    "        fixed_sequence = fixed_sequence[:new_loc] + new_codon + fixed_sequence[new_loc+3:]\n",
    "        \n",
    "    #BsmI GAATGC\n",
    "    for loc in BsmI:\n",
    "        if (fixed_sequence[loc:loc+6] != \"GAATGC\"):\n",
    "            made_changes = True\n",
    "            print(\"Adding original BsmI site\")\n",
    "            if (loc%3 == 0):\n",
    "                fixed_sequence = fixed_sequence[:loc] + \"GAATGC\" + fixed_sequence[loc+6:]\n",
    "            elif (loc%3 == 1):\n",
    "                fixed_sequence = fixed_sequence[:loc-1] + original_codons[loc-1:loc+8] + fixed_sequence[loc+8:]\n",
    "            elif (loc%3 == 2):\n",
    "                fixed_sequence = fixed_sequence[:loc-2] + original_codons[loc-2:loc+7] + fixed_sequence[loc+7:]\n",
    "    # now try to find new instances\n",
    "    for loc in [m.start() for m in re.finditer(\"GAATGC\", fixed_sequence)]:\n",
    "        if (loc in BsmI):\n",
    "            continue\n",
    "        made_changes = True\n",
    "        new_loc = loc - loc%3\n",
    "        old_codon = fixed_sequence[new_loc:new_loc+3]\n",
    "        syn_codons = aa_to_syn_codons_dict[codon_to_aa_dict[old_codon]]\n",
    "        new_codon = \"\"\n",
    "        print(str(loc) + \" \" + old_codon + \" \" + str(syn_codons))\n",
    "        for syn in syn_codons:\n",
    "            if (syn != old_codon):\n",
    "                new_codon = syn\n",
    "        assert new_codon != \"\", \"Can not find a substitution!\"\n",
    "        fixed_sequence = fixed_sequence[:new_loc] + new_codon + fixed_sequence[new_loc+3:]\n",
    "        \n",
    "    ##PflMI CCANNNNNTGG\n",
    "    #for i in [m.start() for m in re.finditer(\"CCA\", original_codons)]:\n",
    "    #if (original_codons[i+8:i+11] == \"TGG\"):\n",
    "    # now try to find new instances\n",
    "    for loc in [m.start() for m in re.finditer(\"CCA\", fixed_sequence)]:\n",
    "        if (fixed_sequence[loc+8:loc+11] != \"TGG\"):\n",
    "            continue\n",
    "        made_changes = True\n",
    "        new_loc = loc - loc%3\n",
    "        old_codon = fixed_sequence[new_loc:new_loc+3]\n",
    "        syn_codons = aa_to_syn_codons_dict[codon_to_aa_dict[old_codon]]\n",
    "        new_codon = \"\"\n",
    "        print(str(loc) + \" \" + old_codon + \" \" + str(syn_codons))\n",
    "        for syn in syn_codons:\n",
    "            if (syn != old_codon):\n",
    "                new_codon = syn\n",
    "        assert new_codon != \"\", \"Can not find a substitution!\"\n",
    "        fixed_sequence = fixed_sequence[:new_loc] + new_codon + fixed_sequence[new_loc+3:]\n",
    "        \n",
    "fixed_back_to_original_AA = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(fixed_sequence),3):\n",
    "    if (fixed_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    fixed_back_to_original_AA += codon_to_aa_dict[fixed_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == fixed_back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "\n",
    "print(fixed_sequence)\n",
    "print(\"num desired codon pairs in original sequence: \" + str(num_costly_codon_pairs(original_codons)))\n",
    "print(\"num desired codon pairs in recoded sequence: \" + str(num_costly_codon_pairs(fixed_sequence)))\n",
    "print(\"frame1 codon pair score of the original sequence: \" + str(get_cps_frame1(original_codons)))\n",
    "print(\"frame1 codon pair score of the recoded sequence: \" + str(get_cps_frame1(fixed_sequence)))\n",
    "print(\"frame2&3 indep codon pair score of the original sequence: \" + str(get_cps_indep(original_codons)))\n",
    "print(\"frame2&3 indep codon pair score of the recoded sequence: \" + str(get_cps_indep(fixed_sequence)))\n",
    "print(\"CAI of the original sequence is: \" + str(get_cai(original_codons)))\n",
    "print(\"CAI of the recoded sequence is: \" + str(get_cai(fixed_sequence)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "311\n",
      "372\n",
      "original codons\n",
      "GGCGCACATGAAAACTCAAATAGAGCGTATGGTAGTTCTACCATTAATTACACCACCATTAATTATTATAGAGATTCAGCTAGTAACGCGGCTTCGAAACAGGACTTCTCTCAAGACCCTTCCAAGTTCACCGAGCCCATCAAGGATGTCCTGATAAAAACAGCCCCAATGCTAAACTCGCCAAACATAGAGGCTTGCGGGTATAGCGATAGAGTACTGCAATTAACACTGGGAAACTCCACTATAACCACACAGGAGGCGGCTAATTCAGTAGTCGCTTATGGGCGTTGGCCTGAATATCTGAGGGACAGCGAAGCCAATCCAGTGGACCAGCCGACAGAACCAGACGTCGCTGCATGCAGGTTTTATACGCTAGACACCGTGTCTTGGACGAAAGAGTCGCGAGGGTGGTGGTGGAAGTTGCCTGATGCACTGAGGGACATGGGACTCTTTGGGCAAAATATGTACTACCACTACCTAGGTAGGTCCGGGTACACCGTGCATGTACAGTGTAACGCCTCCAAATTCCACCAGGGGGCACTAGGGGTATTCGCCGTACCAGAGATGTGTCTGGCCGGGGATAGCAACACCACTACCATGCACACCAGCTATCAAAATGCCAATCCTGGCGAGAAAGGAGGCACTTTCACGGGTACGTTCACTCCTGACAACAACCAGACATCACCTGCCCGCAGGTTCTGCCCGGTGGATTACCTCCTTGGAAATGGCACGTTGTTGGGGAATGCCTTTGTGTTCCCGCACCAGATAATAAACCTACGGACCAACAACTGTGCTACACTGGTACTCCCTTACGTGAACTCCCTCTCGATAGATAGTATGGTAAAGCACAATAATTGGGGAATTGCAATATTACCATTGGCCCCATTAAATTTTGCTAGTGAGTCCTCCCCAGAGATTCCAATCACCTTGACCATAGCCCCTATGTGCTGTGAGTTCAATGGATTAAGAAACATCACCCTGCCACGCTTACAGGGCCTGCCGGTCATGAACACCCCTGGTAGCAATCAATATCTTACTGCAGACAACTTCCAGTCACCGTGTGCGCTGCCTGAATTTGATGTGACCCCACCTATTGACATACCCGGTGAAGTAAAGAACATGATGGAATTGGCAGAAATCGACACCATGATTCCCTTTGACTTAAGTGCCACAAAAAAGAACACCATGGAAATGTATAGGGTTCGGTTAAGTGACAAACCACATACAGACGATCCCATACTCTGCCTGTCACTCTCTCCAGCCTCAGATCCTAGGTTGTCACATACTATGCTTGGAGAAATCCTAAATTACTACACACACTGGGCAGGATCCCTGAAGTTCACGTTTCTGTTCTGTGGATCCATGATGGCAACTGGCAAACTGTTGGTGTCATACGCGCCTCCTGGAGCCGACCCACCAAAGAAGCGTAAGGAGGCGATGTTGGGAACACATGTGATCTGGGACATAGGACTGCAGTCCTCATGTACTATGGTAGTGCCATGGATTAGCAACACCACGTATCGGCAAACCATAGATGATAGTTTCACCGAAGGCGGATACATCAGCGTCTTCTACCAAACTAGAATAGTCGTCCCTCTTTCGACACCCAGAGAGATGGACATCCTTGGTTTTGTGTCAGCGTGTAATGACTTCAGCGTGCGCTTGTTGCGAGATACCACACATATAGAGCAAAAAGCGCTAGCACAGGGGTTAGGTCAGATGCTTGAAAGCATGATTGACAACACAGTCCGTGAAACGGTGGGGGCGGCAACATCTAGAGACGCTCTCCCAAACACTGAAGCCAGTGGACCAACACACTCCAAGGAAATTCCGGCACTCACCGCAGTGGAAACTGGGGCCACAAATCCACTAGTCCCTTCTGATACAGTGCAAACCAGACATGTTGTACAACATAGGTCAAGGTCAGAGTCTAGCATAGAGTCTTTCTTCGCGCGGGGTGCATGCGTGACCATTATGACCGTGGATAACCCAGCTTCCACCACGAATAAGGATAAGCTTTTTGCAGTGTGGAAGATCACTTATAAAGATACTGTCCAGTTACGGAGGAAATTGGAGTTCTTCACCTATTCTAGATTTGATATGGAACTTACCTTTGTGGTTACTGCAAATTTCACTGAGACTAACAATGGCCATGCATTAAATCAAGTGTACCAAATTATGTACGTACCACCAGGCGCTCCAGTGCCCGAAAAATGGGACGACTACACATGGCAAACCTCATCAAATCCATCAATCTTTTACACCTACGGGACAGCTCCAGCCCGGATCTCGGTACCGTATGTTGGTATTTCGAACGCCTATTCACACTTTTACGACGGTTTTTCCAAAGTACCACTGAAGGACCAGTCGGCAGCACTAGGTGACTCCCTTTATGGTGCAGCATCTCTAAATGACTTCGGTATTTTGGCTGTTAGAGTAGTCAATGATCACAACCCGACCAAGGTCACCTCCAAAATCAGAGTGTATCTAAAACCCAAACACATCAGAGTCTGGTGCCCGCGTCCACCGAGGGCAGTGGCGTACTACGGCCCTGGAGTGGATTACAAGGATGGTACGCTTACACCCCTCTCCACCAAGGATCTGACCACATAT\n",
      "recoded sequence\n",
      "GGCGCTCACGAAAACTCGAATCGCGCATACGGATCGTCGACGATCAATTATACGACGATCAATTACTATCGCGATTCCGCGTCGAACGCCGCGTCGAAACAGGATTTTTCGCAAGATCCGTCGAAATTTACCGAACCGATTAAAGACGTACTTATCAAAACCGCGCCGATGCTCAATTCGCCGAATATCGAAGCTTGCGGTTATTCCGATCGCGTATTGCAACTTACGCTCGGTAACTCGACGATTACGACGCAAGAAGCGGCGAATTCGGTCGTCGCGTACGGTCGATGGCCCGAATACTTACGCGATTCCGAAGCGAATCCGGTCGATCAACCGACCGAACCGGACGTCGCCGCTTGTCGATTCTATACGCTCGATACCGTATCGTGGACGAAAGAATCTCGCGGTTGGTGGTGGAAATTACCGGACGCGTTACGCGATATGGGATTATTCGGTCAAAATATGTACTATCACTATCTCGGTCGATCCGGTTATACCGTACACGTACAATGCAACGCGTCGAAATTTCATCAAGGCGCATTGGGCGTATTCGCGGTACCCGAAATGTGTCTCGCCGGCGATTCGAATACGACGACGATGCATACGTCGTATCAAAACGCGAATCCCGGCGAAAAAGGCGGTACGTTTACCGGTACGTTTACGCCGGATAACAATCAAACGTCGCCCGCTCGACGATTTTGTCCGGTCGATTATCTACTCGGAAACGGTACGTTACTCGGAAACGCATTCGTATTTCCGCATCAAATTATCAATCTACGTACGAACAATTGCGCAACGCTCGTATTACCGTACGTTAATTCGCTATCGATCGATTCGATGGTTAAACATAACAATTGGGGTATCGCGATATTACCGCTCGCACCGCTTAATTTCGCAAGCGAATCGTCGCCCGAAATACCGATTACGCTTACGATCGCGCCGATGTGTTGCGAATTCAACGGATTACGCAATATTACGTTACCGCGATTACAGGGATTACCGGTTATGAATACGCCCGGATCGAATCAATATCTTACCGCCGACAATTTTCAATCGCCTTGCGCATTACCGGAATTCGACGTTACGCCGCCGATCGATATACCGGGCGAAGTAAAAAACATGATGGAACTCGCCGAAATCGATACGATGATACCGTTCGATCTATCCGCGACGAAAAAAAATACGATGGAAATGTATCGCGTACGATTATCCGACAAACCGCATACCGACGATCCGATACTATGTCTATCGTTATCGCCCGCAAGCGATCCGCGATTATCGCATACGATGCTCGGCGAAATACTCAATTATTATACGCATTGGGCCGGATCGCTTAAATTTACGTTTCTTTTTTGCGGATCGATGATGGCGACCGGTAAATTACTCGTATCGTACGCCCCCCCCGGCGCCGATCCGCCGAAAAAGCGAAAAGAGGCGATGCTCGGTACGCACGTTATATGGGATATCGGATTGCAATCGTCGTGTACGATGGTCGTACCGTGGATCTCGAATACGACGTATCGACAAACGATCGACGATTCGTTTACCGAAGGCGGATATATATCGGTTTTTTATCAAACGCGAATCGTCGTACCGCTTAGTACGCCTCGCGAAATGGATATACTCGGATTCGTATCCGCTTGCAACGATTTTTCGGTACGATTATTACGCGATACGACGCATATCGAACAAAAAGCGCTCGCGCAAGGTCTCGGTCAAATGCTCGAATCGATGATCGACAATACCGTACGCGAAACGGTCGGTGCCGCGACGTCGCGCGACGCGTTACCGAATACCGAAGCGAGCGGTCCTACGCACTCGAAAGAGATACCGGCGCTTACCGCGGTCGAAACCGGCGCGACGAATCCGCTCGTACCGAGCGATACCGTACAAACGCGACACGTCGTACAACACCGGTCGCGATCCGAATCGTCGATCGAATCGTTTTTCGCTCGCGGCGCTTGCGTTACGATCATGACGGTCGACAATCCGGCGTCGACGACGAATAAGGATAAACTATTCGCGGTATGGAAGATTACGTATAAGGATACCGTACAATTACGACGAAAGCTCGAATTTTTTACGTATTCGCGATTCGATATGGAACTTACGTTCGTCGTTACCGCGAATTTTACCGAAACGAACAACGGTCACGCACTCAATCAGGTATATCAAATTATGTACGTACCGCCCGGCGCACCGGTACCCGAAAAATGGGACGATTATACGTGGCAAACGAGCTCGAATCCGTCGATTTTTTATACGTACGGTACCGCACCCGCTCGTATATCCGTACCGTACGTCGGTATCTCGAACGCGTATTCGCATTTTTACGACGGATTCTCGAAGGTACCGCTTAAGGATCAATCCGCCGCACTCGGCGATTCGCTATACGGTGCCGCGAGTTTAAACGATTTCGGTATACTCGCGGTACGCGTCGTTAACGATCACAATCCGACGAAGGTTACGTCGAAAATTCGCGTATATCTTAAACCGAAACATATACGCGTATGGTGTCCGCGACCGCCTCGCGCGGTCGCGTATTACGGACCGGGGGTCGATTATAAAGACGGTACGCTTACGCCGTTATCGACGAAAGATCTTACGACGTAC\n",
      "total number of codon diferences between original and recoded sequences 615\n",
      "AA sequence remains the same, good\n",
      "num desired codon pairs in original sequence: 92\n",
      "num desired codon pairs in recoded sequence: 888\n",
      "frame1 codon pair score of the original sequence: 13.719914612\n",
      "frame1 codon pair score of the recoded sequence: -7.185032119000012\n",
      "frame2&3 indep codon pair score of the original sequence: -31.61352552200005\n",
      "frame2&3 indep codon pair score of the recoded sequence: -791.3998220849992\n",
      "CAI of the original sequence is: 0.735171695523\n",
      "CAI of the recoded sequence is: 0.563109952212\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "#import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "\n",
    "\n",
    "# number of desired codon paris in sequence of any length - in this case, only in frame 2 and 3\n",
    "def num_costly_codon_pairs(seq):\n",
    "    num_indep = 0\n",
    "    # now do the independent scoring, frame 2\n",
    "    for i in range(1,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] in pairs_to_maximize_indep):\n",
    "            num_indep += 1\n",
    "    # now do the independent scoring, frame 3\n",
    "    for i in range(2,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] in pairs_to_maximize_indep):\n",
    "            num_indep += 1\n",
    "    return(num_indep)\n",
    "\n",
    "# codon pair score of any length sequence\n",
    "def get_cps_indep(seq):\n",
    "    cps_indep = 0 \n",
    "    # for indep_dependent pairs scores - frame 2\n",
    "    for i in range(1,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_indep):\n",
    "            #print(\"Missing \" + seq[i:i+6] + \" from independent scores\")\n",
    "            continue\n",
    "        cps_indep += cp_scores_indep[seq[i:i+6]]\n",
    "    # for indep_dependent pairs scores - frame 3\n",
    "    for i in range(2,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_indep):\n",
    "            #print(\"Missing \" + seq[i:i+6] + \" from independent scores\")\n",
    "            continue\n",
    "        cps_indep += cp_scores_indep[seq[i:i+6]]\n",
    "    return(cps_indep)\n",
    "\n",
    "def get_cps_frame1(seq):\n",
    "    # for dependent pairs scores\n",
    "    cps_dependent = 0 \n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_dependent):\n",
    "            print(\"Missing \" + seq[i:i+6] + \" from dependent scores\")\n",
    "            continue\n",
    "        cps_dependent += cp_scores_dependent[seq[i:i+6]]\n",
    "    return(cps_dependent)\n",
    "\n",
    "# codon adaptation index of any length sequence\n",
    "def get_cai(seq):\n",
    "    # get the CAI as the geometric mean of all the codon's individual CAI values\n",
    "    cai_list = []\n",
    "    for i in range(0,len(seq),3):\n",
    "        cai_list.append(codon_CAI[seq[i:(i+3)]])\n",
    "    return(scistats.gmean(cai_list))\n",
    "\n",
    "# read in synonymous codons\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "\n",
    "# read in the human orfs and make the CAI for all the codons\n",
    "codon_occs = dict()\n",
    "# set up the dict\n",
    "for codon in (codon_to_aa_dict.keys()):\n",
    "    codon_occs[codon] = 0\n",
    "genes_file = open(\"human_CCDS_nucleotide.current.fna\")\n",
    "for line in genes_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\": \n",
    "        if (len(line)%3 != 0):\n",
    "            #print(\"PROBLEM, genes file is not multiple of 3 per line.\")\n",
    "            continue\n",
    "        # get the frequency of each codon\n",
    "        for i in range(0,len(line),3):\n",
    "            codon_occs[line[i:(i+3)]] += 1\n",
    "\n",
    "# get the CAI of each codon\n",
    "codon_CAI = dict()\n",
    "for aa in (aa_to_syn_codons_dict.keys()):\n",
    "    highest_frequency = 0\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (codon_occs[codon] > highest_frequency):\n",
    "            highest_frequency = codon_occs[codon]\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (highest_frequency == 0):\n",
    "            codon_CAI[codon] = 0\n",
    "        else:\n",
    "            codon_CAI[codon] = codon_occs[codon]/highest_frequency\n",
    "            \n",
    "# read in the codon pair scores, both dependent and independent\n",
    "cp_scores_dependent = dict()\n",
    "cp_scores_file = open(\"human_FD_Scores.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "counter = 0\n",
    "pairs_to_maximize_dependent = []\n",
    "for line in cp_reader:\n",
    "    cp_scores_dependent[line[0]] = float(line[1])\n",
    "    counter += 1\n",
    "    if (counter <= 372):\n",
    "        if (line[0][2] == \"T\" and line[0][3] == \"A\" and ((line[0][4] == \"A\" or line[0][4] == \"G\"))):\n",
    "            continue\n",
    "        pairs_to_maximize_dependent.append(line[0])\n",
    "print(len(pairs_to_maximize_dependent))\n",
    "\n",
    "cp_scores_indep = dict()\n",
    "cp_scores_file = open(\"human_FIndep_Scores.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "counter = 0\n",
    "pairs_to_maximize_indep = []\n",
    "for line in cp_reader:\n",
    "    cp_scores_indep[line[0]] = float(line[1])\n",
    "    counter += 1\n",
    "    if (counter <= 372):\n",
    "        if (line[0][2] == \"T\" and line[0][3] == \"A\" and ((line[0][4] == \"A\" or line[0][4] == \"G\"))):\n",
    "            continue\n",
    "        pairs_to_maximize_indep.append(line[0])\n",
    "print(len(pairs_to_maximize_indep))\n",
    "\n",
    "# read in the gene's codon sequence\n",
    "codons_file = open(\"polio_genomics.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "print(\"original codons\")\n",
    "print(original_codons)\n",
    "#original_codons = \"GGCGCACATGAA\"\n",
    "\n",
    "original_AA = \"\"\n",
    "for i in range(0,len(original_codons),3):\n",
    "    original_AA += codon_to_aa_dict[original_codons[i:i+3]]\n",
    "\n",
    "#print(original_AA)\n",
    "\n",
    "# store_best will contain, at each AA position, and at each of the syn codon choices, and the following AA's\n",
    "# syn codon choices, \n",
    "# the best cost of the prefix that starts with this codon, the current codon pair score,\n",
    "# and the next codon that corresponds to this cost\n",
    "store_best = [None] * (len(original_AA)-1)\n",
    "for i in range(0,len(original_AA)-1):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = dict()\n",
    "        for codon_next in aa_to_syn_codons_dict[original_AA[i+1]]:\n",
    "            store_best[i][codon][codon_next] = 0\n",
    "            \n",
    "# initialize the dynamic program by setting store_best for the first amino acid position\n",
    "for codon in store_best[0]:\n",
    "    for codon_next in store_best[1]:\n",
    "        store_best[0][codon][codon_next] = (0,0,\"\",codon)\n",
    "\n",
    "# do the dynamic program, starting from the beginning\n",
    "for pos in range(1, len(store_best)):\n",
    "    for codon in store_best[pos]:\n",
    "        for codon_next in store_best[pos][codon]:\n",
    "            # consider as cost the cost of adding this codon to one of the best sequences\n",
    "            # from the previous suffix\n",
    "            best_prev = \"\"\n",
    "            best_cps = sys.maxsize # keep this\n",
    "            best_cost = -1 #sys.maxsize #-1\n",
    "            best_seq = \"\"\n",
    "            for prev_codon in store_best[pos-1]:\n",
    "                cost = store_best[pos-1][prev_codon][codon][0] + num_costly_codon_pairs(prev_codon + codon + codon_next) \n",
    "                cp_score = get_cps_indep(store_best[pos-1][prev_codon][codon][3] + codon + codon_next)\n",
    "                #print(str(prev_codon))\n",
    "                if ((cost > best_cost) or # > <\n",
    "                    # break the tie with lowest codon pair score\n",
    "                    ((cost == best_cost) and (cp_score < best_cps))):\n",
    "                    best_cost = cost\n",
    "                    best_cps = cp_score\n",
    "                    best_prev = prev_codon\n",
    "                    best_seq = store_best[pos-1][prev_codon][codon][3] + codon\n",
    "            store_best[pos][codon][codon_next] = (best_cost, best_cps, best_prev, best_seq)\n",
    "            #print(\"storing for \" + str(pos) + \" \" + codon +  \" \" + codon_next + \" : \" + str(best_cps) + \" \" + best_prev + \" \" + best_seq)\n",
    "            #GGC GCA CAT GAA\n",
    "\n",
    "# do the final position\n",
    "best_cps = sys.maxsize # keep this\n",
    "best_cost = -1 #sys.maxsize #-1\n",
    "best_seq = \"\"\n",
    "for codon in aa_to_syn_codons_dict[original_AA[-1]]:\n",
    "    for prev_codon in aa_to_syn_codons_dict[original_AA[-2]]:\n",
    "        cost = store_best[-1][prev_codon][codon][0]\n",
    "        cp_score = store_best[-1][prev_codon][codon][1]\n",
    "        if ((cost > best_cost) or # > <\n",
    "            # break the tie with lowest codon pair score\n",
    "            ((cost == best_cost) and (cp_score < best_cps))):\n",
    "            best_cost = cost\n",
    "            best_cps = cp_score\n",
    "            best_seq = store_best[-1][prev_codon][codon][3] + codon\n",
    "#print(str(best_cps) + \" \" + str(best_cost) + \" \" + best_seq)\n",
    "\n",
    "recoded_sequence = best_seq\n",
    "print(\"recoded sequence\")\n",
    "print(recoded_sequence)\n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "#print(len(original_codons))\n",
    "print(\"num desired codon pairs in original sequence: \" + str(num_costly_codon_pairs(original_codons)))\n",
    "print(\"num desired codon pairs in recoded sequence: \" + str(num_costly_codon_pairs(recoded_sequence)))\n",
    "print(\"frame1 codon pair score of the original sequence: \" + str(get_cps_frame1(original_codons)))\n",
    "print(\"frame1 codon pair score of the recoded sequence: \" + str(get_cps_frame1(recoded_sequence)))\n",
    "print(\"frame2&3 indep codon pair score of the original sequence: \" + str(get_cps_indep(original_codons)))\n",
    "print(\"frame2&3 indep codon pair score of the recoded sequence: \" + str(get_cps_indep(recoded_sequence)))\n",
    "print(\"CAI of the original sequence is: \" + str(get_cai(original_codons)))\n",
    "print(\"CAI of the recoded sequence is: \" + str(get_cai(recoded_sequence)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lys2 Version June2017 for Justin paper - design2: maximize 10% of worse frame dependent scores for yeast that do not have nnTAA/Gnn, and 10% of frame independent hexamers for all three frames that do not have nnTAA/Gnn;  the score function is frame dependent + sum(frame_independent_over three frames)/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "331\n",
      "372\n",
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGCTGCCACAACTTGATGTGCCACACGACTCCTTCTCAAACAAATACGCTGTCGCTCTGTCAGTGTGGGCCGCTCTGATCTATCGCGTGACCGGCGACGACGATATCGTGTTATATATCGCGAACAACAAAATCTTACGTTTTAACATACAACCGACCTGGAGCTTTAACGAATTATACTCGACGATTAACAACGAATTAAACAAGTTGAACTCGATCGAAGCAAACTTCTCTTTTGACGAACTCGCTGAGAAAATACAGTCCTGTCAAGACCTTGAGCGGACTCCCCAATTGTTTCGTCTCGCTTTTCTCGAAAACCAAGACTTTAAGTTGGACGAATTTAAACACCACCTGGTGGATTTCGCTCTGAACTTAGACACGTCCAACAACGCCCACGTTCTAAATCTAATCTATAACAGCCTCCTCTACTCCAACGAGCGGGTTACGATTGTCGCCGATCAATTTACACAATATCTAACCGCCGCTCTGTCAGACCCGAGCAACTGTATAACCAAAATCTCGCTTATAACCGCTTCCTCAAAAGATAGCCTCCCGGACCCGACCAAAAATCTCGGGTGGTGTGATTTCGTGGGTTGTATACACGATATCTTTCAAGACAACGCCGAAGCATTCCCGGAGCGGACTTGTGTCGTCGAGACCCCGACCTTGAACTCCGATAAGTCCCGATCTTTTACATATCGCGATATAAATCGGACTTCAAACATAGTCGCCCACTATCTAATAAAAACCGGGATTAAACGCGGCGACGTCGTGATGATATACTCCTCAAGGGGGGTCGATCTTATGGTGTGCGTGATGGGGGTCTTAAAAGCCGGGGCGACCTTCTCGGTCATAGATCCCGCGTACCCCCCCGCGCGTCAGACCATCTATCTCGGGGTCGCGAAGCCCCGCGGCTTGATCGTGATCCGCGCCGCGGGTCAACTTGATCAACTTGTCGAGGATTATATAAACGACGAACTCGAAATCGTGAGCCGAATAAACTCGATCGCTATACAGGAAAACGGTACGATCGAAGGCGGCAAGTTGGATAACGGCGAGGACGTCCTCGCACCGTATGATCACTATAAAGACACGCGGACCGGGGTCGTCGTCGGCCCGGACTCCAACCCGACCTTGTCTTTTACATCGGGTTCCGAAGGCATCCCTAAGGGGGTCTTAGGTCGGCACTTTAGCCTCGCTTATTACTTTAACTGGATGTCAAAACGCTTTAACCTTACAGAGAACGACAAATTTACTATGTTGTCGGGTATCGCCCACGATCCAATCCAGAGGGATATGTTCACCCCGTTGTTTCTGGGGGCCCAACTTTATGTCCCGACCCAAGACGATATCGGGACCCCGGGTCGGCTCGCTGAGTGGATGTCAAAATACGGCTGTACCGTGACCCACCTGACCCCCGCGATGGGTCAGTTGTTGACCGCCCAAGCGACGACTCCTTTTCCAAAACTCCACCACGCTTTTTTCGTGGGTGATATACTTACAAAACGCGATTGTCTCCGGCTCCAGACCCTCGCTGAGAACTGTCGGATCGTGAACATGTACGGTACGACTGAGACGCAGAGGGCAGTGTCTTATTTCGAGGTGAAGTCCAAAAACGACGACCCGAACTTTCTGAAAAAGTTGAAAGACGTCATGCCAGCCGGCAAGGGGATGTTGAACGTGCAGTTGTTGGTCGTCAATCGGAACGATCGGACTCAAATATGCGGCATCGGGGAGATCGGGGAGATCTATGTACGCGCCGGGGGTTTGGCCGAAGGCTATAGGGGTCTGCCAGAATTAAACAAAGAGAAATTCGTAAACAACTGGTTCGTAGAGAAAGATCACTGGAACTACTTAGATAAAGATAACGGCGAGCCGTGGCGTCAATTTTGGCTCGGCCCGAGGGACCGATTATATCGGACCGGGGATCTAGGTCGGTATCTCCCGAACGGCGATTGCGAATGTTGCGGCCGCGCCGATGATCAAGTGAAAATCCGTGGCTTTCGTATCGAATTAGGGGAGATAGATACCCACATCTCACAACACCCGTTGGTCCGGGAGAACATAACCTTGGTCCGTAAAAACGCCGATAACGAACCAACCTTGATAACCTTTATGGTCCCGAGGTTCGATAAGCCCGACGATTTGTCAAAATTTCAGTCCGATGTCCCGAAAGAGGTCGAGACGGATCCAATCGTGAAAGGCTTGATCGGGTATCACTTGTTGTCAAAAGATATCCGTACCTTTCTGAAAAAACGCCTCGCGAGCTACGCTATGCCGAGCCTGATAGTCGTGATGGATAAGTTGCCGTTGAACCCGAACGGCAAGGTGGATAAGCCTAAACTCCAATTCCCGACCCCGAAACAACTTAATCTAGTCGCGGAGAACACCGTTTCCGAAACCGACGATAGTCAGTTTACAAACGTGGAGCGCGAAGTGCGCGATTTGTGGTTGTCAATTCTCCCGACCAAGCCCGCGTCGGTCAGCCCGGACGATAGCTTTTTCGATCTCGGGGGTCATTCAATTCTCGCAACCAAAATGATCTTTACCTTGAAAAAAAAACTCCAAGTGGATCTCCCGTTGGGTACGATCTTTAAATACCCGACGATTAAAGCATTCGCAGCGGAGATAGATCGGATCAAGTCCTCAGGGGGTTCCTCACAAGGCGAGGTCGTCGAGAACGTGACCGCTAACTACGCCGAGGACGCCAAAAAGTTGGTCGAGACCCTCCCGTCGTCTTACCCGTCGCGTGAGTATTTCGTAGAACCAAACTCGGCCGAAGGCAAGACGACGATTAACGTGTTCGTGACCGGGGTGACCGGCTTTCTGGGGTCTTATATATTAGCCGATTTGTTGGGTCGGAGCCCGAAAAACTATAGCTTTAAAGTGTTCGCACACGTGCGCGCGAAAGACGAAGAAGCAGCGTTCGCCCGACTCCAAAAAGCCGGGATTACCTACGGTACGTGGAACGAGAAATTCGCGAGCAACATAAAAGTCGTGTTAGGGGATTTGTCAAAAAGTCAGTTCGGCTTGTCCGATGAGAAGTGGATGGATCTCGCTAACACCGTTGATATAATAATACACAACGGAGCCCTGGTACACTGGGTGTACCCCTACGCCAAGTTGCGCGACCCGAACGTGATCTCGACGATTAACGTGATGAGCCTCGCAGCAGTGGGCAAGCCTAAATTTTTCGATTTCGTATCCTCAACCTCAACCTTGGACACGGAGTATTACTTTAACTTGTCCGATAAGTTGGTCTCCGAAGGCAAGCCCGGGATTCTTGAGTCCGATGATCTTATGAACTCGGCTTCGGGTTTGACCGGGGGTTACGGTCAGTCCAAGTGGGCCGCGGAGTATATAATAAGACGCGCGGGTGAGCGAGGTCTCCGGGGGTGTATCGTTCGACCGGGGTATGTGACCGGCGCGAGCGCAAACGGCTCCTCAAACACCGACGATTTTTTGCTCCGGTTTCTGAAAGGCTCGGTCCAACTTGGCAAGATCCCCGATATCGAAAACTCGGTCAACATGGTGCCAGTTGATCACGTCGCCCGAGTCGTCGTCGCGACCAGCCTGAACCCCCCGAAAGAGAACGAGCTCGCTGTCGCCCAAGTGACCGGTCATCCCCGGATCTTGTTTAAAGACTACTTATATACCCTCCACGATTACGGCTATGACGTCGAGATCGAATCGTACTCCAAGTGGAAAAAAAGCCTCGAAGCATCGGTCATAGATCGGAACGAAGAAAACGCCCTGTACCCGTTGCTCCACATGGTGTTAGATAACCTCCCGGAGTCGACTAAAGCACCGGAGTTAGACGATCGGAACGCCGTCGCGAGCCTGAAAAAAGACACGGCTTGGACCGGGGTCGATTGGTCAAACGGCATCGGGGTGACCCCCGAAGAAGTCGGGATTTATATCGCTTTTCTGAACAAAGTGGGTTTTCTGCCCCCCCCGACCCACAACGACAAGTTGCCACTGCCGAGCATCGAACTTACACAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "\n",
      "\n",
      "\n",
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGCTGCCACAACTTGATGTGCCACACGACTCCTTCTCAAACAAATACGCTGTCGCTCTGTCAGTGTGGGCCGCTCTGATCTATCGCGTGACCGGCGACGACGATATCGTGTTATATATCGCGAACAACAAAATCTTACGTTTTAACATACAACCGACCTGGAGCTTTAACGAATTATACTCGACGATTAACAACGAATTAAACAAGTTGAACTCGATCGAAGCAAACTTCTCTTTTGACGAACTCGCTGAGAAAATACAGTCCTGTCAAGACCTTGAGCGGACTCCCCAATTGTTTCGTCTCGCTTTTCTCGAAAACCAAGACTTTAAGTTGGACGAATTTAAACACCACCTGGTGGATTTCGCTCTGAACTTAGACACGTCCAACAACGCCCACGTTCTAAATCTAATCTATAACAGCCTCCTCTACTCCAACGAGCGGGTTACGATTGTCGCCGATCAATTTACACAATATCTAACCGCCGCTCTGTCAGACCCGAGCAACTGTATAACCAAAATCTCGCTTATAACCGCTTCCTCAAAAGATAGCCTCCCGGACCCGACCAAAAATCTCGGGTGGTGTGATTTCGTGGGTTGTATACACGATATCTTTCAAGACAACGCCGAAGCATTCCCGGAGCGGACTTGTGTCGTCGAGACCCCGACCTTGAACTCCGATAAGTCCCGATCTTTTACATATCGCGATATAAATCGGACTTCAAACATAGTCGCCCACTATCTAATAAAAACCGGGATTAAACGCGGCGACGTCGTGATGATATACTCCTCAAGGGGGGTCGATCTTATGGTGTGCGTGATGGGGGTCTTAAAAGCCGGGGCGACCTTCTCGGTCATAGATCCCGCGTACCCCCCCGCGCGTCAGACCATCTATCTCGGGGTCGCGAAGCCCCGCGGCTTGATCGTGATCCGCGCCGCGGGTCAACTTGATCAACTTGTCGAGGATTATATAAACGACGAACTCGAAATCGTGAGCCGAATAAACTCGATCGCTATACAGGAAAACGGTACGATCGAAGGCGGCAAGTTGGATAACGGCGAGGACGTCCTCGCACCGTATGATCACTATAAAGACACGCGGACCGGGGTCGTCGTCGGCCCGGACTCCAACCCGACCTTGTCTTTTACATCGGGTTCCGAAGGCATCCCTAAGGGGGTCTTAGGTCGGCACTTTAGCCTCGCTTATTACTTTAACTGGATGTCAAAACGCTTTAACCTTACAGAGAACGACAAATTTACTATGTTGTCGGGTATCGCCCACGATCCAATCCAGAGGGATATGTTCACCCCGTTGTTTCTGGGGGCCCAACTTTATGTCCCGACCCAAGACGATATCGGGACCCCGGGTCGGCTCGCTGAGTGGATGTCAAAATACGGCTGTACCGTGACCCACCTGACCCCCGCGATGGGTCAGTTGTTGACCGCCCAAGCGACGACTCCTTTTCCAAAACTCCACCACGCTTTTTTCGTGGGTGATATACTTACAAAACGCGATTGTCTCCGGCTCCAGACCCTCGCTGAGAACTGTCGGATCGTGAACATGTACGGTACGACTGAGACGCAGAGGGCAGTGTCTTATTTCGAGGTGAAGTCCAAAAACGACGACCCGAACTTTCTGAAAAAGTTGAAAGACGTCATGCCAGCCGGCAAGGGGATGTTGAACGTGCAGTTGTTGGTCGTCAATCGGAACGATCGGACTCAAATATGCGGCATCGGGGAGATCGGGGAGATCTATGTACGCGCCGGGGGTTTGGCCGAAGGCTATAGGGGTCTGCCAGAATTAAACAAAGAGAAATTCGTAAACAACTGGTTCGTAGAGAAAGATCACTGGAACTACTTAGATAAAGATAACGGCGAGCCGTGGCGTCAATTTTGGCTCGGCCCGAGGGACCGATTATATCGGACCGGGGATCTAGGTCGGTATCTCCCGAACGGCGATTGCGAATGTTGCGGCCGCGCCGATGATCAAGTGAAAATCCGTGGCTTTCGTATCGAATTAGGGGAGATAGATACCCACATCTCACAACACCCGTTGGTCCGGGAGAACATAACCTTGGTCCGTAAAAACGCCGATAACGAACCAACCTTGATAACCTTTATGGTCCCGAGGTTCGATAAGCCCGACGATTTGTCAAAATTTCAGTCCGATGTCCCGAAAGAGGTCGAGACGGATCCAATCGTGAAAGGCTTGATCGGGTATCACTTGTTGTCAAAAGATATCCGTACCTTTCTGAAAAAACGCCTCGCGAGCTACGCTATGCCGAGCCTGATAGTCGTGATGGATAAGTTGCCGTTGAACCCGAACGGCAAGGTGGATAAGCCTAAACTCCAATTCCCGACCCCGAAACAACTTAATCTAGTCGCGGAGAACACCGTTTCCGAAACCGACGATAGTCAGTTTACAAACGTGGAGCGCGAAGTGCGCGATTTGTGGTTGTCAATTCTCCCGACCAAGCCCGCGTCGGTCAGCCCGGACGATAGCTTTTTCGATCTCGGGGGTCATTCAATTCTCGCAACCAAAATGATCTTTACCTTGAAAAAAAAACTCCAAGTGGATCTCCCGTTGGGTACGATCTTTAAATACCCGACGATTAAAGCATTCGCAGCGGAGATAGATCGGATCAAGTCCTCAGGGGGTTCCTCACAAGGCGAGGTCGTCGAGAACGTGACCGCTAACTACGCCGAGGACGCCAAAAAGTTGGTCGAGACCCTCCCGTCGTCTTACCCGTCGCGTGAGTATTTCGTAGAACCAAACTCGGCCGAAGGCAAGACGACGATTAACGTGTTCGTGACCGGGGTGACCGGCTTTCTGGGGTCTTATATATTAGCCGATTTGTTGGGTCGGAGCCCGAAAAACTATAGCTTTAAAGTGTTCGCACACGTGCGCGCGAAAGACGAAGAAGCAGCGTTCGCCCGACTCCAAAAAGCCGGGATTACCTACGGTACGTGGAACGAGAAATTCGCGAGCAACATAAAAGTCGTGTTAGGGGATTTGTCAAAAAGTCAGTTCGGCTTGTCCGATGAGAAGTGGATGGATCTCGCTAACACCGTTGATATAATAATACACAACGGAGCCCTGGTACACTGGGTGTACCCCTACGCCAAGTTGCGCGACCCGAACGTGATCTCGACGATTAACGTGATGAGCCTCGCAGCAGTGGGCAAGCCTAAATTTTTCGATTTCGTATCCTCAACCTCAACCTTGGACACGGAGTATTACTTTAACTTGTCCGATAAGTTGGTCTCCGAAGGCAAGCCCGGGATTCTTGAGTCCGATGATCTTATGAACTCGGCTTCGGGTTTGACCGGGGGTTACGGTCAGTCCAAGTGGGCCGCGGAGTATATAATAAGACGCGCGGGTGAGCGAGGTCTCCGGGGGTGTATCGTTCGACCGGGGTATGTGACCGGCGCGAGCGCAAACGGCTCCTCAAACACCGACGATTTTTTGCTCCGGTTTCTGAAAGGCTCGGTCCAACTTGGCAAGATCCCCGATATCGAAAACTCGGTCAACATGGTGCCAGTTGATCACGTCGCCCGAGTCGTCGTCGCGACCAGCCTGAACCCCCCGAAAGAGAACGAGCTCGCTGTCGCCCAAGTGACCGGTCATCCCCGGATCTTGTTTAAAGACTACTTATATACCCTCCACGATTACGGCTATGACGTCGAGATCGAATCGTACTCCAAGTGGAAAAAAAGCCTCGAAGCATCGGTCATAGATCGGAACGAAGAAAACGCCCTGTACCCGTTGCTCCACATGGTGTTAGATAACCTCCCGGAGTCGACTAAAGCACCGGAGTTAGACGATCGGAACGCCGTCGCGAGCCTGAAAAAAGACACGGCTTGGACCGGGGTCGATTGGTCAAACGGCATCGGGGTGACCCCCGAAGAAGTCGGGATTTATATCGCTTTTCTGAACAAAGTGGGTTTTCTGCCCCCCCCGACCCACAACGACAAGTTGCCACTGCCGAGCATCGAACTTACACAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "total number of codon diferences between original and recoded sequences 915\n",
      "AA sequence remains the same, good\n",
      "4179\n",
      "num desired codon pairs in original sequence: 137.33333333333334\n",
      "num desired codon pairs in recoded sequence: 761.6666666666667\n",
      "Missing GCTTAA from dependent scores\n",
      "codon pair score of the original sequence: 55.645373368000044\n",
      "Missing GCTTAA from dependent scores\n",
      "codon pair score of the recoded sequence: -242.0292744626668\n",
      "CAI of the original sequence is: 0.783213534519\n",
      "CAI of the recoded sequence is: 0.599447584394\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "#import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "\n",
    "\n",
    "# number of desired codon paris in sequence of any length\n",
    "def num_costly_codon_pairs(seq, count_negatives):\n",
    "    num_dependent = 0\n",
    "    # the dependent pairs score\n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] in pairs_to_maximize_dependent):\n",
    "            num_dependent += 1\n",
    "    # now do the independent scoring\n",
    "    num_indep = 0\n",
    "    for i in range(0,(len(seq)-5),1):\n",
    "        if (seq[i:i+6] in pairs_to_maximize_indep):\n",
    "            num_indep += 1\n",
    "    num = num_dependent + num_indep/3\n",
    "    return(num)\n",
    "\n",
    "# codon pair score of any length sequence\n",
    "def get_cps(seq):\n",
    "    # for dependent pairs scores\n",
    "    cps_dependent = 0 \n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_dependent):\n",
    "            print(\"Missing \" + seq[i:i+6] + \" from dependent scores\")\n",
    "            continue\n",
    "        cps_dependent += cp_scores_dependent[seq[i:i+6]]\n",
    "    # for indep_dependent pairs scores\n",
    "    cps_indep = 0 \n",
    "    for i in range(0,(len(seq)-5),1):\n",
    "        if (seq[i:i+6] not in cp_scores_indep):\n",
    "            #print(\"Missing \" + seq[i:i+6] + \" from independent scores\")\n",
    "            continue\n",
    "        cps_indep += cp_scores_indep[seq[i:i+6]]\n",
    "    cps = cps_dependent + cps_indep/3\n",
    "    return(cps)\n",
    "\n",
    "# codon adaptation index of any length sequence\n",
    "def get_cai(seq):\n",
    "    # get the CAI as the geometric mean of all the codon's individual CAI values\n",
    "    cai_list = []\n",
    "    for i in range(0,len(seq),3):\n",
    "        cai_list.append(codon_CAI[seq[i:(i+3)]])\n",
    "    return(scistats.gmean(cai_list))\n",
    "\n",
    "# read in synonymous codons\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "\n",
    "# read in the yeast orfs and make the CAI for all the codons\n",
    "codon_occs = dict()\n",
    "# set up the dict\n",
    "for codon in (codon_to_aa_dict.keys()):\n",
    "    codon_occs[codon] = 0\n",
    "genes_file = open(\"s_cerevisae_orf_coding_no_Mito_no_Plasmid.fasta\")\n",
    "for line in genes_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\": \n",
    "        if (len(line)%3 != 0):\n",
    "            print(\"PROBLEM, genes file is not multiple of 3 per line.\")\n",
    "        # get the frequency of each codon\n",
    "        for i in range(0,len(line),3):\n",
    "            codon_occs[line[i:(i+3)]] += 1\n",
    "            \n",
    "# get the CAI of each codon\n",
    "codon_CAI = dict()\n",
    "for aa in (aa_to_syn_codons_dict.keys()):\n",
    "    highest_frequency = 0\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (codon_occs[codon] > highest_frequency):\n",
    "            highest_frequency = codon_occs[codon]\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (highest_frequency == 0):\n",
    "            codon_CAI[codon] = 0\n",
    "        else:\n",
    "            codon_CAI[codon] = codon_occs[codon]/highest_frequency\n",
    "        \n",
    "# read in the codon pair scores, both dependent and independent\n",
    "cp_scores_dependent = dict()\n",
    "cp_scores_file = open(\"cerevisiae_FD_Scores.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "counter = 0\n",
    "pairs_to_maximize_dependent = []\n",
    "for line in cp_reader:\n",
    "    cp_scores_dependent[line[0]] = float(line[1])\n",
    "    counter += 1\n",
    "    if (counter <= 372):\n",
    "        if (line[0][2] == \"T\" and line[0][3] == \"A\" and ((line[0][4] == \"A\" or line[0][4] == \"G\"))):\n",
    "            continue\n",
    "        pairs_to_maximize_dependent.append(line[0])\n",
    "print(len(pairs_to_maximize_dependent))\n",
    "\n",
    "cp_scores_indep = dict()\n",
    "cp_scores_file = open(\"cerevisiae_FIndep_Scores.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "counter = 0\n",
    "pairs_to_maximize_indep = []\n",
    "for line in cp_reader:\n",
    "    cp_scores_indep[line[0]] = float(line[1])\n",
    "    counter += 1\n",
    "    if (counter <= 372):\n",
    "        if (line[0][2] == \"T\" and line[0][3] == \"A\" and ((line[0][4] == \"A\" or line[0][4] == \"G\"))):\n",
    "            continue\n",
    "        pairs_to_maximize_indep.append(line[0])\n",
    "print(len(pairs_to_maximize_indep))\n",
    "\n",
    "\n",
    "# read in the gene's codon sequence\n",
    "codons_file = open(\"S288C_YBR115C_LYS2_genomic.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "# read in the gene's amino acid sequence\n",
    "aa_file = open(\"S288C_YBR115C_LYS2_protein.fsa\")\n",
    "original_AA = \"\"\n",
    "for l in aa_file:\n",
    "    l = l.rstrip()\n",
    "    if l[0] != \">\":\n",
    "        original_AA = original_AA + l\n",
    "aa_file.close()\n",
    "\n",
    "if (len(original_AA)*3 != len(original_codons)):\n",
    "    print(\"PROBLEM: codons and AA don't match\")\n",
    "\n",
    "    \n",
    "# store the sequences, and then work with the ones that have \"chopped off\" start and end\n",
    "long_original_AA = original_AA\n",
    "original_AA = original_AA[40:-20]\n",
    "long_original_codons = original_codons\n",
    "original_codons = original_codons[120:-60]\n",
    "   \n",
    "# store_best will contain, at each AA position, and at each of the syn codon choices,\n",
    "# the best cost of the suffix that starts with this codon, the current codon pair score,\n",
    "# and the next codon that corresponds to this cost\n",
    "store_best = [None] * len(original_AA)\n",
    "for i in range(0,len(original_AA)):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = 0\n",
    "                \n",
    "\n",
    "# initialize the dynamic program by setting store_best for the last amino acid position\n",
    "for codon in store_best[-1]:\n",
    "    store_best[i][codon] = (0,0,\"\",\"\")\n",
    "\n",
    "# do the dynamic program, stating from the one before the end of the list, until the beginning\n",
    "for pos in reversed(range(0, len(store_best)-1)):\n",
    "    for codon in store_best[pos]:\n",
    "        # consider as cost the cost of adding this codon to one of the best sequences\n",
    "        # from the previous suffix\n",
    "        best_prev = \"\"\n",
    "        best_cps = sys.maxsize # keep this\n",
    "        best_cost = -1 #sys.maxsize #-1\n",
    "        best_seq = \"\"\n",
    "        for prev_codon in store_best[pos+1]:\n",
    "            cost = num_costly_codon_pairs(codon+prev_codon,1) + store_best[pos+1][prev_codon][0]\n",
    "            #cp_score = get_cps(codon+prev_codon) + store_best[pos+1][prev_codon][1]\n",
    "            cp_score = get_cps(codon + store_best[pos+1][prev_codon][3])\n",
    "            #print(cp_score)\n",
    "            # maximizer the cost\n",
    "            #if (cp_score < best_cps):\n",
    "            if ((cost > best_cost) or # > <\n",
    "                # break the tie with lowest codon pair score\n",
    "                ((cost == best_cost) and (cp_score < best_cps))):\n",
    "                best_cost = cost\n",
    "                best_cps = cp_score\n",
    "                best_prev = prev_codon\n",
    "                best_seq = prev_codon + store_best[pos+1][prev_codon][3]\n",
    "        store_best[pos][codon] = (best_cost, best_cps, best_prev, best_seq)\n",
    "        #print(\"storing for \" + str(pos) + \" \" + codon + \": \" + str(best_cps) + \" \" + best_prev)\n",
    "\n",
    "# going from the beginning to the end, read off the recoded sequences from the store_best\n",
    "# re-doing the computation for the first position, to pick the best of the syn codons\n",
    "# the rest of the seuqnces is just read out\n",
    "recoded_sequence = \"\"\n",
    "next_codon = \"\"\n",
    "for pos in range(0, len(store_best)):\n",
    "    if pos == 0:\n",
    "        # maximum\n",
    "        best_first = \"\"\n",
    "        best_first_cps = sys.maxsize # keep this\n",
    "        best_first_cost = -1 # sys.maxsize #-1\n",
    "        for codon in store_best[0]:\n",
    "            # maximize the cost\n",
    "            #if (store_best[pos][codon][1] < best_first_cps):\n",
    "            if ((store_best[pos][codon][0] > best_first_cost) or  #> <\n",
    "               ((store_best[pos][codon][0] == best_first_cost) and (store_best[pos][codon][1] < best_first_cps))):\n",
    "                best_first = codon\n",
    "                best_first_cost = store_best[pos][codon][0]\n",
    "                best_first_cps = store_best[pos][codon][1]\n",
    "                next_codon = store_best[pos][codon][2]\n",
    "        recoded_sequence += best_first\n",
    "    else:\n",
    "        recoded_sequence += next_codon\n",
    "        next_codon = store_best[pos][next_codon][2]\n",
    "\n",
    "# now restore the original sequences and patch up the recoded sequence with the beginning and end\n",
    "original_AA = long_original_AA\n",
    "original_codons = long_original_codons\n",
    "recoded_sequence = original_codons[0:120] + recoded_sequence + original_codons[-60:]\n",
    "\n",
    "print(original_codons[0:120] + best_first + store_best[0][best_first][3] + original_codons[-60:])\n",
    "\n",
    "print(\"\\n\\n\")\n",
    "\n",
    "\n",
    "#print(original_codons)\n",
    "print(recoded_sequence)    \n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "print(len(original_codons))\n",
    "print(\"num desired codon pairs in original sequence: \" + str(num_costly_codon_pairs(original_codons, 0)))\n",
    "print(\"num desired codon pairs in recoded sequence: \" + str(num_costly_codon_pairs(recoded_sequence, 0)))\n",
    "print(\"codon pair score of the original sequence: \" + str(get_cps(original_codons)))\n",
    "print(\"codon pair score of the recoded sequence: \" + str(get_cps(recoded_sequence)))\n",
    "print(\"CAI of the original sequence is: \" + str(get_cai(original_codons)))\n",
    "print(\"CAI of the recoded sequence is: \" + str(get_cai(recoded_sequence)))\n",
    "\n",
    "\n",
    "#'''and None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lys2 Version June2017 for Justin paper -  design 1: maximize 20% of worse frame dependent scores for yeast that do not have nnTAA/Gnn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "676\n",
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGTTGCCACAACTCGACGTGCCACATGATAGCTTCTCAAATAAGTACGCAGTCGCCTTGTCAGTGTGGGCCGCACTTATATACCGAGTGACCGGCGACGACGACATCGTGTTGTATATCGCTAATAATAAGATTTTGCGTTTTAATATACAACCAACCTGGAGCTTTAACGAGTTGTATAGTACGATTAATAACGAGCTTAATAAGTTGAATTCAATCGAAGCAAACTTTAGCTTCGATGAACTCGCCGAAAAGATCCAATCCTGTCAAGACCTCGAACGTACGCCACAATTGTTTCGTCTCGCTTTTCTCGAAAATCAAGACTTTAAGTTGGATGAATTCAAACACCATCTGGTCGATTTCGCACTTAATTTGGATACGTCAAATAACGCCCATGTGCTTAATCTTATATATAATTCCTTGTTGTATTCAAACGAGAGGGTTACAATCGTCGCCGATCAATTTACACAATATCTGACCGCCGCCTTGTCCGATCCATCAAACTGCATTACAAAGATCTCACTTATAACCGCGAGCTCAAAGGATAGCCTGCCAGATCCAACCAAGAATTTGGGTTGGTGCGACTTCGTGGGCTGCATCCACGACATCTTTCAAGATAACGCCGAAGCATTCCCGGAACGCACTTGCGTCGTCGAAACCCCAACCTTGAATTCCGATAAGTCTAGGAGCTTTACATACCGAGATATTAACCGTACGTCAAATATCGTCGCCCACTACCTTATAAAGACCGGGATTAAGCGGGGCGACGTCGTGATGATCTATTCAAGTCGTGGGGTCGATCTTATGGTGTGCGTGATGGGCGTGCTTAAGGCCGGGGCTACCTTCTCAGTGATCGATCCAGCCTACCCCCCGGCCCGACAGACTATCTATCTCGGAGTCGCTAAGCCACGGGGCTTGATCGTGATTAGGGCCGCGGGTCAACTTGATCAACTCGTCGAAGATTACATTAACGATGAACTCGAAATCGTGTCACGTATAAACTCAATCGCTATACAAGAAAACGGTACGATCGAAGGCGGCAAGTTGGATAACGGCGAAGACGTGCTCGCACCATACGATCATTACAAAGATACGCGGACCGGGGTCGTCGTGGGCCCGGATTCAAATCCAACCTTGTCTTTTACATCGGGTTCCGAAGGCATTCCAAAAGGCGTGTTGGGTCGACACTTTAGCCTCGCCTACTACTTTAACTGGATGTCAAAACGCTTTAACCTGACCGAAAACGATAAGTTTACTATGTTGTCCGGAATCGCCCACGACCCAATCCAAAGGGATATGTTTACACCATTGTTTCTCGGAGCCCAATTGTACGTCCCGACCCAAGACGATATAGGCACGCCAGGGAGACTCGCAGAGTGGATGTCAAAGTACGGCTGCACCGTGACCCACCTGACCCCAGCTATGGGTCAACTCCTGACCGCCCAAGCAACTACTCCTTTTCCAAAACTCCACCATGCTTTCTTCGTGGGCGACATTTTGACAAAACGCGATTGCTTGCGTCTCCAAACCCTCGCCGAAAACTGCAGGATCGTGAATATGTACGGCACGACCGAAACCCAGAGGGCCGTGAGCTACTTTGAAGTTAAGTCAAAGAACGATGACCCAAACTTTCTTAAGAAGTTGAAAGACGTGATGCCAGCCGGCAAAGGCATGTTGAACGTGCAACTCCTCGTCGTGAATAGGAATGATCGGACTCAGATTTGCGGCATTGGAGAAATAGGCGAGATTTACGTGCGCGCCGGGGGTTTGGCCGAAGGCTACCGAGGTTTGCCAGAACTTAATAAAGAAAAGTTCGTGAATAACTGGTTCGTTGAAAAGGATCATTGGAACTACTTGGATAAGGATAACGGCGAACCGTGGCGTCAATTTTGGTTGGGTCCACGTGATCGCTTGTACCGGACTGGCGACTTGGGTCGGTATTTGCCAAACGGCGATTGCGAATGCTGCGGTCGCGCCGATGATCAAGTTAAGATCCGTGGTTTTCGTATCGAACTCGGAGAAATCGATACCCATATTAGCCAACACCCTCTCGTGAGGGAAAATATTACACTCGTGCGTAAGAACGCCGATAACGAACCAACACTTATAACCTTTATGGTGCCACGTTTCGATAAACCAGATGATTTGTCAAAGTTTCAATCCGACGTGCCAAAAGAAGTCGAAACCGATCCAATCGTGAAAGGCCTTATAGGCTATCACCTGTTGTCAAAAGACATCCGTACATTCCTGAAAAAACGCCTCGCATCATACGCTATGCCAAGCCTGATCGTCGTGATGGATAAGTTGCCACTTAATCCAAACGGCAAGGTCGATAAACCAAAACTCCAATTTCCAACCCCAAAACAACTTAATTTGGTCGCCGAAAATACCGTGTCCGAAACCGATGATAGCCAATTTACAAATGTCGAACGCGAAGTTAGGGATCTGTGGTTGTCAATTTTGCCAACCAAACCAGCCTCAGTGAGCCCAGATGATAGCTTCTTCGACCTCGGAGGTCATTCAATTCTCGCAACTAAGATGATCTTTACACTTAAGAAGAAGCTCCAAGTTGATCTGCCATTGGGTACGATCTTTAAGTACCCAACGATTAAAGCATTCGCCGCCGAAATCGATAGGATTAAGTCCTCGGGCGGCTCAAGTCAGGGTGAGGTCGTCGAAAACGTGACCGCTAACTACGCCGAAGACGCAAAGAAGTTGGTCGAAACCTTGCCATCCTCATACCCATCACGTGAATATTTCGTTGAACCAAACTCAGCCGAAGGCAAAACTACTATTAACGTGTTCGTGACCGGCGTGACCGGTTTTCTGGGCTCATATATACTCGCCGATTTGTTGGGTCGGAGCCCAAAAAATTATAGCTTTAAGGTGTTCGCCCATGTGCGCGCTAAGGATGAAGAAGCAGCTTTCGCTAGACTTCAGAAGGCCGGGATTACATACGGCACGTGGAACGAGAAGTTCGCCTCAAATATAAAGGTCGTGTTGGGCGACTTGTCAAAGTCACAATTCGGTTTGTCCGATGAAAAGTGGATGGACCTCGCTAATACGGTCGATATAATTATCCACAACGGCGCACTCGTGCACTGGGTGTACCCATACGCTAAGTTGCGTGATCCAAACGTGATCTCGACTATTAACGTGATGAGCCTCGCAGCAGTGGGCAAACCAAAATTCTTCGATTTCGTGAGCTCAACCTCGACCTTGGATACCGAATATTACTTTAACCTGTCCGATAAGTTGGTGTCCGAAGGCAAACCAGGGATTCTCGAATCCGATGACCTTATGAACTCAGCCTCGGGTCTGACCGGGGGTTACGGTCAATCAAAGTGGGCCGCCGAATATATAATTAGGCGCGCCGGGGAACGCGGTTTGCGGGGCTGCATCGTGCGTCCAGGGTACGTGACCGGGGCCTCGGCTAACGGTTCCTCAAATACCGATGACTTTCTCCTCCGTTTTCTTAAAGGCTCAGTTCAATTGGGCAAGATCCCTGATATCGAAAACTCAGTTAACATGGTGCCAGTCGATCATGTCGCGAGGGTCGTCGTCGCGACCTCCCTCAACCCACCAAAAGAAAACGAGCTCGCAGTCGCCCAAGTTACGGGTCACCCTAGGATTTTGTTTAAGGATTACTTGTATACTCTCCATGATTACGGATACGATGTCGAAATCGAATCCTATTCAAAGTGGAAGAAGTCCCTCGAAGCATCAGTGATCGATAGGAATGAAGAAAACGCCTTGTACCCACTCCTCCATATGGTGTTGGATAACCTGCCAGAATCGACTAAGGCCCCGGAATTGGATGATAGGAACGCCGTCGCGAGCCTGAAGAAGGATACGGCCTGGACCGGGGTCGATTGGTCAAACGGCATAGGCGTGACCCCAGAAGAAGTGGGCATTTATATCGCTTTTCTTAATAAGGTGGGTTTTCTGCCGCCGCCGACCCATAACGATAAGTTGCCATTGCCATCAATCGAACTTACACAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "\n",
      "\n",
      "\n",
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGTTGCCACAACTCGACGTGCCACATGATAGCTTCTCAAATAAGTACGCAGTCGCCTTGTCAGTGTGGGCCGCACTTATATACCGAGTGACCGGCGACGACGACATCGTGTTGTATATCGCTAATAATAAGATTTTGCGTTTTAATATACAACCAACCTGGAGCTTTAACGAGTTGTATAGTACGATTAATAACGAGCTTAATAAGTTGAATTCAATCGAAGCAAACTTTAGCTTCGATGAACTCGCCGAAAAGATCCAATCCTGTCAAGACCTCGAACGTACGCCACAATTGTTTCGTCTCGCTTTTCTCGAAAATCAAGACTTTAAGTTGGATGAATTCAAACACCATCTGGTCGATTTCGCACTTAATTTGGATACGTCAAATAACGCCCATGTGCTTAATCTTATATATAATTCCTTGTTGTATTCAAACGAGAGGGTTACAATCGTCGCCGATCAATTTACACAATATCTGACCGCCGCCTTGTCCGATCCATCAAACTGCATTACAAAGATCTCACTTATAACCGCGAGCTCAAAGGATAGCCTGCCAGATCCAACCAAGAATTTGGGTTGGTGCGACTTCGTGGGCTGCATCCACGACATCTTTCAAGATAACGCCGAAGCATTCCCGGAACGCACTTGCGTCGTCGAAACCCCAACCTTGAATTCCGATAAGTCTAGGAGCTTTACATACCGAGATATTAACCGTACGTCAAATATCGTCGCCCACTACCTTATAAAGACCGGGATTAAGCGGGGCGACGTCGTGATGATCTATTCAAGTCGTGGGGTCGATCTTATGGTGTGCGTGATGGGCGTGCTTAAGGCCGGGGCTACCTTCTCAGTGATCGATCCAGCCTACCCCCCGGCCCGACAGACTATCTATCTCGGAGTCGCTAAGCCACGGGGCTTGATCGTGATTAGGGCCGCGGGTCAACTTGATCAACTCGTCGAAGATTACATTAACGATGAACTCGAAATCGTGTCACGTATAAACTCAATCGCTATACAAGAAAACGGTACGATCGAAGGCGGCAAGTTGGATAACGGCGAAGACGTGCTCGCACCATACGATCATTACAAAGATACGCGGACCGGGGTCGTCGTGGGCCCGGATTCAAATCCAACCTTGTCTTTTACATCGGGTTCCGAAGGCATTCCAAAAGGCGTGTTGGGTCGACACTTTAGCCTCGCCTACTACTTTAACTGGATGTCAAAACGCTTTAACCTGACCGAAAACGATAAGTTTACTATGTTGTCCGGAATCGCCCACGACCCAATCCAAAGGGATATGTTTACACCATTGTTTCTCGGAGCCCAATTGTACGTCCCGACCCAAGACGATATAGGCACGCCAGGGAGACTCGCAGAGTGGATGTCAAAGTACGGCTGCACCGTGACCCACCTGACCCCAGCTATGGGTCAACTCCTGACCGCCCAAGCAACTACTCCTTTTCCAAAACTCCACCATGCTTTCTTCGTGGGCGACATTTTGACAAAACGCGATTGCTTGCGTCTCCAAACCCTCGCCGAAAACTGCAGGATCGTGAATATGTACGGCACGACCGAAACCCAGAGGGCCGTGAGCTACTTTGAAGTTAAGTCAAAGAACGATGACCCAAACTTTCTTAAGAAGTTGAAAGACGTGATGCCAGCCGGCAAAGGCATGTTGAACGTGCAACTCCTCGTCGTGAATAGGAATGATCGGACTCAGATTTGCGGCATTGGAGAAATAGGCGAGATTTACGTGCGCGCCGGGGGTTTGGCCGAAGGCTACCGAGGTTTGCCAGAACTTAATAAAGAAAAGTTCGTGAATAACTGGTTCGTTGAAAAGGATCATTGGAACTACTTGGATAAGGATAACGGCGAACCGTGGCGTCAATTTTGGTTGGGTCCACGTGATCGCTTGTACCGGACTGGCGACTTGGGTCGGTATTTGCCAAACGGCGATTGCGAATGCTGCGGTCGCGCCGATGATCAAGTTAAGATCCGTGGTTTTCGTATCGAACTCGGAGAAATCGATACCCATATTAGCCAACACCCTCTCGTGAGGGAAAATATTACACTCGTGCGTAAGAACGCCGATAACGAACCAACACTTATAACCTTTATGGTGCCACGTTTCGATAAACCAGATGATTTGTCAAAGTTTCAATCCGACGTGCCAAAAGAAGTCGAAACCGATCCAATCGTGAAAGGCCTTATAGGCTATCACCTGTTGTCAAAAGACATCCGTACATTCCTGAAAAAACGCCTCGCATCATACGCTATGCCAAGCCTGATCGTCGTGATGGATAAGTTGCCACTTAATCCAAACGGCAAGGTCGATAAACCAAAACTCCAATTTCCAACCCCAAAACAACTTAATTTGGTCGCCGAAAATACCGTGTCCGAAACCGATGATAGCCAATTTACAAATGTCGAACGCGAAGTTAGGGATCTGTGGTTGTCAATTTTGCCAACCAAACCAGCCTCAGTGAGCCCAGATGATAGCTTCTTCGACCTCGGAGGTCATTCAATTCTCGCAACTAAGATGATCTTTACACTTAAGAAGAAGCTCCAAGTTGATCTGCCATTGGGTACGATCTTTAAGTACCCAACGATTAAAGCATTCGCCGCCGAAATCGATAGGATTAAGTCCTCGGGCGGCTCAAGTCAGGGTGAGGTCGTCGAAAACGTGACCGCTAACTACGCCGAAGACGCAAAGAAGTTGGTCGAAACCTTGCCATCCTCATACCCATCACGTGAATATTTCGTTGAACCAAACTCAGCCGAAGGCAAAACTACTATTAACGTGTTCGTGACCGGCGTGACCGGTTTTCTGGGCTCATATATACTCGCCGATTTGTTGGGTCGGAGCCCAAAAAATTATAGCTTTAAGGTGTTCGCCCATGTGCGCGCTAAGGATGAAGAAGCAGCTTTCGCTAGACTTCAGAAGGCCGGGATTACATACGGCACGTGGAACGAGAAGTTCGCCTCAAATATAAAGGTCGTGTTGGGCGACTTGTCAAAGTCACAATTCGGTTTGTCCGATGAAAAGTGGATGGACCTCGCTAATACGGTCGATATAATTATCCACAACGGCGCACTCGTGCACTGGGTGTACCCATACGCTAAGTTGCGTGATCCAAACGTGATCTCGACTATTAACGTGATGAGCCTCGCAGCAGTGGGCAAACCAAAATTCTTCGATTTCGTGAGCTCAACCTCGACCTTGGATACCGAATATTACTTTAACCTGTCCGATAAGTTGGTGTCCGAAGGCAAACCAGGGATTCTCGAATCCGATGACCTTATGAACTCAGCCTCGGGTCTGACCGGGGGTTACGGTCAATCAAAGTGGGCCGCCGAATATATAATTAGGCGCGCCGGGGAACGCGGTTTGCGGGGCTGCATCGTGCGTCCAGGGTACGTGACCGGGGCCTCGGCTAACGGTTCCTCAAATACCGATGACTTTCTCCTCCGTTTTCTTAAAGGCTCAGTTCAATTGGGCAAGATCCCTGATATCGAAAACTCAGTTAACATGGTGCCAGTCGATCATGTCGCGAGGGTCGTCGTCGCGACCTCCCTCAACCCACCAAAAGAAAACGAGCTCGCAGTCGCCCAAGTTACGGGTCACCCTAGGATTTTGTTTAAGGATTACTTGTATACTCTCCATGATTACGGATACGATGTCGAAATCGAATCCTATTCAAAGTGGAAGAAGTCCCTCGAAGCATCAGTGATCGATAGGAATGAAGAAAACGCCTTGTACCCACTCCTCCATATGGTGTTGGATAACCTGCCAGAATCGACTAAGGCCCCGGAATTGGATGATAGGAACGCCGTCGCGAGCCTGAAGAAGGATACGGCCTGGACCGGGGTCGATTGGTCAAACGGCATAGGCGTGACCCCAGAAGAAGTGGGCATTTATATCGCTTTTCTTAATAAGGTGGGTTTTCTGCCGCCGCCGACCCATAACGATAAGTTGCCATTGCCATCAATCGAACTTACACAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "total number of codon diferences between original and recoded sequences 879\n",
      "AA sequence remains the same, good\n",
      "4179\n",
      "num desired codon pairs in original sequence: 177\n",
      "num desired codon pairs in recoded sequence: 787\n",
      "codon pair score of the original sequence: 24.055618279000008\n",
      "codon pair score of the recoded sequence: -319.2213258829996\n",
      "CAI of the original sequence is: 0.783213534519\n",
      "CAI of the recoded sequence is: 0.670211771995\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "#import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "\n",
    "\n",
    "# number of desired codon paris in sequence of any length\n",
    "def num_costly_codon_pairs(seq, count_negatives):\n",
    "    num = 0\n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] in pairs_to_maximize):\n",
    "            num += 1\n",
    "    return(num)\n",
    "\n",
    "# codon pair score of any length sequence\n",
    "def get_cps(seq):\n",
    "    cps = 0\n",
    "    if (len(seq) == 6):\n",
    "        if (seq not in cp_scores):\n",
    "            print(\"Missing \" + seq[i:i+6])\n",
    "        return(cp_scores[seq])  \n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        if (seq[i:i+6] not in cp_scores):\n",
    "            print(\"Missing \" + seq[i:i+6])\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    return(cps)\n",
    "\n",
    "# codon adaptation index of any length sequence\n",
    "def get_cai(seq):\n",
    "    # get the CAI as the geometric mean of all the codon's individual CAI values\n",
    "    cai_list = []\n",
    "    for i in range(0,len(seq),3):\n",
    "        cai_list.append(codon_CAI[seq[i:(i+3)]])\n",
    "    return(scistats.gmean(cai_list))\n",
    "\n",
    "# read in synonymous codons\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "\n",
    "# read in the yeast orfs and make the CAI for all the codons\n",
    "codon_occs = dict()\n",
    "# set up the dict\n",
    "for codon in (codon_to_aa_dict.keys()):\n",
    "    codon_occs[codon] = 0\n",
    "genes_file = open(\"s_cerevisae_orf_coding_no_Mito_no_Plasmid.fasta\")\n",
    "for line in genes_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\": \n",
    "        if (len(line)%3 != 0):\n",
    "            print(\"PROBLEM, genes file is not multiple of 3 per line.\")\n",
    "        # get the frequency of each codon\n",
    "        for i in range(0,len(line),3):\n",
    "            codon_occs[line[i:(i+3)]] += 1\n",
    "# get the CAI of each codon\n",
    "codon_CAI = dict()\n",
    "for aa in (aa_to_syn_codons_dict.keys()):\n",
    "    highest_frequency = 0\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (codon_occs[codon] > highest_frequency):\n",
    "            highest_frequency = codon_occs[codon]\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (highest_frequency == 0):\n",
    "            codon_CAI[codon] = 0\n",
    "        else:\n",
    "            codon_CAI[codon] = codon_occs[codon]/highest_frequency\n",
    "        \n",
    "# read in the codon pair scores\n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"cerevisiae_FD_Scores.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "counter = 0\n",
    "pairs_to_maximize = []\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[0]] = float(line[1])\n",
    "    counter += 1\n",
    "    if (counter <= 744):\n",
    "        if (line[0][2] == \"T\" and line[0][3] == \"A\" and ((line[0][4] == \"A\" or line[0][4] == \"G\"))):\n",
    "            continue\n",
    "        pairs_to_maximize.append(line[0])\n",
    "print(len(pairs_to_maximize))\n",
    "\n",
    "# read in the gene's codon sequence\n",
    "codons_file = open(\"S288C_YBR115C_LYS2_genomic.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "# read in the gene's amino acid sequence\n",
    "aa_file = open(\"S288C_YBR115C_LYS2_protein.fsa\")\n",
    "original_AA = \"\"\n",
    "for l in aa_file:\n",
    "    l = l.rstrip()\n",
    "    if l[0] != \">\":\n",
    "        original_AA = original_AA + l\n",
    "aa_file.close()\n",
    "\n",
    "if (len(original_AA)*3 != len(original_codons)):\n",
    "    print(\"PROBLEM: codons and AA don't match\")\n",
    "\n",
    "    \n",
    "# store the sequences, and then work with the ones that have \"chopped off\" start and end\n",
    "long_original_AA = original_AA\n",
    "original_AA = original_AA[40:-20]\n",
    "long_original_codons = original_codons\n",
    "original_codons = original_codons[120:-60]\n",
    "   \n",
    "# store_best will contain, at each AA position, and at each of the syn codon choices,\n",
    "# the best cost of the suffix that starts with this codon, the current codon pair score,\n",
    "# and the next codon that corresponds to this cost\n",
    "store_best = [None] * len(original_AA)\n",
    "for i in range(0,len(original_AA)):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = 0\n",
    "                \n",
    "\n",
    "# initialize the dynamic program by setting store_best for the last amino acid position\n",
    "for codon in store_best[-1]:\n",
    "    store_best[i][codon] = (0,0,\"\",\"\")\n",
    "\n",
    "\n",
    "# do the dynamic program, stating from the one before the end of the list, until the beginning\n",
    "for pos in reversed(range(0, len(store_best)-1)):\n",
    "    for codon in store_best[pos]:\n",
    "        # consider as cost the cost of adding this codon to one of the best sequences\n",
    "        # from the previous suffix\n",
    "        best_prev = \"\"\n",
    "        best_cps = sys.maxsize # keep this\n",
    "        best_cost = -1 #sys.maxsize #-1\n",
    "        best_seq = \"\"\n",
    "        for prev_codon in store_best[pos+1]:\n",
    "            cost = num_costly_codon_pairs(codon+prev_codon,1) + store_best[pos+1][prev_codon][0]\n",
    "            cp_score = get_cps(codon+prev_codon) + store_best[pos+1][prev_codon][1]\n",
    "            #print(cp_score)\n",
    "            # maximizer the cost\n",
    "            #if (cp_score < best_cps):\n",
    "            if ((cost > best_cost) or # > <\n",
    "                # break the tie with lowest codon pair score\n",
    "                ((cost == best_cost) and (cp_score < best_cps))):\n",
    "                best_cost = cost\n",
    "                best_cps = cp_score\n",
    "                best_prev = prev_codon\n",
    "                best_seq = prev_codon + store_best[pos+1][prev_codon][3]\n",
    "        store_best[pos][codon] = (best_cost, best_cps, best_prev, best_seq)\n",
    "        #print(\"storing for \" + str(pos) + \" \" + codon + \": \" + str(best_cps) + \" \" + best_prev)\n",
    "\n",
    "# going from the beginning to the end, read off the recoded sequences from the store_best\n",
    "# re-doing the computation for the first position, to pick the best of the syn codons\n",
    "# the rest of the seuqnces is just read out\n",
    "recoded_sequence = \"\"\n",
    "next_codon = \"\"\n",
    "for pos in range(0, len(store_best)):\n",
    "    if pos == 0:\n",
    "        # maximum\n",
    "        best_first = \"\"\n",
    "        best_first_cps = sys.maxsize # keep this\n",
    "        best_first_cost = -1 # sys.maxsize #-1\n",
    "        for codon in store_best[0]:\n",
    "            # maximize the cost\n",
    "            #if (store_best[pos][codon][1] < best_first_cps):\n",
    "            if ((store_best[pos][codon][0] > best_first_cost) or  #> <\n",
    "               ((store_best[pos][codon][0] == best_first_cost) and (store_best[pos][codon][1] < best_first_cps))):\n",
    "                best_first = codon\n",
    "                best_first_cost = store_best[pos][codon][0]\n",
    "                best_first_cps = store_best[pos][codon][1]\n",
    "                next_codon = store_best[pos][codon][2]\n",
    "        recoded_sequence += best_first\n",
    "    else:\n",
    "        recoded_sequence += next_codon\n",
    "        next_codon = store_best[pos][next_codon][2]\n",
    "\n",
    "# now restore the original sequences and patch up the recoded sequence with the beginning and end\n",
    "original_AA = long_original_AA\n",
    "original_codons = long_original_codons\n",
    "recoded_sequence = original_codons[0:120] + recoded_sequence + original_codons[-60:]\n",
    "\n",
    "print(original_codons[0:120] + best_first + store_best[0][best_first][3] + original_codons[-60:])\n",
    "\n",
    "print(\"\\n\\n\")\n",
    "\n",
    "\n",
    "#print(original_codons)\n",
    "print(recoded_sequence)    \n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "print(len(original_codons))\n",
    "print(\"num desired codon pairs in original sequence: \" + str(num_costly_codon_pairs(original_codons, 0)))\n",
    "print(\"num desired codon pairs in recoded sequence: \" + str(num_costly_codon_pairs(recoded_sequence, 0)))\n",
    "print(\"codon pair score of the original sequence: \" + str(get_cps(original_codons)))\n",
    "print(\"codon pair score of the recoded sequence: \" + str(get_cps(recoded_sequence)))\n",
    "print(\"CAI of the original sequence is: \" + str(get_cai(original_codons)))\n",
    "print(\"CAI of the recoded sequence is: \" + str(get_cai(recoded_sequence)))\n",
    "\n",
    "\n",
    "#'''and None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lys2 Version 2 - sent to Bruce on Jan 27 \n",
    "## try to minimize! the codon pair score in the cost function\n",
    "## no longer care about keeping the rest of the gene wt\n",
    "The minimum possible codon pair score for for [120:-60], without caring for any other sequences is -293.61"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGCTCCCACAACTTGATGTCCCGCATGATAGCTTTAGTAATAAGTATGCAGTTGCACTTAGTGTGTGGGCAGCCCTAATCTATAGAGTTACAGGGGATGATGATATAGTCTTGTATATAGCTAATAATAAGATACTTAGATTTAATATACAACCGACCTGGAGCTTTAATGAATTGTATAGTACGATTAATAATGAACTTAATAAGCTTAATAGTATAGAAGCTAACTTTAGCTTTGATGAATTAGCTGAGAAGATACAATCCTGTCAGGATCTTGAGCGGACCCCCCAATTGTTTAGACTAGCTTTTCTTGAGAATCAAGACTTTAAGTTAGATGAATTTAAACACCACCTAGTTGACTTTGCACTTAATCTAGATACTAGTAATAATGCCCATGTACTTAATCTTATCTATAATAGCTTGTTGTATAGTAATGAGCGGGTTACGATTGTAGCTGATCAATTTACACAATACCTTACAGCTGCACTTAGTGATCCTAGTAATTGTATAACTAAGATTAGTCTAATTACAGCTAGTAGTAAGGATAGCCTCCCAGACCCGACTAAGAATCTAGGTTGGTGTGACTTTGTAGGTTGTATACATGACATCTTTCAGGATAATGCTGAGGCTTTTCCAGAGCGGACTTGTGTAGTTGAGACCCCGACCCTTAATAGTGATAAGTCTAGATCTTTTACATATAGAGACATTAATAGGACTAGTAACATAGTAGCCCACTACCTAATTAAGACTGGGATTAAGCGGGGGGATGTAGTTATGATCTATAGTAGTAGGGGGGTTGATCTTATGGTGTGTGTAATGGGAGTCCTTAAAGCGGGGGCGACCTTTAGTGTTATAGATCCAGCCTACCCCCCGGCTAGACAGACTATCTATCTAGGAGTAGCTAAGCCTAGGGGACTTATAGTAATTAGAGCGGCGGGTCAGTTAGATCAACTAGTTGAGGATTACATTAATGATGAACTTGAGATAGTTAGTAGGATTAATAGTATAGCTATACAAGAGAATGGTACGATTGAGGGGGGTAAGTTAGATAATGGTGAGGATGTGTTAGCCCCGTATGATCATTATAAAGATACTAGGACTGGAGTAGTTGTAGGTCCGGATAGTAACCCGACCCTTAGCTTTACTAGTGGTAGTGAGGGGATCCCTAAGGGGGTCCTAGGTAGGCACTTTAGTCTAGCCTATTACTTTAATTGGATGTCTAAACGCTTTAATCTTACAGAGAATGATAAGTTTACTATGCTTAGTGGGATAGCCCATGATCCTATACAGAGGGATATGTTTACACCCCTGTTTCTAGGAGCCCAGTTGTATGTCCCGACCCAGGATGATATAGGGACCCCGGGTAGGTTAGCTGAGTGGATGTCTAAGTATGGGTGTACAGTTACACACCTAACCCCAGCTATGGGTCAGTTGCTTACAGCCCAGGCAACGACCCCTTTTCCTAAGTTACATCATGCTTTTTTTGTAGGGGATATACTTACTAAGCGGGATTGCCTTAGACTTCAGACCCTAGCTGAGAATTGTAGAATAGTTAACATGTATGGCACGACGGAGACCCAGAGAGCAGTTAGTTACTTTGAGGTTAAGTCTAAGAATGATGATCCTAACTTCCTTAAAAAACTTAAGGATGTTATGCCGGCGGGTAAGGGGATGCTTAATGTACAACTCCTTGTAGTTAATCGTAATGATAGGACACAAATCTGTGGAATAGGTGAGATAGGTGAGATCTATGTTAGAGCGGGGGGTCTGGCTGAGGGTTATAGGGGTCTGCCAGAACTTAATAAAGAAAAATTTGTTAATAATTGGTTTGTAGAAAAAGATCATTGGAATTACTTAGATAAGGATAATGGAGAACCCTGGCGTCAATTCTGGTTGGGTCCTAGGGATAGGTTGTATAGGACAGGGGATCTAGGTAGGTACCTCCCTAATGGGGATTGTGAATGTTGTGGTAGGGCGGATGATCAGGTTAAGATTAGGGGTTTTAGGATAGAATTAGGGGAGATAGATACCCATATTAGTCAGCACCCTCTAGTTAGGGAGAATATAACCCTAGTTAGAAAAAATGCAGATAATGAGCCGACCCTAATTACCTTTATGGTCCCTAGATTTGATAAGCCAGATGATCTTAGTAAGTTTCAGTCTGATGTGCCTAAAGAAGTTGAGACTGACCCGATAGTTAAAGGACTTATAGGTTACCACCTCCTTAGTAAGGATATTAGGACATTCCTTAAAAAACGCCTAGCTAGCTATGCTATGCCTAGCCTTATAGTAGTTATGGATAAGTTGCCACTTAATCCTAATGGTAAGGTAGATAAGCCTAAGTTGCAATTCCCGACCCCTAAACAACTTAATCTTGTAGCTGAGAATACAGTTAGTGAGACTGATGATAGTCAATTTACTAATGTAGAACGTGAGGTTAGGGATCTGTGGCTTAGTATACTCCCGACTAAGCCAGCTAGTGTTAGTCCAGATGATAGCTTTTTTGATCTAGGGGGTCATAGTATTCTAGCGACTAAGATGATCTTTACACTTAAAAAAAAACTTCAGGTTGATCTCCCTCTAGGCACGATCTTTAAGTACCCGACGATTAAAGCTTTTGCAGCTGAGATAGATAGGATTAAGTCTAGTGGGGGTAGTAGTCAGGGGGAAGTAGTTGAGAATGTTACAGCTAATTATGCAGAGGATGCTAAAAAGTTAGTTGAGACCCTCCCTAGTAGTTACCCTAGTAGAGAATACTTTGTAGAACCTAATAGTGCTGAGGGTAAAACGACGATTAATGTGTTTGTGACTGGAGTTACGGGTTTTCTAGGTAGTTACATTCTAGCTGACTTGTTAGGTAGGAGTCCTAAGAATTATAGCTTTAAGGTGTTTGCACATGTTAGAGCTAAGGATGAAGAAGCAGCTTTTGCTAGACTCCAAAAAGCGGGGATTACCTATGGCACGTGGAATGAAAAATTTGCTAGTAATATTAAGGTAGTCCTAGGGGATCTTAGTAAGTCCCAATTTGGACTTAGTGATGAGAAGTGGATGGATCTAGCTAACACGGTTGACATAATTATACATAATGGAGCCTTAGTACACTGGGTGTACCCCTATGCTAAGCTTAGGGATCCTAATGTGATTAGTACGATTAATGTTATGAGTCTAGCAGCGGTTGGTAAGCCTAAGTTCTTTGACTTTGTTAGTAGTACTAGTACCCTAGATACGGAATATTACTTTAATCTTAGTGATAAGTTAGTTAGTGAGGGTAAGCCAGGGATCCTTGAGTCGGATGATCTTATGAATAGTGCTAGTGGACTTACAGGGGGTTATGGTCAGTCTAAGTGGGCAGCGGAATATATAATTAGACGAGCGGGGGAGAGGGGACTTAGGGGGTGTATAGTTAGACCGGGTTATGTAACTGGAGCTAGTGCTAATGGTAGTAGTAATACTGATGACTTCTTGCTTAGATTCCTTAAGGGTAGTGTCCAACTAGGTAAGATCCCAGACATAGAGAATAGTGTTAACATGGTCCCGGTTGATCATGTAGCTAGGGTTGTAGTTGCGACTAGCCTTAATCCCCCTAAGGAGAATGAATTAGCTGTAGCCCAGGTTACGGGTCACCCTAGGATCTTGTTTAAGGATTACTTATACACCCTACATGATTATGGGTATGATGTTGAGATAGAATCCTATAGTAAGTGGAAAAAGTCCCTTGAGGCTAGTGTTATAGATAGGAATGAAGAGAATGCCTTGTACCCCCTGTTACATATGGTGTTAGATAACCTCCCGGAATCGACTAAAGCCCCGGAATTAGATGATAGGAATGCAGTTGCTAGCCTTAAAAAAGATACAGCCTGGACTGGAGTTGACTGGTCTAATGGGATAGGAGTTACCCCTGAGGAAGTAGGGATCTATATAGCTTTTCTTAATAAGGTAGGTTTTCTACCCCCCCCGACCCATAATGATAAGTTGCCACTCCCTAGCATAGAACTTACACAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "\n",
      "\n",
      "\n",
      "ATGACTAACGAAAAGGTCTGGATAGAGAAGTTGGATAATCCAACTCTTTCAGTGTTACCACATGACTTTTTACGCCCACAACAAGAACCTTATACGAAACAAGCTACATATTCGTTACAGCTCCCACAACTTGATGTCCCGCATGATAGCTTTAGTAATAAGTATGCAGTTGCACTTAGTGTGTGGGCAGCCCTAATCTATAGAGTTACAGGGGATGATGATATAGTCTTGTATATAGCTAATAATAAGATACTTAGATTTAATATACAACCGACCTGGAGCTTTAATGAATTGTATAGTACGATTAATAATGAACTTAATAAGCTTAATAGTATAGAAGCTAACTTTAGCTTTGATGAATTAGCTGAGAAGATACAATCCTGTCAGGATCTTGAGCGGACCCCCCAATTGTTTAGACTAGCTTTTCTTGAGAATCAAGACTTTAAGTTAGATGAATTTAAACACCACCTAGTTGACTTTGCACTTAATCTAGATACTAGTAATAATGCCCATGTACTTAATCTTATCTATAATAGCTTGTTGTATAGTAATGAGCGGGTTACGATTGTAGCTGATCAATTTACACAATACCTTACAGCTGCACTTAGTGATCCTAGTAATTGTATAACTAAGATTAGTCTAATTACAGCTAGTAGTAAGGATAGCCTCCCAGACCCGACTAAGAATCTAGGTTGGTGTGACTTTGTAGGTTGTATACATGACATCTTTCAGGATAATGCTGAGGCTTTTCCAGAGCGGACTTGTGTAGTTGAGACCCCGACCCTTAATAGTGATAAGTCTAGATCTTTTACATATAGAGACATTAATAGGACTAGTAACATAGTAGCCCACTACCTAATTAAGACTGGGATTAAGCGGGGGGATGTAGTTATGATCTATAGTAGTAGGGGGGTTGATCTTATGGTGTGTGTAATGGGAGTCCTTAAAGCGGGGGCGACCTTTAGTGTTATAGATCCAGCCTACCCCCCGGCTAGACAGACTATCTATCTAGGAGTAGCTAAGCCTAGGGGACTTATAGTAATTAGAGCGGCGGGTCAGTTAGATCAACTAGTTGAGGATTACATTAATGATGAACTTGAGATAGTTAGTAGGATTAATAGTATAGCTATACAAGAGAATGGTACGATTGAGGGGGGTAAGTTAGATAATGGTGAGGATGTGTTAGCCCCGTATGATCATTATAAAGATACTAGGACTGGAGTAGTTGTAGGTCCGGATAGTAACCCGACCCTTAGCTTTACTAGTGGTAGTGAGGGGATCCCTAAGGGGGTCCTAGGTAGGCACTTTAGTCTAGCCTATTACTTTAATTGGATGTCTAAACGCTTTAATCTTACAGAGAATGATAAGTTTACTATGCTTAGTGGGATAGCCCATGATCCTATACAGAGGGATATGTTTACACCCCTGTTTCTAGGAGCCCAGTTGTATGTCCCGACCCAGGATGATATAGGGACCCCGGGTAGGTTAGCTGAGTGGATGTCTAAGTATGGGTGTACAGTTACACACCTAACCCCAGCTATGGGTCAGTTGCTTACAGCCCAGGCAACGACCCCTTTTCCTAAGTTACATCATGCTTTTTTTGTAGGGGATATACTTACTAAGCGGGATTGCCTTAGACTTCAGACCCTAGCTGAGAATTGTAGAATAGTTAACATGTATGGCACGACGGAGACCCAGAGAGCAGTTAGTTACTTTGAGGTTAAGTCTAAGAATGATGATCCTAACTTCCTTAAAAAACTTAAGGATGTTATGCCGGCGGGTAAGGGGATGCTTAATGTACAACTCCTTGTAGTTAATCGTAATGATAGGACACAAATCTGTGGAATAGGTGAGATAGGTGAGATCTATGTTAGAGCGGGGGGTCTGGCTGAGGGTTATAGGGGTCTGCCAGAACTTAATAAAGAAAAATTTGTTAATAATTGGTTTGTAGAAAAAGATCATTGGAATTACTTAGATAAGGATAATGGAGAACCCTGGCGTCAATTCTGGTTGGGTCCTAGGGATAGGTTGTATAGGACAGGGGATCTAGGTAGGTACCTCCCTAATGGGGATTGTGAATGTTGTGGTAGGGCGGATGATCAGGTTAAGATTAGGGGTTTTAGGATAGAATTAGGGGAGATAGATACCCATATTAGTCAGCACCCTCTAGTTAGGGAGAATATAACCCTAGTTAGAAAAAATGCAGATAATGAGCCGACCCTAATTACCTTTATGGTCCCTAGATTTGATAAGCCAGATGATCTTAGTAAGTTTCAGTCTGATGTGCCTAAAGAAGTTGAGACTGACCCGATAGTTAAAGGACTTATAGGTTACCACCTCCTTAGTAAGGATATTAGGACATTCCTTAAAAAACGCCTAGCTAGCTATGCTATGCCTAGCCTTATAGTAGTTATGGATAAGTTGCCACTTAATCCTAATGGTAAGGTAGATAAGCCTAAGTTGCAATTCCCGACCCCTAAACAACTTAATCTTGTAGCTGAGAATACAGTTAGTGAGACTGATGATAGTCAATTTACTAATGTAGAACGTGAGGTTAGGGATCTGTGGCTTAGTATACTCCCGACTAAGCCAGCTAGTGTTAGTCCAGATGATAGCTTTTTTGATCTAGGGGGTCATAGTATTCTAGCGACTAAGATGATCTTTACACTTAAAAAAAAACTTCAGGTTGATCTCCCTCTAGGCACGATCTTTAAGTACCCGACGATTAAAGCTTTTGCAGCTGAGATAGATAGGATTAAGTCTAGTGGGGGTAGTAGTCAGGGGGAAGTAGTTGAGAATGTTACAGCTAATTATGCAGAGGATGCTAAAAAGTTAGTTGAGACCCTCCCTAGTAGTTACCCTAGTAGAGAATACTTTGTAGAACCTAATAGTGCTGAGGGTAAAACGACGATTAATGTGTTTGTGACTGGAGTTACGGGTTTTCTAGGTAGTTACATTCTAGCTGACTTGTTAGGTAGGAGTCCTAAGAATTATAGCTTTAAGGTGTTTGCACATGTTAGAGCTAAGGATGAAGAAGCAGCTTTTGCTAGACTCCAAAAAGCGGGGATTACCTATGGCACGTGGAATGAAAAATTTGCTAGTAATATTAAGGTAGTCCTAGGGGATCTTAGTAAGTCCCAATTTGGACTTAGTGATGAGAAGTGGATGGATCTAGCTAACACGGTTGACATAATTATACATAATGGAGCCTTAGTACACTGGGTGTACCCCTATGCTAAGCTTAGGGATCCTAATGTGATTAGTACGATTAATGTTATGAGTCTAGCAGCGGTTGGTAAGCCTAAGTTCTTTGACTTTGTTAGTAGTACTAGTACCCTAGATACGGAATATTACTTTAATCTTAGTGATAAGTTAGTTAGTGAGGGTAAGCCAGGGATCCTTGAGTCGGATGATCTTATGAATAGTGCTAGTGGACTTACAGGGGGTTATGGTCAGTCTAAGTGGGCAGCGGAATATATAATTAGACGAGCGGGGGAGAGGGGACTTAGGGGGTGTATAGTTAGACCGGGTTATGTAACTGGAGCTAGTGCTAATGGTAGTAGTAATACTGATGACTTCTTGCTTAGATTCCTTAAGGGTAGTGTCCAACTAGGTAAGATCCCAGACATAGAGAATAGTGTTAACATGGTCCCGGTTGATCATGTAGCTAGGGTTGTAGTTGCGACTAGCCTTAATCCCCCTAAGGAGAATGAATTAGCTGTAGCCCAGGTTACGGGTCACCCTAGGATCTTGTTTAAGGATTACTTATACACCCTACATGATTATGGGTATGATGTTGAGATAGAATCCTATAGTAAGTGGAAAAAGTCCCTTGAGGCTAGTGTTATAGATAGGAATGAAGAGAATGCCTTGTACCCCCTGTTACATATGGTGTTAGATAACCTCCCGGAATCGACTAAAGCCCCGGAATTAGATGATAGGAATGCAGTTGCTAGCCTTAAAAAAGATACAGCCTGGACTGGAGTTGACTGGTCTAATGGGATAGGAGTTACCCCTGAGGAAGTAGGGATCTATATAGCTTTTCTTAATAAGGTAGGTTTTCTACCCCCCCCGACCCATAATGATAAGTTGCCACTCCCTAGCATAGAACTTACACAAGCGCAAATAAGTCTAGTTGCTTCAGGTGCTGGTGCTCGTGGAAGCTCCGCAGCAGCTTAA\n",
      "total number of codon diferences between original and recoded sequences 853\n",
      "AA sequence remains the same, good\n",
      "4179\n",
      "num desired codon pairs in original sequence: 59\n",
      "num desired codon pairs in recoded sequence: 245\n",
      "codon pair score of the original sequence: 34.78041769633767\n",
      "codon pair score of the recoded sequence: -210.32227221086487\n",
      "CAI of the original sequence is: 0.784752669237\n",
      "CAI of the recoded sequence is: 0.70392201631\n",
      "num opposite in original: 59\n",
      "num opposite in recoded: 245\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "#import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "\n",
    "def get_count(seq):\n",
    "    num = 0\n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "            num += 1\n",
    "    return(num)\n",
    "\n",
    "# number of desired codon paris in sequence of any length\n",
    "def num_costly_codon_pairs(seq, count_negatives):\n",
    "    num = 0\n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "            num += 1\n",
    "        if (count_negatives):\n",
    "            if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "                num -= 1\n",
    "        '''if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "            num += 1\n",
    "        if (count_negatives):\n",
    "            if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "                num -= 1'''\n",
    "        '''if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"T\" or seq[i+4] == \"C\"))):\n",
    "            num += 1\n",
    "        if (count_negatives):\n",
    "            if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "                num -= 1\n",
    "        if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "            num -= 1'''\n",
    "    return(num)\n",
    "\n",
    "# codon pair score of any length sequence\n",
    "def get_cps(seq):\n",
    "    cps = 0\n",
    "    if (len(seq) == 6):\n",
    "        return(cp_scores[seq])  \n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    return(cps)\n",
    "\n",
    "# codon adaptation index of any length sequence\n",
    "def get_cai(seq):\n",
    "    # get the CAI as the geometric mean of all the codon's individual CAI values\n",
    "    cai_list = []\n",
    "    for i in range(0,len(seq),3):\n",
    "        cai_list.append(codon_CAI[seq[i:(i+3)]])\n",
    "    return(scistats.gmean(cai_list))\n",
    "\n",
    "# read in synonymous codons\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "\n",
    "# read in the yeast orfs and make the CAI for all the codons\n",
    "codon_occs = dict()\n",
    "# set up the dict\n",
    "for codon in (codon_to_aa_dict.keys()):\n",
    "    codon_occs[codon] = 0\n",
    "genes_file = open(\"s_cerevisae_orf_coding_no_Mito_no_Plasmid.fasta\")\n",
    "for line in genes_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\": \n",
    "        if (len(line)%3 != 0):\n",
    "            print(\"PROBLEM, genes file is not multiple of 3 per line.\")\n",
    "        # get the frequency of each codon\n",
    "        for i in range(0,len(line),3):\n",
    "            codon_occs[line[i:(i+3)]] += 1\n",
    "# get the CAI of each codon\n",
    "codon_CAI = dict()\n",
    "for aa in (aa_to_syn_codons_dict.keys()):\n",
    "    highest_frequency = 0\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (codon_occs[codon] > highest_frequency):\n",
    "            highest_frequency = codon_occs[codon]\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (highest_frequency == 0):\n",
    "            codon_CAI[codon] = 0\n",
    "        else:\n",
    "            codon_CAI[codon] = codon_occs[codon]/highest_frequency\n",
    "        \n",
    "# read in the codon pair scores\n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"yeast_cps.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[0]] = float(line[1])\n",
    "\n",
    "# read in the gene's codon sequence\n",
    "codons_file = open(\"S288C_YBR115C_LYS2_genomic.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "# read in the gene's amino acid sequence\n",
    "aa_file = open(\"S288C_YBR115C_LYS2_protein.fsa\")\n",
    "original_AA = \"\"\n",
    "for l in aa_file:\n",
    "    l = l.rstrip()\n",
    "    if l[0] != \">\":\n",
    "        original_AA = original_AA + l\n",
    "aa_file.close()\n",
    "\n",
    "if (len(original_AA)*3 != len(original_codons)):\n",
    "    print(\"PROBLEM: codons and AA don't match\")\n",
    "\n",
    "    \n",
    "# store the sequences, and then work with the ones that have \"chopped off\" start and end\n",
    "long_original_AA = original_AA\n",
    "original_AA = original_AA[40:-20]\n",
    "long_original_codons = original_codons\n",
    "original_codons = original_codons[120:-60]\n",
    "   \n",
    "# store_best will contain, at each AA position, and at each of the syn codon choices,\n",
    "# the best cost of the suffix that starts with this codon, the current codon pair score,\n",
    "# and the next codon that corresponds to this cost\n",
    "store_best = [None] * len(original_AA)\n",
    "for i in range(0,len(original_AA)):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = 0\n",
    "                \n",
    "\n",
    "# initialize the dynamic program by setting store_best for the last amino acid position\n",
    "for codon in store_best[-1]:\n",
    "    store_best[i][codon] = (0,0,\"\",\"\")\n",
    "\n",
    "\n",
    "# do the dynamic program, stating from the one before the end of the list, until the beginning\n",
    "for pos in reversed(range(0, len(store_best)-1)):\n",
    "    for codon in store_best[pos]:\n",
    "        # consider as cost the cost of adding this codon to one of the best sequences\n",
    "        # from the previous suffix\n",
    "        best_prev = \"\"\n",
    "        best_cps = sys.maxsize # keep this\n",
    "        best_cost = -1 #sys.maxsize #-1\n",
    "        best_seq = \"\"\n",
    "        for prev_codon in store_best[pos+1]:\n",
    "            cost = num_costly_codon_pairs(codon+prev_codon,1) + store_best[pos+1][prev_codon][0]\n",
    "            cp_score = get_cps(codon+prev_codon) + store_best[pos+1][prev_codon][1]\n",
    "            #print(cp_score)\n",
    "            # maximizer the cost\n",
    "            #if (cp_score < best_cps):\n",
    "            if ((cost > best_cost) or # > <\n",
    "                # break the tie with lowest codon pair score\n",
    "                ((cost == best_cost) and (cp_score < best_cps))):\n",
    "                best_cost = cost\n",
    "                best_cps = cp_score\n",
    "                best_prev = prev_codon\n",
    "                best_seq = prev_codon + store_best[pos+1][prev_codon][3]\n",
    "        store_best[pos][codon] = (best_cost, best_cps, best_prev, best_seq)\n",
    "        #print(\"storing for \" + str(pos) + \" \" + codon + \": \" + str(best_cps) + \" \" + best_prev)\n",
    "\n",
    "# going from the beginning to the end, read off the recoded sequences from the store_best\n",
    "# re-doing the computation for the first position, to pick the best of the syn codons\n",
    "# the rest of the seuqnces is just read out\n",
    "recoded_sequence = \"\"\n",
    "next_codon = \"\"\n",
    "for pos in range(0, len(store_best)):\n",
    "    if pos == 0:\n",
    "        # maximum\n",
    "        best_first = \"\"\n",
    "        best_first_cps = sys.maxsize # keep this\n",
    "        best_first_cost = -1 # sys.maxsize #-1\n",
    "        for codon in store_best[0]:\n",
    "            # maximize the cost\n",
    "            #if (store_best[pos][codon][1] < best_first_cps):\n",
    "            if ((store_best[pos][codon][0] > best_first_cost) or  #> <\n",
    "               ((store_best[pos][codon][0] == best_first_cost) and (store_best[pos][codon][1] < best_first_cps))):\n",
    "                best_first = codon\n",
    "                best_first_cost = store_best[pos][codon][0]\n",
    "                best_first_cps = store_best[pos][codon][1]\n",
    "                next_codon = store_best[pos][codon][2]\n",
    "        recoded_sequence += best_first\n",
    "    else:\n",
    "        recoded_sequence += next_codon\n",
    "        next_codon = store_best[pos][next_codon][2]\n",
    "\n",
    "# now restore the original sequences and patch up the recoded sequence with the beginning and end\n",
    "original_AA = long_original_AA\n",
    "original_codons = long_original_codons\n",
    "recoded_sequence = original_codons[0:120] + recoded_sequence + original_codons[-60:]\n",
    "\n",
    "print(original_codons[0:120] + best_first + store_best[0][best_first][3] + original_codons[-60:])\n",
    "\n",
    "print(\"\\n)\n",
    "\n",
    "\n",
    "#print(original_codons)\n",
    "print(recoded_sequence)    \n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "print(len(original_codons))\n",
    "print(\"num desired codon pairs in original sequence: \" + str(num_costly_codon_pairs(original_codons, 0)))\n",
    "print(\"num desired codon pairs in recoded sequence: \" + str(num_costly_codon_pairs(recoded_sequence, 0)))\n",
    "print(\"codon pair score of the original sequence: \" + str(get_cps(original_codons)))\n",
    "print(\"codon pair score of the recoded sequence: \" + str(get_cps(recoded_sequence)))\n",
    "print(\"CAI of the original sequence is: \" + str(get_cai(original_codons)))\n",
    "print(\"CAI of the recoded sequence is: \" + str(get_cai(recoded_sequence)))\n",
    "print(\"num opposite in original: \" + str(get_count(original_codons)))\n",
    "print(\"num opposite in recoded: \" + str(get_count(recoded_sequence)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Version 1\n",
    "vienna package - something like mfold - look into with justin - Bruce not worried about this\n",
    "vienna package verbose\n",
    "\n",
    "splice sites - Justin,\n",
    "https://users.soe.ucsc.edu/~leslie/runIntronSearch.html\n",
    "http://harlequin.jax.org/drupal/?q=yeastRNApred\n",
    "\n",
    "transcriptional terminators - Justin did not do anything;  they are not well defined....\n",
    "\n",
    "restriction sites - Bruce (look at what genscript uses)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "        \n",
    "def get_cps(seq):\n",
    "    cps = 0\n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    return(cps)        \n",
    "\n",
    "def num_costly_codon_pairs(seq, do_print):\n",
    "    num = 0\n",
    "    to_print = \"\"\n",
    "    for i in range(0,(len(seq)-3),3):\n",
    "        if (seq[i+2] == \"C\" and seq[i+3] == \"G\"):\n",
    "        #if (seq[i+2] == \"T\" and seq[i+3] == \"A\" and ((seq[i+4] == \"A\" or seq[i+4] == \"G\"))):\n",
    "            num += 1\n",
    "            if (do_print == 1):\n",
    "                to_print += str(i) + \" \" #seq[i:i+6] + \" \" + \n",
    "    if (do_print):\n",
    "        print(to_print)\n",
    "    return(num)\n",
    "\n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"yeast_cps.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[0]] = float(line[1])\n",
    "    \n",
    "codons_file = open(\"S288C_YBR115C_LYS2_genomic.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "aa_file = open(\"S288C_YBR115C_LYS2_protein.fsa\")\n",
    "original_AA = \"\"\n",
    "for l in aa_file:\n",
    "    l = l.rstrip()\n",
    "    if l[0] != \">\":\n",
    "        original_AA = original_AA + l\n",
    "aa_file.close()\n",
    "\n",
    "if (len(original_AA)*3 != len(original_codons)):\n",
    "    print(\"PROBLEM: codons and AA don't match\")\n",
    "\n",
    "    \n",
    "# store the sequences, and then work with the ones that have \"chopped off\" start and end\n",
    "long_original_AA = original_AA\n",
    "original_AA = original_AA[40:-20]\n",
    "long_original_codons = original_codons\n",
    "original_codons = original_codons[120:-60]\n",
    "    \n",
    "store_best = [None] * len(original_AA)\n",
    "for i in range(0,len(original_AA)):\n",
    "    store_best[i] = dict()\n",
    "    for codon in aa_to_syn_codons_dict[original_AA[i]]:\n",
    "        store_best[i][codon] = 0\n",
    "                \n",
    "\n",
    "# do the last AA\n",
    "for codon in store_best[-1]:\n",
    "    store_best[i][codon] = (0,\"\")\n",
    "\n",
    "\n",
    "# do the dynamic program, stating from the one before the end of the list, until the end\n",
    "for pos in reversed(range(0, len(store_best)-1)):\n",
    "    for codon in store_best[pos]:\n",
    "        # consider as cost the cost of adding this codon to one of the best sequences\n",
    "        # from the previous suffix\n",
    "        # maximizer the dimers\n",
    "        best_prev = \"\"\n",
    "        best_cost = -1 #sys.maxsize #-1\n",
    "        for prev_codon in store_best[pos+1]:\n",
    "            cost = num_costly_codon_pairs(codon+prev_codon,0) + store_best[pos+1][prev_codon][0]\n",
    "            # maximizer the dimers\n",
    "            if ((cost > best_cost) or # > <\n",
    "                ((cost == best_cost) and (prev_codon == original_codons[(pos+1)*3:(pos+1)*3+3]))): #try to keep same seq\n",
    "                best_cost = cost\n",
    "                best_prev = prev_codon\n",
    "        store_best[pos][codon] = (best_cost, best_prev)\n",
    "        #print(\"storing for \" + str(pos) + \" \" + codon + \": \" + str(best_cost) + \" \" + best_prev)\n",
    "\n",
    "recoded_sequence = \"\"\n",
    "next_codon = \"\"\n",
    "for pos in range(0, len(store_best)):\n",
    "    if pos == 0:\n",
    "        # maximum\n",
    "        best_first = \"\"\n",
    "        best_first_cost = -1 # sys.maxsize #-1\n",
    "        for codon in store_best[0]:\n",
    "            # maximum\n",
    "            if ((store_best[pos][codon][0] > best_first_cost) or  #> <\n",
    "                ((cost == best_cost) and (codon == original_codons[(pos+1)*3:(pos+1)*3+3]))): #try to keep same seq\n",
    "                best_first = codon\n",
    "                best_first_cost = store_best[pos][codon][0]\n",
    "                next_codon = store_best[pos][codon][1]\n",
    "        recoded_sequence += best_first\n",
    "    else:\n",
    "        recoded_sequence += next_codon\n",
    "        next_codon = store_best[pos][next_codon][1]\n",
    "\n",
    "# now restore the original sequences and patch up the recoded sequence with the beginning and end\n",
    "original_AA = long_original_AA\n",
    "original_codons = long_original_codons\n",
    "recoded_sequence = original_codons[0:120] + recoded_sequence + original_codons[-60:]\n",
    "\n",
    "#print(original_codons)\n",
    "print(recoded_sequence)    \n",
    "print(\"num costly codon pairs: \" + str(num_costly_codon_pairs(recoded_sequence, 0)))\n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "print(len(original_codons))\n",
    "print(\"num costly codon pairs in original sequences: \" + str(num_costly_codon_pairs(original_codons, 0)))\n",
    "print(\"codon pair score of the original sequence: \" + str(get_cps(original_codons)))\n",
    "print(\"codon pair score of the recoded sequence: \" + str(get_cps(recoded_sequence)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
