{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hi\n"
     ]
    }
   ],
   "source": [
    "print(\"hi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of codon diferences between original and recoded sequences 576\n",
      "AA sequence remains the same, good\n",
      "coleman codon pair bias of the original sequence: -0.03513233601841192\n",
      "coleman codon pair bias of the recoded sequence: -0.036011507479861975\n",
      "frame2&3 indep codon pair score of the original sequence: -31.61352552200005\n",
      "frame2&3 indep codon pair score of the recoded sequence: -488.30513615799987\n",
      "C\n",
      "[195, 357, 510, 567, 699, 789, 945, 948, 1059, 1242, 1353, 1482, 1641, 1962, 2514]\n",
      "[2514, 357, 1242, 1353, 195, 945, 1482, 1641, 510, 948, 1962, 1059, 567, 789, 699]\n",
      "195   4\n",
      "357   1\n",
      "510   8\n",
      "567   12\n",
      "699   14\n",
      "789   13\n",
      "945   5\n",
      "948   9\n",
      "1059   11\n",
      "1242   2\n",
      "1353   3\n",
      "1482   6\n",
      "1641   7\n",
      "1962   10\n",
      "2514   0\n",
      "GAHENSNRAYGSSTINYTTINYYRDSASNAASKQDFSQDPSKFTEPIKDVLIKTAPMLNSPNIEACGYSDRVLQLTLGNSTITTQEAANSVVAYGRWPEYLRDSEANPVDQPTEPDVAACRFYTLDTVSWTKESRGWWWKLPDALRDMGLFGQNMYYHYLGRSGYTVHVQCNASKFHQGALGVFAVPEMCLAGDSNTTTMHTSYQNANPGEKGGTFTGTFTPDNNQTSPARRFCPVDYLLGNGTLLGNAFVFPHQIINLRTNNCATLVLPYVNSLSIDSMVKHNNWGIAILPLAPLNFASESSPEIPITLTIAPMCCEFNGLRNITLPRLQGLPVMNTPGSNQYLTADNFQSPCALPEFDVTPPIDIPGEVKNMMELAEIDTMIPFDLSATKKNTMEMYRVRLSDKPHTDDPILCLSLSPASDPRLSHTMLGEILNYYTHWAGSLKFTFLFCGSMMATGKLLVSYAPPGADPPKKRKEAMLGTHVIWDIGLQSSCTMVVPWISNTTYRQTIDDSFTEGGYISVFYQTRIVVPLSTPREMDILGFVSACNDFSVRLLRDTTHIEQKALAQGLGQMLESMIDNTVRETVGAATSRDALPNTEASGPTHSKEIPALTAVETGATNPLVPSDTVQTRHVVQHRSRSESSIESFFARGACVTIMTVDNPASTTNKDKLFAVWKITYKDTVQLRRKLEFFTYSRFDMELTFVVTANFTETNNGHALNQVYQIMYVPPGAPVPEKWDDYTWQTSSNPSIFYTYGTAPARISVPYVGISNAYSHFYDGFSKVPLKDQSAALGDSLYGAASLNDFGILAVRVVNDHNPTKVTSKIRVYLKPKHIRVWCPRPPRAVAYYGPGVDYKDGTLTPLSTKDLTTY\n",
      "NSGRNNQTGRLRASVYAFNCDNSRSTRDSRASAEKEVNARSRLEVTRRDRLRASVANTGATYVIELNVYARSLENRTLANRGRGVSQTRLRTATREGTRRSKRLRTTVTLSRVGTGRNTVKTGLYNSTLNTLTDVVATCRRNRKKCARRRNSNAHSTTSCQHGTKNQRLLRKTRNALGLYLTAIERDLNNTDVGINYLTKCASAARSRARAEEIREAKIRIKTDRHRGSVIARLRLILSSLRRYTLGTACKRIEAARGTGTSRTNLRKDRIIERLSARLALSKHILEHIRARRRIESKTRTTEDRRTLSTGTRVRSEASSHSNLNASSASKRIRASCTSAQACVRDRVLLSTNKVAIESRVYREARSRIKTRRARLTTYARLGSSRTLTRHRSTQLVEHAEGSYVHQSLNILTYCEYNDSLVYDVRNERVTSETRLVLLVHRSLTVSRLQVAALTEHSSGGRDTRRRNRLLLTKIRLTTSDKNNISKRDGISNRHGCRNYRSRTSRETKDRLARKTAVTYIIHAARDTESCHDKTYSSSCVAYTATRTISNRHAHILTTSRNGTNRYQRANATLYARIEKASRERRSDCDATYSTDKSTSACTDRLSTNTASHHLYSRCNSTVVNNIRITIRRTRTRALSRQRSSRLKVHNSRSKASYKVDHVTLRTKSAVSRRDYETKDCGTYDLIRRYVCYTVRENRERHRKAETQDLARKCKNTHHARRLSAGTNTEQRSSRRNCSQSEIKVVTAESDLGRNSNASGYSRRGLRGSRVTLDHKSLTTRARRHGASRYGTRHEIKGDSITSVSGRLQRSSNRVDSRRRLSRTEKRRANVLSNSLVVGTVCASKVITRYGSSNKTLTIRETVHAKRTDTR\n",
      "coleman codon pair bias of the candidate sequence: -0.13114959723820485\n",
      "frame2&3 indep codon pair score of the candidate sequence: -475.8033532609998\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "#import pandas as pd\n",
    "import sys        \n",
    "from scipy import stats as scistats\n",
    "from random import shuffle\n",
    "\n",
    "def get_cps(seq):\n",
    "    cps = 0\n",
    "    num_pairs = 0\n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        num_pairs += 1\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    return(cps)\n",
    "\n",
    "def get_cpb(seq):\n",
    "    cps = 0\n",
    "    num_pairs = 0\n",
    "    for i in range(0,(len(seq)-6),3):\n",
    "        num_pairs += 1\n",
    "        cps += cp_scores[seq[i:i+6]]\n",
    "    #print(num_pairs)\n",
    "    #print(cps)\n",
    "    return((cps/num_pairs) )\n",
    "\n",
    "# codon pair score of any length sequence\n",
    "def get_cps_indep(seq):\n",
    "    cps_indep = 0 \n",
    "    # for indep_dependent pairs scores - frame 2\n",
    "    for i in range(1,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_indep):\n",
    "            #print(\"Missing \" + seq[i:i+6] + \" from independent scores\")\n",
    "            continue\n",
    "        cps_indep += cp_scores_indep[seq[i:i+6]]\n",
    "    # for indep_dependent pairs scores - frame 3\n",
    "    for i in range(2,(len(seq)-3),3):\n",
    "        if (seq[i:i+6] not in cp_scores_indep):\n",
    "            #print(\"Missing \" + seq[i:i+6] + \" from independent scores\")\n",
    "            continue\n",
    "        cps_indep += cp_scores_indep[seq[i:i+6]]\n",
    "    return(cps_indep)\n",
    "\n",
    "# codon adaptation index of any length sequence\n",
    "def get_cai(seq):\n",
    "    # get the CAI as the geometric mean of all the codon's individual CAI values\n",
    "    cai_list = []\n",
    "    for i in range(0,len(seq),3):\n",
    "        cai_list.append(codon_CAI[seq[i:(i+3)]])\n",
    "    return(scistats.gmean(cai_list))\n",
    "\n",
    "# read in synonymous codons\n",
    "aa_positions = dict()\n",
    "aa_to_syn_codons_dict = dict()\n",
    "codon_to_aa_dict = dict()\n",
    "csvfile = open('AA_CODONS_MAP')\n",
    "reader = csv.reader(csvfile, delimiter='\\t')\n",
    "for row in reader:\n",
    "    mylist = row[3].split(',')\n",
    "    aa_to_syn_codons_dict[row[0]] = mylist\n",
    "    aa_positions[row[0]] = []\n",
    "    for codon in mylist:\n",
    "        codon_to_aa_dict[codon] = row[0]\n",
    "\n",
    "# read in the human orfs and make the CAI for all the codons\n",
    "codon_occs = dict()\n",
    "# set up the dict\n",
    "for codon in (codon_to_aa_dict.keys()):\n",
    "    codon_occs[codon] = 0\n",
    "genes_file = open(\"human_CCDS_nucleotide.current.fna\")\n",
    "for line in genes_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\": \n",
    "        if (len(line)%3 != 0):\n",
    "            #print(\"PROBLEM, genes file is not multiple of 3 per line.\")\n",
    "            continue\n",
    "        # get the frequency of each codon\n",
    "        for i in range(0,len(line),3):\n",
    "            codon_occs[line[i:(i+3)]] += 1\n",
    "\n",
    "# get the CAI of each codon\n",
    "codon_CAI = dict()\n",
    "for aa in (aa_to_syn_codons_dict.keys()):\n",
    "    highest_frequency = 0\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (codon_occs[codon] > highest_frequency):\n",
    "            highest_frequency = codon_occs[codon]\n",
    "    for codon in (aa_to_syn_codons_dict[aa]):\n",
    "        if (highest_frequency == 0):\n",
    "            codon_CAI[codon] = 0\n",
    "        else:\n",
    "            codon_CAI[codon] = codon_occs[codon]/highest_frequency\n",
    "            \n",
    "cp_scores = dict()\n",
    "cp_scores_file = open(\"human_cps_coleman.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter='\\t')\n",
    "for line in cp_reader:\n",
    "    cp_scores[line[1]] = float(line[5])\n",
    "\n",
    "\n",
    "cp_scores_indep = dict()\n",
    "cp_scores_file = open(\"human_FIndep_Scores.csv\")\n",
    "cp_reader = csv.reader(cp_scores_file, delimiter=',')\n",
    "for line in cp_reader:\n",
    "    cp_scores_indep[line[0]] = float(line[1])\n",
    "            \n",
    "            \n",
    "# read in the gene's codon sequence\n",
    "codons_file = open(\"polio_genomics.fsa\")\n",
    "original_codons = \"\"\n",
    "for line in codons_file:\n",
    "    line = line.rstrip()\n",
    "    if line[0] != \">\":\n",
    "        original_codons = original_codons + line\n",
    "codons_file.close()\n",
    "\n",
    "original_AA = \"\"\n",
    "for i in range(0,len(original_codons),3):\n",
    "    original_AA += codon_to_aa_dict[original_codons[i:i+3]]\n",
    "    \n",
    "recoded_sequence = \"GGTGCTCACGAAAACTCGAATCGCGCGTACGGCTCGTCGACGATCAACTACACGACGATCAACTACTATCGCGATTCGGCGTCGAACGCGGCGTCGAAACAAGACTTCTCGCAAGACCCGTCGAAATTTACCGAACCGATTAAAGATGTTTTAATAAAAACGGCGCCGATGTTAAATTCGCCGAACATAGAAGCTTGCGGTTATTCGGATCGCGTATTACAACTTACGCTGGGCAACTCGACGATCACCACGCAAGAAGCGGCGAACTCGGTCGTCGCGTACGGTCGATGGCCAGAATATTTACGCGACAGTGAGGCGAATCCGGTCGATCAACCAACCGAACCGGACGTCGCGGCGTGCCGATTTTATACGCTCGATACGGTCTCCTGGACGAAAGAATCTCGAGGCTGGTGGTGGAAACTACCGGACGCGTTACGCGACATGGGCCTCTTTGGACAAAACATGTACTACCACTACTTGGGTCGATCCGGTTATACCGTACATGTTCAGTGCAACGCGTCGAAATTTCATCAAGGTGCTTTGGGTGTTTTCGCGGTACCGGAGATGTGCCTGGCGGGGGACTCGAATACGACGACGATGCACACGTCGTACCAAAACGCGAACCCAGGCGAAAAGGGGGGCACCTTCACGGGCACCTTCACGCCGGACAACAACCAGACGTCGCCGGCGCGACGATTCTGCCCGGTCGATTATTTACTTGGAAACGGTACGCTGCTGGGGAACGCGTTCGTATTTCCTCATCAAATAATAAATTTACGAACCAACAATTGCGCGACGCTTGTTTTACCGTACGTCAACAGCCTCTCGATCGACTCGATGGTGAAGCACAACAACTGGGGTATCGCGATCCTCCCGCTGGCGCCGTTAAATTTCGCGAGCGAATCGTCGCCGGAGATCCCGATTACGCTCACGATCGCGCCGATGTGTTGCGAATTTAATGGTTTACGAAATATTACGCTGCCGCGATTACAAGGACTACCGGTCATGAATACGCCGGGCTCGAACCAGTACCTCACGGCGGACAACTTCCAGAGCCCTTGCGCGTTACCGGAATTCGACGTTACGCCGCCGATCGATATTCCTGGAGAAGTAAAAAACATGATGGAGCTGGCGGAGATCGATACGATGATTCCTTTCGATCTTTCCGCGACGAAAAAAAATACGATGGAGATGTATCGCGTACGATTATCCGACAAGCCGCATACCGACGATCCGATCCTCTGCCTCTCGCTCTCGCCCGCGAGCGATCCGCGCCTCTCGCATACGATGCTTGGAGAAATATTAAATTATTATACGCACTGGGCGGGCAGCCTCAAGTTCACCTTCCTCTTCTGTGGCTCGATGATGGCGACGGGGAAATTACTTGTCTCGTACGCGCCGCCGGGGGCGGACCCGCCGAAAAAGCGAAAGGAGGCGATGCTTGGAACTCATGTCATCTGGGATATCGGTTTACAAAGCAGCTGCACCATGGTGGTGCCCTGGATCTCGAATACGACGTACCGCCAGACGATCGACGATTCCTTCACGGAGGGGGGCTACATCTCGGTCTTCTACCAAACGCGAATCGTCGTACCGCTCTCGACGCCGCGCGAAATGGACATTCTTGGTTTCGTATCGGCCTGCAATGACTTCTCGGTACGATTATTACGCGACACCACGCATATCGAACAAAAAGCGCTGGCGCAAGGACTTGGACAAATGCTGGAGTCGATGATCGACAACACGGTACGCGAAACGGTTGGTGCCGCGACGTCGCGCGACGCGTTACCGAATACGGAGGCGTCGGGGCCCACGCACTCGAAAGAAATACCGGCGCTCACGGCGGTCGAAACGGGCGCGACGAACCCGCTTGTTCCAAGCGATACCGTACAAACGCGACATGTTGTACAACATCGATCGCGAAGTGAATCGTCGATCGAAAGTTTTTTCGCGAGGGGGGCCTGTGTCACGATCATGACGGTCGACAACCCGGCGTCGACGACGAACAAAGATAAACTTTTCGCGGTCTGGAAGATCACGTACAAAGATACCGTACAATTACGACGAAAATTAGAATTCTTCACGTACTCGCGATTCGATATGGAGCTCACCTTTGTTGTTACCGCGAACTTCACGGAGACGAACAATGGACATGCGCTCAATCAAGTATATCAAATAATGTACGTACCGCCGGGGGCGCCGGTACCGGAGAAGTGGGATGACTACACGTGGCAGACGTCGTCGAACCCGTCGATCTTCTACACGTACGGTACCGCGCCGGCGCGAATATCGGTACCGTACGTGGGCATCTCGAACGCGTACTCGCACTTTTACGACGGCTTCTCGAAAGTTCCTTTAAAAGATCAATCGGCGGCGCTGGGGGACTCGTTATACGGGGCGGCGTCGCTCAACGATTTCGGTATTCTCGCGGTACGAGTAGTAAATGACCACAACCCGACGAAAGTCACGTCGAAAATTCGCGTATATTTAAAACCGAAACATATTCGCGTATGGTGCCCGCGGCCGCCGCGCGCGGTCGCGTATTACGGTCCTGGGGTCGATTATAAAGATGGTACGCTTACGCCGCTCTCGACGAAAGACCTCACGACGTAC\"\n",
    "\n",
    "back_to_original = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (recoded_sequence[i:i+3] != original_codons[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_original += codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "print(\"total number of codon diferences between original and recoded sequences \" + str(diff_total))\n",
    "if (original_AA == back_to_original):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "\n",
    "print(\"coleman codon pair bias of the original sequence: \" + str(get_cpb(original_codons)))\n",
    "print(\"coleman codon pair bias of the recoded sequence: \" + str(get_cpb(recoded_sequence)))\n",
    "print(\"frame2&3 indep codon pair score of the original sequence: \" + str(get_cps_indep(original_codons)))\n",
    "print(\"frame2&3 indep codon pair score of the recoded sequence: \" + str(get_cps_indep(recoded_sequence)))\n",
    "\n",
    "# now, for each amino acid, find all of the positions to scramble\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    aa = codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "    aa_positions[aa].append(i)\n",
    "\n",
    "#print(aa_positions)\n",
    "number_permutations = 1\n",
    "\n",
    "best_so_far_sequence = recoded_sequence\n",
    "best_so_far_indep_cps = get_cps_indep(recoded_sequence)\n",
    "for p in range(0, number_permutations):\n",
    "    aa_permuted_positions = dict()\n",
    "    # permute all the postions\n",
    "    for aa in (aa_to_syn_codons_dict.keys()):\n",
    "        if (aa == \"C\"):\n",
    "            print(aa)\n",
    "        temp_list = list(aa_positions[aa])\n",
    "        if (aa == \"C\"):\n",
    "            print(temp_list)\n",
    "        shuffle(temp_list)\n",
    "        if (aa == \"C\"):\n",
    "            print(temp_list)\n",
    "        aa_permuted_positions[aa] = temp_list\n",
    "    # construct the new string with all the permutations\n",
    "    candidate_sequence = \"\"\n",
    "    for i in range(0,len(recoded_sequence),3):\n",
    "        aa = codon_to_aa_dict[recoded_sequence[i:i+3]]\n",
    "        pos = aa_permuted_positions[aa].index(i)\n",
    "        if (aa == \"C\"):\n",
    "            print(i, \" \", pos)\n",
    "        candidate_sequence += recoded_sequence[pos:pos+3]\n",
    "    candidate_cps = get_cps_indep(candidate_sequence)\n",
    "    if(candidate_cps > best_so_far_indep_cps):\n",
    "        best_so_far_sequence = candidate_sequence\n",
    "        best_so_far_indep_cps = candidate_cps\n",
    "\n",
    "back_to_recoded = \"\"\n",
    "diff_total = 0\n",
    "for i in range(0,len(recoded_sequence),3):\n",
    "    if (best_so_far_sequence[i:i+3] != recoded_sequence[i:i+3]):\n",
    "        diff_total += 1\n",
    "    back_to_recoded += codon_to_aa_dict[best_so_far_sequence[i:i+3]]\n",
    "if (original_AA == back_to_recoded):\n",
    "    print(\"AA sequence remains the same, good\")\n",
    "else:\n",
    "    print(original_AA)\n",
    "    print(back_to_recoded)\n",
    "\n",
    "print(\"coleman codon pair bias of the candidate sequence: \" + str(get_cpb(candidate_sequence)))\n",
    "print(\"frame2&3 indep codon pair score of the candidate sequence: \" + str(get_cps_indep(candidate_sequence)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
