{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AGGAG']\n",
      "AGGAG\n"
     ]
    }
   ],
   "source": [
    "profile_output = \"GCA_000013405.1_ASM1340v1_profile_output\"\n",
    "#\"GCA_000011185.1_ASM1118v1_profile_output\"\n",
    "lookup = { 0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "found_matrix = False \n",
    "found_start = False\n",
    "found_end = False\n",
    "SDs = []\n",
    "SD = \"\"\n",
    "with open(profile_output, \"r+\") as f:\n",
    "    for line in f:\n",
    "        if (line.startswith(\"% Weight matrix composed\")):\n",
    "            found_matrix = True\n",
    "        elif (line.startswith(\"     A\")):\n",
    "            pass\n",
    "        elif (found_matrix):\n",
    "            numbers = line.strip().split(\"\\t\")\n",
    "            if (len(numbers) != 6) :\n",
    "                continue # bug in output\n",
    "            if (not found_start and float(numbers[5]) >= 0.4): #0.35): #0.4\n",
    "                found_start = True\n",
    "                found_end = False\n",
    "                SD = \"\"\n",
    "                #print(\"found start\")\n",
    "            if (found_start and (float(numbers[5]) < 0.4)): #0.35)): #0.4\n",
    "                found_end = True\n",
    "                found_start = False\n",
    "                #print(\"found end\")\n",
    "                SDs.append(SD)\n",
    "            if (found_start and not found_end):\n",
    "                max_num = float(numbers[0])\n",
    "                max_i = 0\n",
    "                for i in range(1,4):\n",
    "                    if (float(numbers[i]) > max_num):\n",
    "                        max_num = float(numbers[i])\n",
    "                        max_i = i\n",
    "                SD = SD + lookup[max_i]\n",
    "maxSD = \"\"\n",
    "for SD in SDs:\n",
    "    if(len(SD) > len(maxSD)):\n",
    "        maxSD = SD\n",
    "print(SDs)\n",
    "print(maxSD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000008885.1_ASM888v1\n",
      "GCA_000012765.1_ASM1276v1\n",
      "GCA_000008365.1_ASM836v1\n",
      "GCA_000007725.1_ASM772v1\n",
      "GCA_000006625.1_ASM662v1\n",
      "GCA_000011225.1_ASM1122v1\n",
      "GCA_000195875.1_ASM19587v1\n",
      "GCA_000012225.1_ASM1222v1\n",
      "GCA_000008305.1_ASM830v1\n",
      "GCA_000007325.1_ASM732v1\n"
     ]
    }
   ],
   "source": [
    "# This is an OLD VERSION, see cell below for new one\n",
    "#### FOR TESTING, remove keep only valid orfs from Tompa search\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import subprocess\n",
    "import csv\n",
    "import scipy\n",
    "from os.path import join\n",
    "\n",
    "# expected perl: v5\n",
    "# expected input files:\n",
    "# / *_genomic.fna.gz\n",
    "# keep_only_valid_orfs.pl, tompa_test installed, \n",
    "\n",
    "def run_Tompa_pipeline(file_prefix, input_path, output_path): \n",
    "\n",
    "    print(file_prefix)\n",
    "    rbs_input = output_path + file_prefix + '_RBSidentify_input'\n",
    "    rbs_output = output_path + file_prefix + '_RBSidentify_output'\n",
    "    profile_input = output_path + file_prefix + '_profile_input'\n",
    "    profile_output = output_path + file_prefix + '_profile_output' \n",
    "    \n",
    "    # pre-process to get rid of bad orfs\n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    \n",
    "    #p = Popen(\"gunzip\" + \" -c \" + cds_file + \" > \" + processed_cds_file, shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "    #ret = p.communicate()[0]\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing keep_only_valid_orfs for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    #TODO later make sequence upstream of genes handle multiple chromosomes!!!!    \n",
    "    full_genome_file = input_path + file_prefix + \"_genomic.fna.gz\"\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_sequence_upstream_of_gene.pl\", processed_cds_file,\\\n",
    "                                      full_genome_file, rbs_input], stderr=subprocess.STDOUT)\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_sequence_upstream_of_gene.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    \n",
    "    # run RBSidentify\n",
    "    p = Popen(['tompa_test/RBSidentify/RBSidentify'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input= (rbs_input + '\\n' + rbs_output + '\\nn\\n7\\n1\\nn\\n').encode())[0]\n",
    "\n",
    "    \n",
    "    # process output of RBSidentify \n",
    "    p = Popen(\"tail -n +27 \" + rbs_output  + \"|  awk '{print $NF,$0}' | sort -nr | cut -f2- -d' ' >\"  + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "\n",
    "    # run profile\n",
    "    p = Popen(['tompa_test/profile/profile'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input=(profile_input + '\\n' + rbs_input + '\\n'  + profile_output + '\\n20\\n7\\n1\\nn\\n').encode())[0]\n",
    "\n",
    "    lookup = { 0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "    found_matrix = False \n",
    "    found_start = False\n",
    "    found_end = False\n",
    "    SD = \"\"\n",
    "    with open(profile_output, \"r+\") as f:\n",
    "        for line in f:\n",
    "            if (line.startswith(\"% Weight matrix composed\")):\n",
    "                found_matrix = True\n",
    "            elif (line.startswith(\"     A\")):\n",
    "                pass\n",
    "            elif (found_matrix):\n",
    "                numbers = line.strip().split()\n",
    "                if (len(numbers) != 6) :\n",
    "                    break # bug in output\n",
    "                if (float(numbers[5]) >= 0.35):\n",
    "                    found_start = True\n",
    "                if (found_start and (float(numbers[5]) < 0.35)):\n",
    "                    found_end = True\n",
    "                if (found_start and not found_end):\n",
    "                    max_num = float(numbers[0])\n",
    "                    max_i = 0\n",
    "                    for i in range(1,4):\n",
    "                        if (float(numbers[i]) > max_num):\n",
    "                            max_num = float(numbers[i])\n",
    "                            max_i = i\n",
    "                    SD = SD + lookup[max_i]\n",
    "                    \n",
    "    # clean up the files\n",
    "    #p = Popen(\"rm \" + processed_cds_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + rbs_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    p = Popen(\"rm \" + rbs_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    #p = Popen(\"rm \" + profile_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    return(SD)\n",
    "\n",
    "\n",
    "\n",
    "# expected file\n",
    "csv_with_file_prefixes = \"DiwanDataFilesInfo.csv\"\n",
    "csv_with_tompa_sd_added = 'DiwanDataFiles_WithTompa_no_filter.csv'\n",
    "input_file_path = \"/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "output_file_path = \"./\"\n",
    "#file_prefix = \"Sulfolobus_acidocaldarius\"\n",
    "\n",
    "count = 0\n",
    "with open(csv_with_file_prefixes, 'r') as in_file:\n",
    "    with open(csv_with_tompa_sd_added, 'w') as out_file:\n",
    "        csvReader = csv.reader(in_file, delimiter=',')\n",
    "        csvWriter = csv.writer(out_file, delimiter=',')\n",
    "        for row in csvReader:\n",
    "            if (row[0] == \"name\"):\n",
    "                row.append(\"tompa_sd\")\n",
    "                csvWriter.writerow(row)\n",
    "                continue\n",
    "            count += 1\n",
    "            if (count > 10):\n",
    "                break\n",
    "            file_prefix = row[1]\n",
    "            #if (file_prefix != \"GCA_000946735.1_Vibrio_cholerae_G_298_Guinea\"):\n",
    "             #   continue\n",
    "            SD = run_Tompa_pipeline(file_prefix, input_file_path, output_file_path)\n",
    "            if ((len(SD) < 4) or (len(SD) > 6)):\n",
    "                SD = \"------\"\n",
    "            row.append(SD)\n",
    "            csvWriter.writerow(row)\n",
    "            \n",
    "            \n",
    "# TODO: provide path to the installed tompa algo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000020605.1_ASM2060v1\n",
      "GCA_000580335.1_ASM58033v1\n",
      "GCA_000581715.1_ASM58171v1\n",
      "GCA_000046845.1_ASM4684v1\n",
      "GCA_000092025.1_ASM9202v1\n",
      "GCA_000009365.1_ASM936v1\n",
      "GCA_000014785.1_ASM1478v1\n",
      "GCA_000204075.1_ASM20407v1\n",
      "GCA_000013385.1_ASM1338v1\n",
      "GCA_000011945.1_ASM1194v1\n",
      "GCA_000013125.1_ASM1312v1\n",
      "GCA_900018365.1_IMG-taxon_2615840617_annotated_assembly\n",
      "GCA_000008625.1_ASM862v1\n",
      "GCA_000747155.1_ASM74715v1\n",
      "GCA_000025965.1_ASM2596v1\n",
      "GCA_000012225.1_ASM1222v1\n",
      "GCA_000008165.1_ASM816v1\n",
      "GCA_000307855.1_BAZO1.0\n",
      "GCA_000307875.1_BABA1.0\n",
      "GCA_000007825.1_ASM782v1\n",
      "GCA_000399225.1_Baci_cere_IS195_V1\n",
      "GCA_000399245.1_Baci_cere_MC118_V1\n",
      "GCA_000290955.1_Baci_cere_VD022_G8177_V1\n",
      "GCA_000009825.1_ASM982v1\n",
      "GCA_000011145.1_ASM1114v1\n",
      "GCA_000011645.1_ASM1164v1\n",
      "GCA_001307105.1_ASM130710v1\n",
      "GCA_000009045.1_ASM904v1\n",
      "GCA_000940785.1_ASM94078v1\n",
      "GCA_001304035.1_ASM130403v1\n",
      "GCA_001552775.1_ASM155277v1\n",
      "GCA_000025985.1_ASM2598v1\n",
      "GCA_000699665.1_ASM69966v1\n",
      "GCA_000011065.1_ASM1106v1\n",
      "GCA_001567255.1_ASM156725v1\n",
      "GCA_001525625.1_ASM152562v1\n",
      "GCA_000046685.1_ASM4668v1\n",
      "GCA_000013185.1_ASM1318v1\n",
      "GCA_000196175.1_ASM19617v1\n",
      "GCA_000301075.1_Berg_zooh_ATCC_43767_V1\n",
      "GCA_000301095.1_Berg_zooh_CCUG_30536_V1\n",
      "GCA_000741215.1_Bifgalrum\n",
      "GCA_000007525.1_ASM752v1\n",
      "GCA_000334405.1_ASM33440v1\n",
      "GCA_000022605.2_ASM2260v1\n",
      "GCA_000236405.1_ASM23640v1\n",
      "GCA_000233435.1_ASM23343v1\n",
      "GCA_000471965.1_ASM47196v1\n",
      "GCA_000348805.1_ASM34880v1\n",
      "GCA_000093165.1_ASM9316v1\n",
      "GCA_000070465.1_ASM7046v1\n",
      "GCA_000195675.1_ASM19567v1\n",
      "GCA_000317935.1_ASM31793v1\n",
      "GCA_000195715.1_ASM19571v1\n",
      "GCA_000165595.2_ASM16559v2\n",
      "GCA_000008685.2_ASM868v2\n",
      "GCA_000300045.1_ASM30004v1\n",
      "GCA_000284375.1_ASM28437v1\n",
      "GCA_000008145.1_ASM814v1\n",
      "GCA_000192725.1_ASM19272v1\n",
      "GCA_000223195.1_ASM22319v1\n",
      "GCA_000007725.1_ASM772v1\n",
      "GCA_000014085.1_ASM1408v1\n",
      "GCA_000974835.2_ASM97483v2\n",
      "GCA_000011705.1_ASM1170v1\n",
      "GCA_000770565.1_ASM77056v1\n",
      "GCA_000011545.1_ASM1154v1\n",
      "GCA_000152285.1_ASM15228v1\n",
      "gzip: /scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000152285.1_ASM15228v1_cds_from_genomic.fna.gz: No such file or directory\n",
      "\n",
      "GCA_000007085.1_ASM708v1\n",
      "GCA_000009085.1_ASM908v1\n",
      "GCA_001655775.1_ASM165577v1\n",
      "GCA_001001895.1_ASM100189v1\n",
      "GCA_000043285.1_ASM4328v1\n",
      "GCA_001274515.1_ASM127451v1\n",
      "GCA_000287235.1_ASM28723v1\n",
      "GCA_000287255.1_ASM28725v1\n",
      "GCA_000441575.1_ASM44157v1\n",
      "GCA_000287275.1_ASM28727v1\n",
      "GCA_000287295.1_ASM28729v1\n",
      "GCA_000287315.1_ASM28731v1\n",
      "GCA_000010365.1_ASM1036v1\n",
      "GCA_001273995.1_ASM127399v1\n",
      "GCA_001645765.1_ASM164576v1\n",
      "GCA_000281235.1_ASM28123v1\n",
      "GCA_000319365.1_ASM31936v1\n",
      "GCA_001447885.1_ASM144788v1\n",
      "GCA_000442605.1_ASM44260v1\n",
      "GCA_000012345.1_ASM1234v1\n",
      "GCA_000999445.1_ASM99944v1\n",
      "GCA_000989585.1_ASM98958v1\n",
      "GCA_001430885.1_ASM143088v1\n",
      "GCA_001004735.1_ASM100473v1\n",
      "GCA_001004775.1_ASM100477v1\n",
      "GCA_000011565.1_ASM1156v1\n",
      "GCA_000800435.1_ASM80043v1\n",
      "GCA_000025785.1_ASM2578v1\n",
      "GCA_000442635.1_ASM44263v1\n",
      "GCA_000331975.1_ASM33197v1\n",
      "GCA_001003505.1_ASM100350v1\n",
      "GCA_000012865.1_ASM1286v1\n",
      "GCA_000689375.1_CHV_1\n",
      "GCA_000006905.1_ASM690v1\n",
      "GCA_000014245.1_ASM1424v1\n",
      "GCA_000006685.1_ASM668v1\n",
      "GCA_000008725.1_ASM872v1\n",
      "GCA_000007605.1_ASM760v1\n",
      "GCA_000007605.1_ASM760v1\n",
      "GCA_000091085.1_ASM9108v1\n",
      "GCA_000024145.1_ASM2414v1\n",
      "GCA_000012585.1_ASM1258v1\n",
      "GCA_000012485.1_ASM1248v1\n",
      "GCA_000006985.1_ASM698v1\n",
      "GCA_000018865.1_ASM1886v1\n",
      "GCA_000007705.1_ASM770v1\n",
      "GCA_000055785.1_ASM5578v1\n",
      "GCA_001045465.1_ASM104546v1\n",
      "GCA_000729985.1_ASM72998v1\n",
      "GCA_001420285.1_ASM142028v1\n",
      "GCA_000737665.1_ASM73766v1\n",
      "GCA_001021975.1_ASM102197v1\n",
      "GCA_000143785.1_ASM14378v1\n",
      "GCA_001507325.1_ASM150732v1\n",
      "GCA_001295265.1_ASM129526v1\n",
      "GCA_000520835.1_ASM52083v1\n",
      "GCA_000812865.1_ASM81286v1\n",
      "GCA_001045435.1_ASM104543v1\n",
      "GCA_001563495.1_ASM156349v1\n",
      "GCA_000737785.1_ASM73778v1\n",
      "GCA_000820645.1_PRJEB1544_assembly_1\n",
      "GCA_000737775.1_ASM73777v1\n",
      "GCA_000737705.1_ASM73770v1\n",
      "GCA_000812875.1_ASM81287v1\n",
      "GCA_001677955.1_ASM167795v1\n",
      "GCA_001045445.1_ASM104544v1\n",
      "GCA_000282115.1_Chryseobacterium.strCF314_v1.0\n",
      "GCA_001297705.1_ASM129770v1\n",
      "GCA_001045455.1_ASM104545v1\n",
      "GCA_001593385.1_ASM159338v1\n",
      "GCA_001187685.1_ASM118768v1\n",
      "GCA_001648155.1_ASM164815v1\n",
      "GCA_001456155.1_ASM145615v1\n",
      "GCA_001507335.1_ASM150733v1\n",
      "GCA_000737715.1_ASM73771v1\n",
      "GCA_001424585.1_Leaf180\n",
      "GCA_001421435.1_Leaf201\n",
      "GCA_001424105.1_Leaf394\n",
      "GCA_001424145.1_Leaf404\n",
      "GCA_001425355.1_Leaf405\n",
      "GCA_001677935.1_ASM167793v1\n",
      "GCA_000738495.1_Chry1.0\n",
      "GCA_000829375.1_ASM82937v1\n",
      "GCA_000813825.1_ASM81382v1\n",
      "GCA_000737765.1_ASM73776v1\n",
      "GCA_000008765.1_ASM876v1\n",
      "GCA_000013845.1_ASM1384v1\n",
      "GCA_000007625.1_ASM762v1\n",
      "GCA_000012325.1_ASM1232v1\n",
      "GCA_000195815.1_ASM19581v1\n",
      "GCA_000011305.1_ASM1130v1\n",
      "GCA_000196335.1_ASM19633v1\n",
      "GCA_000006605.1_ASM660v1\n",
      "GCA_000007765.1_ASM776v1\n",
      "GCA_000196015.1_ASM19601v1\n",
      "GCA_000014145.1_ASM1414v1\n",
      "GCA_000012425.1_ASM1242v1\n",
      "GCA_000011905.1_ASM1190v1\n",
      "GCA_000009025.1_ASM902v1\n",
      "GCA_000196275.1_ASM19627v1\n",
      "GCA_001638825.1_ASM163882v1\n",
      "GCA_000010045.1_ASM1004v1\n",
      "GCA_000025945.1_ASM2594v1\n",
      "GCA_000012665.1_ASM1266v1\n",
      "GCA_000195755.1_ASM19575v1\n",
      "GCA_000012565.1_ASM1256v1\n",
      "GCA_000013145.1_ASM1314v1\n",
      "GCA_000050425.1_ASM5042v1\n",
      "GCA_001521815.1_ASM152181v1\n",
      "GCA_000447375.1_C874_spades\n",
      "GCA_001011675.1_ASM101167v1\n",
      "GCA_000495935.2_ASM49593v2\n",
      "GCA_001521745.1_ASM152174v1\n",
      "GCA_001521735.1_ASM152173v1\n",
      "GCA_001521765.1_ASM152176v1\n",
      "GCA_000773465.1_ASM77346v1\n",
      "GCA_001483145.1_ASM148314v1\n",
      "GCA_000390805.1_Ente_faec_7330257-1_V1\n",
      "GCA_000148105.1_ASM14810v1\n",
      "GCA_000147575.1_ASM14757v1\n",
      "GCA_000148205.1_ASM14820v1\n",
      "GCA_000147295.1_ASM14729v1\n",
      "GCA_000407305.1_Ente_faec_V583_V2\n",
      "GCA_000499925.1_1st_assembly\n",
      "GCA_000735105.1_ASM73510v1\n",
      "GCA_000735135.1_ASM73513v1\n",
      "GCA_001643375.1_ASM164337v1\n",
      "GCA_000714795.1_Ery_DSM_8509_v1\n",
      "GCA_000013005.1_ASM1300v1\n",
      "GCA_000700585.1_ASM70058v1\n",
      "GCA_000010245.1_ASM1024v1\n",
      "GCA_000331105.1_ASM33110v1\n",
      "GCA_000296815.2_ASM29681v2\n",
      "GCA_000269865.1_ASM26986v1\n",
      "GCA_000023725.1_ASM2372v1\n",
      "GCA_000023725.1_ASM2372v1\n",
      "GCA_000194605.1_ASM19460v1\n",
      "GCA_000833535.1_ASM83353v1\n",
      "GCA_000058485.1_ASM5848v1\n",
      "GCA_000013345.1_ASM1334v1\n",
      "GCA_000007325.1_ASM732v1\n",
      "GCA_000300875.1_ASM30087v1\n",
      "GCA_000009785.1_ASM978v1\n",
      "GCA_000012925.1_ASM1292v1\n",
      "GCA_000007985.2_ASM798v2\n",
      "GCA_000011385.1_ASM1138v1\n",
      "GCA_000011685.1_ASM1168v1\n",
      "GCA_001302565.1_ASM130256v1\n",
      "GCA_000014285.1_ASM1428v1\n",
      "GCA_000007945.1_ASM794v1\n",
      "GCA_000169755.1_ASM16975v1\n",
      "GCA_000027305.1_ASM2730v1\n",
      "GCA_000011785.1_ASM1178v1\n",
      "GCA_000012985.1_ASM1298v1\n",
      "GCA_000009305.1_ASM930v1\n",
      "GCA_000007905.1_ASM790v1\n",
      "GCA_000008785.1_ASM878v1\n",
      "GCA_000019165.1_ASM1916v1\n",
      "GCA_000008465.1_ASM846v1\n",
      "GCA_000013565.1_ASM1356v1\n",
      "GCA_000439185.1_UKKV901664\n",
      "GCA_000469705.1_LaCIP7613V1\n",
      "GCA_000011985.1_ASM1198v1\n",
      "GCA_000056065.1_ASM5606v1\n",
      "GCA_000008065.1_ASM806v1\n",
      "GCA_000242275.1_ASM24227v1\n",
      "GCA_000347515.1_Assemblage_de_novo\n",
      "GCA_000203855.3_ASM20385v3\n",
      "GCA_000026065.1_ASM2606v1\n",
      "GCA_000008925.1_ASM892v1\n",
      "GCA_001455265.1_ASM145526v1\n",
      "GCA_000006865.1_ASM686v1\n",
      "GCA_000055945.1_ASM5594v1\n",
      "GCA_000048665.1_ASM4866v1\n",
      "GCA_000007665.1_ASM766v1\n",
      "GCA_000007685.1_ASM768v1\n",
      "GCA_000195795.1_ASM19579v1\n",
      "GCA_000008285.1_ASM828v1\n",
      "GCA_000009985.1_ASM998v1\n",
      "GCA_000007745.1_ASM774v1\n",
      "GCA_000014745.1_ASM1474v1\n",
      "GCA_000165735.1_ASM16573v1\n",
      "GCA_000008305.1_ASM830v1\n",
      "GCA_000709395.1_ASM70939v1\n",
      "GCA_000009625.1_ASM962v1\n",
      "GCA_000013705.1_ASM1370v1\n",
      "GCA_000008325.1_ASM832v1\n",
      "GCA_000412595.1_spc777-v1\n",
      "GCA_000013105.1_ASM1310v1\n",
      "GCA_000007865.1_ASM786v1\n",
      "GCA_000195835.1_ASM19583v1\n",
      "GCA_000195855.1_ASM19585v1\n",
      "GCA_000014165.1_ASM1416v1\n",
      "GCA_000277735.2_ASM27773v2\n",
      "GCA_000389905.1_ASM38990v1\n",
      "GCA_000012765.1_ASM1276v1\n",
      "GCA_000025365.1_ASM2536v1\n",
      "GCA_000027325.1_ASM2732v1\n",
      "GCA_000238995.1_ASM23899v1\n",
      "GCA_000186985.3_ASM18698v3\n",
      "GCA_000200735.1_ASM20073v1\n",
      "GCA_000200735.1_ASM20073v1\n",
      "GCA_000008225.1_ASM822v1\n",
      "GCA_000008365.1_ASM836v1\n",
      "GCA_000143865.1_ASM14386v1\n",
      "GCA_000508245.1_ASM50824v1\n",
      "GCA_000477415.1_ASM47741v1\n",
      "GCA_000011225.1_ASM1122v1\n",
      "GCA_000027345.1_ASM2734v1\n",
      "GCA_000195875.1_ASM19587v1\n",
      "GCA_000203215.1_ASM20321v1\n",
      "GCA_000179035.2_ASM17903v2\n",
      "GCA_000179035.2_ASM17903v2\n",
      "GCA_000008245.1_ASM824v1\n",
      "GCA_000277795.1_ASM27779v1\n",
      "GCA_000012685.1_ASM1268v1\n",
      "GCA_000006845.1_ASM684v1\n",
      "GCA_000009105.1_ASM910v1\n",
      "GCA_000013165.1_ASM1316v1\n",
      "GCA_000013885.1_ASM1388v1\n",
      "GCA_000012725.1_ASM1272v1\n",
      "GCA_000012805.1_ASM1280v1\n",
      "GCA_000009145.1_ASM914v1\n",
      "GCA_000014765.1_ASM1476v1\n",
      "GCA_000196355.1_ASM19635v1\n",
      "GCA_000009805.1_ASM980v1\n",
      "GCA_000009705.1_ASM970v1\n",
      "GCA_000013325.1_ASM1332v1\n",
      "GCA_000011245.1_ASM1124v1\n",
      "GCA_000009845.1_ASM984v1\n",
      "GCA_000265465.1_ASM26546v1\n",
      "GCA_000764515.1_ASM76451v1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000756505.1_ASM75650v1\n",
      "GCA_001001785.1_ASM100178v1\n",
      "GCA_000992445.1_ASM99244v1\n",
      "GCA_000996245.1_ASM99624v1\n",
      "GCA_001002825.1_ASM100282v1\n",
      "GCA_001002925.1_ASM100292v1\n",
      "GCA_000993485.1_ASM99348v1\n",
      "GCA_000006825.1_ASM682v1\n",
      "GCA_000011605.1_ASM1160v1\n",
      "GCA_000012885.1_ASM1288v1\n",
      "GCA_000932055.1_ASM93205v1\n",
      "GCA_000196255.1_ASM19625v1\n",
      "GCA_000196155.1_ASM19615v1\n",
      "GCA_000013865.1_ASM1386v1\n",
      "GCA_000007585.1_ASM758v1\n",
      "GCA_000011465.1_ASM1146v1\n",
      "GCA_000008345.1_ASM834v1\n",
      "GCA_000014225.1_ASM1422v1\n",
      "GCA_000026085.1_ASM2608v1\n",
      "GCA_000414275.1_Wild_Type_454AllContigs\n",
      "gzip: /scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000414275.1_Wild_Type_454AllContigs_cds_from_genomic.fna.gz: No such file or directory\n",
      "\n",
      "GCA_001641345.1_ASM164134v1\n",
      "GCA_000026105.1_ASM2610v1\n",
      "GCA_000012445.1_ASM1244v1\n",
      "GCA_000007565.2_ASM756v2\n",
      "GCA_000012245.1_ASM1224v1\n",
      "GCA_000012305.1_ASM1230v1\n",
      "GCA_000013905.1_ASM1390v1\n",
      "GCA_000203875.1_ASM20387v1\n",
      "GCA_000009125.1_ASM912v1\n",
      "GCA_000092045.1_ASM9204v1\n",
      "GCA_000009265.1_ASM926v1\n",
      "GCA_000269625.2_PB_Rhod_Spha_2_4_1_V1\n",
      "gzip: /scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000269625.2_PB_Rhod_Spha_2_4_1_V1_cds_from_genomic.fna.gz: No such file or directory\n",
      "\n",
      "GCA_000014565.1_ASM1456v1\n",
      "GCA_000013605.1_ASM1360v1\n",
      "GCA_000196115.1_ASM19611v1\n",
      "GCA_000013685.1_ASM1368v1\n",
      "GCA_000013085.1_ASM1308v1\n",
      "GCA_000012385.1_ASM1238v1\n",
      "GCA_000007025.1_ASM702v1\n",
      "GCA_000012145.1_ASM1214v1\n",
      "GCA_000195735.1_ASM19573v1\n",
      "GCA_000008045.1_ASM804v1\n",
      "GCA_001670865.1_ASM167086v1\n",
      "GCA_000183155.1_ASM18315v1\n",
      "GCA_000252855.1_ASM25285v1\n",
      "GCA_000734055.1_ASM73405v1\n",
      "GCA_000295655.1_ASM29565v1\n",
      "GCA_000331695.1_ASM33169v1\n",
      "GCA_000191565.1_ASM19156v1\n",
      "GCA_001077795.1_ASM107779v1\n",
      "GCA_000014045.1_ASM1404v1\n",
      "GCA_000014185.1_ASM1418v1\n",
      "GCA_000011965.2_ASM1196v2\n",
      "GCA_000014065.1_ASM1406v1\n",
      "GCA_000210035.1_ASM21003v1\n",
      "GCA_000013665.1_ASM1366v1\n",
      "GCA_000505125.1_CFSAN001338_01.0\n",
      "GCA_000011885.1_ASM1188v1\n",
      "GCA_000486185.1_ASM48618v1\n",
      "GCA_000006945.2_ASM694v2\n",
      "GCA_000013765.1_ASM1376v1\n",
      "GCA_000014705.1_ASM1470v1\n",
      "GCA_000146165.2_ASM14616v2\n",
      "GCA_000014685.1_ASM1468v1\n",
      "GCA_000006965.1_ASM696v1\n",
      "GCA_000010085.1_ASM1008v1\n",
      "GCA_000013985.1_ASM1398v1\n",
      "GCA_000009005.1_ASM900v1\n",
      "GCA_000007645.1_ASM764v1\n",
      "GCA_000009865.1_ASM986v1\n",
      "GCA_000010125.1_ASM1012v1\n",
      "GCA_000012705.1_ASM1270v1\n",
      "GCA_000007465.2_ASM746v2\n",
      "GCA_000963275.1_ASM96327v1\n",
      "GCA_000007045.1_ASM704v1\n",
      "GCA_000006785.2_ASM678v2\n",
      "GCA_000011825.1_ASM1182v1\n",
      "GCA_000009765.2_ASM976v2\n",
      "GCA_000203835.1_ASM20383v1\n",
      "GCA_000012965.1_ASM1296v1\n",
      "GCA_000009905.1_ASM990v1\n",
      "GCA_000010065.1_ASM1006v1\n",
      "GCA_000012505.1_ASM1250v1\n",
      "GCA_000009725.1_ASM972v1\n",
      "GCA_000014725.1_ASM1472v1\n",
      "GCA_000013405.1_ASM1340v1\n",
      "GCA_000012405.1_ASM1240v1\n",
      "GCA_000011345.1_ASM1134v1\n",
      "GCA_000008545.1_ASM854v1\n",
      "GCA_000018945.1_ASM1894v1\n",
      "GCA_000187005.1_ASM18700v1\n",
      "GCA_000008125.1_ASM812v1\n",
      "GCA_000012745.1_ASM1274v1\n",
      "GCA_000012605.1_ASM1260v1\n",
      "GCA_000008185.1_ASM818v1\n",
      "GCA_000008605.1_ASM860v1\n",
      "GCA_000014265.1_ASM1426v1\n",
      "GCA_000196075.1_ASM19607v1\n",
      "GCA_000006625.1_ASM662v1\n",
      "GCA_000946735.1_Vibrio_cholerae_G_298_Guinea\n",
      "gzip: /scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000946735.1_Vibrio_cholerae_G_298_Guinea_cds_from_genomic.fna.gz: No such file or directory\n",
      "\n",
      "GCA_000011805.1_ASM1180v1\n",
      "GCA_000196095.1_ASM19609v1\n",
      "GCA_000039765.1_ASM3976v1\n",
      "GCA_000189415.1_ASM18941v1\n",
      "GCA_000189415.1_ASM18941v1\n",
      "GCA_000008885.1_ASM888v1\n",
      "GCA_000008385.1_ASM838v1\n",
      "GCA_000196135.1_ASM19613v1\n",
      "GCA_000007165.1_ASM716v1\n",
      "GCA_000007145.1_ASM714v1\n",
      "GCA_000010025.1_ASM1002v1\n",
      "GCA_000007245.1_ASM724v1\n",
      "GCA_000009065.1_ASM906v1\n",
      "GCA_000834295.1_ASM83429v1\n",
      "GCA_000007105.1_ASM710v1\n"
     ]
    }
   ],
   "source": [
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import subprocess\n",
    "import csv\n",
    "import scipy\n",
    "from os.path import join\n",
    "import glob\n",
    "\n",
    "# expected perl: v5\n",
    "# expected input files:\n",
    "# / *_genomic.fna.gz\n",
    "# keep_only_valid_orfs.pl, tompa_test installed, \n",
    "\n",
    "def run_Tompa_pipeline(file_prefix, input_path, output_path): \n",
    "\n",
    "    print(file_prefix)\n",
    "    rbs_input = output_path + file_prefix + '_RBSidentify_input'\n",
    "    rbs_output = output_path + file_prefix + '_RBSidentify_output'\n",
    "    profile_input = output_path + file_prefix + '_profile_input'\n",
    "    profile_output = output_path + file_prefix + '_profile_output' \n",
    "    \n",
    "    # pre-process to get rid of bad orfs\n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing keep_only_valid_orfs for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    #TODO later make sequence upstream of genes handle multiple chromosomes!!!!    \n",
    "    full_genome_file = input_path + file_prefix + \"_genomic.fna.gz\"\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_sequence_upstream_of_gene.pl\", processed_cds_file,\\\n",
    "                                      full_genome_file, rbs_input], stderr=subprocess.STDOUT)\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_sequence_upstream_of_gene.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    \n",
    "    # run RBSidentify\n",
    "    p = Popen(['tompa_test/RBSidentify/RBSidentify'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input= (rbs_input + '\\n' + rbs_output + '\\nn\\n7\\n1\\nn\\n').encode())[0]\n",
    "    \n",
    "    # check that the output file was created\n",
    "    names= glob.glob(rbs_output)\n",
    "    if (len(names) == 0):\n",
    "        return(\"\") # problem with input\n",
    "     #   print(\"Error, more than 1 file with prefix:\", first)\n",
    "    \n",
    "    # process output of RBSidentify \n",
    "    p = Popen(\"tail -n +27 \" + rbs_output  + \"|  awk '{print $NF,$0}' | sort -nr | cut -f2- -d' ' >\"  + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "\n",
    "    # run profile\n",
    "    p = Popen(['tompa_test/profile/profile'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input=(profile_input + '\\n' + rbs_input + '\\n'  + profile_output + '\\n20\\n7\\n1\\nn\\n').encode())[0]\n",
    "\n",
    "    lookup = { 0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "    found_matrix = False \n",
    "    found_start = False\n",
    "    found_end = False\n",
    "    \n",
    "                    \n",
    "    SDs = []\n",
    "    SD = \"\"\n",
    "    with open(profile_output, \"r+\") as f:\n",
    "        last_threshold = 0\n",
    "        for line in f:\n",
    "            if (line.startswith(\"% Weight matrix composed\")):\n",
    "                found_matrix = True\n",
    "            elif (line.startswith(\"     A\")):\n",
    "                pass\n",
    "            elif (found_matrix):\n",
    "                numbers = line.strip().split(\"\\t\")\n",
    "                if (len(numbers) != 6) :\n",
    "                    break # continue # bug in output\n",
    "                if (not found_start and float(numbers[5]) >= 0.4): \n",
    "                    found_start = True\n",
    "                    found_end = False\n",
    "                    SD = \"\"\n",
    "                    #print(\"found start\")\n",
    "                if (found_start and (float(numbers[5]) < 0.4) and (last_threshold < 0.4)):\n",
    "                    found_end = True\n",
    "                    found_start = False\n",
    "                    #print(\"found end\")\n",
    "                    SDs.append(SD[:-1])\n",
    "                if (found_start and not found_end):\n",
    "                    max_num = float(numbers[0])\n",
    "                    max_i = 0\n",
    "                    for i in range(1,4):\n",
    "                        if (float(numbers[i]) > max_num):\n",
    "                            max_num = float(numbers[i])\n",
    "                            max_i = i\n",
    "                    SD = SD + lookup[max_i]\n",
    "                last_threshold = float(numbers[5])\n",
    "    maxSD = \"\"\n",
    "    for SD in SDs:\n",
    "        if(len(SD) > len(maxSD)):\n",
    "            maxSD = SD              \n",
    "                    \n",
    "    # clean up the files\n",
    "    p = Popen(\"rm \" + processed_cds_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + rbs_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + rbs_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    #p = Popen(\"rm \" + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + profile_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    return(maxSD)\n",
    "\n",
    "\n",
    "\n",
    "# expected file\n",
    "#csv_with_file_prefixes = \"\"\n",
    "#csv_with_tompa_sd_added = ''\n",
    "csv_with_file_prefixes = \"all_just_tompa_Bacteria_DataFilesInfo.csv\"\n",
    "csv_with_tompa_sd_added = 'all_just_tompa_Bacteria_DataFilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"166_bad_tompa_Bacteria_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = '166_bad_tompa_Bacteria_DataFilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Our_NOT_IN_DIWAN_SD_Bacteria_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Our_NOT_IN_DIWAN_SD_Bacteria_DataFilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Diwan_MissingArchaea_FilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Diwan_MissingArchaea_FilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Diwan_MissingBacteria_FilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Diwan_MissingBacteria_FilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Ruhul_Annotate_153_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Ruhul_Annotate_153_DataFiles_WithTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"DiwanDataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'DiwanDataFiles_WithTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Our_SD_Bacteria_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Our_SD_Bacteria_DataFilesInfo_with_Tompa.csv'\n",
    "#input_file_path = \"/scratch4/moamin/ncbi_archaea_db/GbArchaea_FNA/\"\n",
    "input_file_path = \"/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "output_file_path = \"./\"\n",
    "#file_prefix = \"Sulfolobus_acidocaldarius\"\n",
    "\n",
    "count = 0\n",
    "with open(csv_with_file_prefixes, 'r') as in_file:\n",
    "    with open(csv_with_tompa_sd_added, 'w') as out_file:\n",
    "        csvReader = csv.reader(in_file, delimiter=',')\n",
    "        csvWriter = csv.writer(out_file, delimiter=',')\n",
    "        for row in csvReader:\n",
    "            if (row[0] == \"name\"):\n",
    "                #row.append(\"tompa_sd\")\n",
    "                csvWriter.writerow(row)\n",
    "                continue\n",
    "            count += 1\n",
    "            #if (count > 2):\n",
    "             #   break\n",
    "            file_prefix = row[1]\n",
    "            #if (file_prefix != \"GCA_000282115.1_Chryseobacterium.strCF314_v1.0\"):\n",
    "             #   continue\n",
    "            SD = run_Tompa_pipeline(file_prefix, input_file_path, output_file_path)\n",
    "            if (len(SD) < 4): \n",
    "                SD = \"------\"\n",
    "            if (len(SD) > 6):\n",
    "                SD = SD[:6]\n",
    "            #print(row[0])\n",
    "            #print(SD)\n",
    "            #print(row[3])\n",
    "            #row.append(SD)\n",
    "            if (SD == row[3]):\n",
    "                csvWriter.writerow(row)\n",
    "            else:\n",
    "                new_row = row[:3]\n",
    "                new_row.append(SD)\n",
    "                csvWriter.writerow(new_row)\n",
    "            \n",
    "# TODO: provide path to the installed tompa algo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hi\n"
     ]
    }
   ],
   "source": [
    "print(\"hi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
