{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AGGAG']\n",
      "AGGAG\n"
     ]
    }
   ],
   "source": [
    "profile_output = \"GCA_000013405.1_ASM1340v1_profile_output\"\n",
    "#\"GCA_000011185.1_ASM1118v1_profile_output\"\n",
    "lookup = { 0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "found_matrix = False \n",
    "found_start = False\n",
    "found_end = False\n",
    "SDs = []\n",
    "SD = \"\"\n",
    "with open(profile_output, \"r+\") as f:\n",
    "    for line in f:\n",
    "        if (line.startswith(\"% Weight matrix composed\")):\n",
    "            found_matrix = True\n",
    "        elif (line.startswith(\"     A\")):\n",
    "            pass\n",
    "        elif (found_matrix):\n",
    "            numbers = line.strip().split(\"\\t\")\n",
    "            if (len(numbers) != 6) :\n",
    "                continue # bug in output\n",
    "            if (not found_start and float(numbers[5]) >= 0.4): #0.35): #0.4\n",
    "                found_start = True\n",
    "                found_end = False\n",
    "                SD = \"\"\n",
    "                #print(\"found start\")\n",
    "            if (found_start and (float(numbers[5]) < 0.4)): #0.35)): #0.4\n",
    "                found_end = True\n",
    "                found_start = False\n",
    "                #print(\"found end\")\n",
    "                SDs.append(SD)\n",
    "            if (found_start and not found_end):\n",
    "                max_num = float(numbers[0])\n",
    "                max_i = 0\n",
    "                for i in range(1,4):\n",
    "                    if (float(numbers[i]) > max_num):\n",
    "                        max_num = float(numbers[i])\n",
    "                        max_i = i\n",
    "                SD = SD + lookup[max_i]\n",
    "maxSD = \"\"\n",
    "for SD in SDs:\n",
    "    if(len(SD) > len(maxSD)):\n",
    "        maxSD = SD\n",
    "print(SDs)\n",
    "print(maxSD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000008885.1_ASM888v1\n",
      "GCA_000012765.1_ASM1276v1\n",
      "GCA_000008365.1_ASM836v1\n",
      "GCA_000007725.1_ASM772v1\n",
      "GCA_000006625.1_ASM662v1\n",
      "GCA_000011225.1_ASM1122v1\n",
      "GCA_000195875.1_ASM19587v1\n",
      "GCA_000012225.1_ASM1222v1\n",
      "GCA_000008305.1_ASM830v1\n",
      "GCA_000007325.1_ASM732v1\n"
     ]
    }
   ],
   "source": [
    "# This is an OLD VERSION, see cell below for new one\n",
    "#### FOR TESTING, remove keep only valid orfs from Tompa search\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import subprocess\n",
    "import csv\n",
    "import scipy\n",
    "from os.path import join\n",
    "\n",
    "# expected perl: v5\n",
    "# expected input files:\n",
    "# / *_genomic.fna.gz\n",
    "# keep_only_valid_orfs.pl, tompa_test installed, \n",
    "\n",
    "def run_Tompa_pipeline(file_prefix, input_path, output_path): \n",
    "\n",
    "    print(file_prefix)\n",
    "    rbs_input = output_path + file_prefix + '_RBSidentify_input'\n",
    "    rbs_output = output_path + file_prefix + '_RBSidentify_output'\n",
    "    profile_input = output_path + file_prefix + '_profile_input'\n",
    "    profile_output = output_path + file_prefix + '_profile_output' \n",
    "    \n",
    "    # pre-process to get rid of bad orfs\n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    \n",
    "    #p = Popen(\"gunzip\" + \" -c \" + cds_file + \" > \" + processed_cds_file, shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "    #ret = p.communicate()[0]\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing keep_only_valid_orfs for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    #TODO later make sequence upstream of genes handle multiple chromosomes!!!!    \n",
    "    full_genome_file = input_path + file_prefix + \"_genomic.fna.gz\"\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_sequence_upstream_of_gene.pl\", processed_cds_file,\\\n",
    "                                      full_genome_file, rbs_input], stderr=subprocess.STDOUT)\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_sequence_upstream_of_gene.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    \n",
    "    # run RBSidentify\n",
    "    p = Popen(['tompa_test/RBSidentify/RBSidentify'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input= (rbs_input + '\\n' + rbs_output + '\\nn\\n7\\n1\\nn\\n').encode())[0]\n",
    "\n",
    "    \n",
    "    # process output of RBSidentify \n",
    "    p = Popen(\"tail -n +27 \" + rbs_output  + \"|  awk '{print $NF,$0}' | sort -nr | cut -f2- -d' ' >\"  + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "\n",
    "    # run profile\n",
    "    p = Popen(['tompa_test/profile/profile'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input=(profile_input + '\\n' + rbs_input + '\\n'  + profile_output + '\\n20\\n7\\n1\\nn\\n').encode())[0]\n",
    "\n",
    "    lookup = { 0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "    found_matrix = False \n",
    "    found_start = False\n",
    "    found_end = False\n",
    "    SD = \"\"\n",
    "    with open(profile_output, \"r+\") as f:\n",
    "        for line in f:\n",
    "            if (line.startswith(\"% Weight matrix composed\")):\n",
    "                found_matrix = True\n",
    "            elif (line.startswith(\"     A\")):\n",
    "                pass\n",
    "            elif (found_matrix):\n",
    "                numbers = line.strip().split()\n",
    "                if (len(numbers) != 6) :\n",
    "                    break # bug in output\n",
    "                if (float(numbers[5]) >= 0.35):\n",
    "                    found_start = True\n",
    "                if (found_start and (float(numbers[5]) < 0.35)):\n",
    "                    found_end = True\n",
    "                if (found_start and not found_end):\n",
    "                    max_num = float(numbers[0])\n",
    "                    max_i = 0\n",
    "                    for i in range(1,4):\n",
    "                        if (float(numbers[i]) > max_num):\n",
    "                            max_num = float(numbers[i])\n",
    "                            max_i = i\n",
    "                    SD = SD + lookup[max_i]\n",
    "                    \n",
    "    # clean up the files\n",
    "    #p = Popen(\"rm \" + processed_cds_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + rbs_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    p = Popen(\"rm \" + rbs_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    #p = Popen(\"rm \" + profile_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    return(SD)\n",
    "\n",
    "\n",
    "\n",
    "# expected file\n",
    "csv_with_file_prefixes = \"DiwanDataFilesInfo.csv\"\n",
    "csv_with_tompa_sd_added = 'DiwanDataFiles_WithTompa_no_filter.csv'\n",
    "input_file_path = \"/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "output_file_path = \"./\"\n",
    "#file_prefix = \"Sulfolobus_acidocaldarius\"\n",
    "\n",
    "count = 0\n",
    "with open(csv_with_file_prefixes, 'r') as in_file:\n",
    "    with open(csv_with_tompa_sd_added, 'w') as out_file:\n",
    "        csvReader = csv.reader(in_file, delimiter=',')\n",
    "        csvWriter = csv.writer(out_file, delimiter=',')\n",
    "        for row in csvReader:\n",
    "            if (row[0] == \"name\"):\n",
    "                row.append(\"tompa_sd\")\n",
    "                csvWriter.writerow(row)\n",
    "                continue\n",
    "            count += 1\n",
    "            if (count > 10):\n",
    "                break\n",
    "            file_prefix = row[1]\n",
    "            #if (file_prefix != \"GCA_000946735.1_Vibrio_cholerae_G_298_Guinea\"):\n",
    "             #   continue\n",
    "            SD = run_Tompa_pipeline(file_prefix, input_file_path, output_file_path)\n",
    "            if ((len(SD) < 4) or (len(SD) > 6)):\n",
    "                SD = \"------\"\n",
    "            row.append(SD)\n",
    "            csvWriter.writerow(row)\n",
    "            \n",
    "            \n",
    "# TODO: provide path to the installed tompa algo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_001273995.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_001273995.1_ASM127399v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_001273995.1_ASM127399v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "ACACCGAT\n",
      "GCA_000735325.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000735325.1_ASM73532v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000735325.1_ASM73532v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AAGGA\n",
      "GCA_001438645.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_001438645.1_ASM143864v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_001438645.1_ASM143864v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "TGAGGA\n",
      "GCA_000268045.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000268045.1_ASM26804v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000268045.1_ASM26804v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_000171155.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000171155.1_ASM17115v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000171155.1_ASM17115v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "GGAGG\n",
      "GCA_000709915.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000709915.1_JSD_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000709915.1_JSD_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_000769515.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000769515.1_ASM76951v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000769515.1_ASM76951v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAGG\n",
      "GCA_000300175.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000300175.1_ASM30017v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000300175.1_ASM30017v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGA\n",
      "GCA_000439515.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439515.1_ASM43951v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439515.1_ASM43951v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_000020585.3\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000020585.3_ASM2058v3_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000020585.3_ASM2058v3_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_000439555.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439555.1_ASM43955v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439555.1_ASM43955v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_000299995.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000299995.1_ASM29999v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000299995.1_ASM29999v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGA\n",
      "GCA_000439595.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439595.1_ASM43959v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439595.1_ASM43959v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_000439535.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439535.1_ASM43953v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000439535.1_ASM43953v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_001562295.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_001562295.1_ASM156229v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_001562295.1_ASM156229v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AGGAG\n",
      "GCA_000268005.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000268005.1_ASM26800v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000268005.1_ASM26800v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "AAGGAG\n",
      "GCA_000171215.1\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000171215.1_ASM17121v1_cds_from_genomic.fna.gz\n",
      "done with multi\n",
      "/home/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000171215.1_ASM17121v1_genomic.fna.gz\n",
      "done with get sequence upstream\n",
      "done with RBSidentify\n",
      "done with profile\n",
      "GGAGG\n"
     ]
    }
   ],
   "source": [
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import subprocess\n",
    "import csv\n",
    "import scipy\n",
    "from os.path import join\n",
    "import glob\n",
    "import time\n",
    "\n",
    "# expected perl: v5\n",
    "# expected input files:\n",
    "# / *_genomic.fna.gz\n",
    "# keep_only_valid_orfs.pl, tompa_test installed, \n",
    "\n",
    "def run_Tompa_pipeline(file_prefix, input_path, output_path): \n",
    "\n",
    "    print(file_prefix)\n",
    "    rbs_input = output_path + file_prefix + '_RBSidentify_input'\n",
    "    rbs_output = output_path + file_prefix + '_RBSidentify_output'\n",
    "    profile_input = output_path + file_prefix + '_profile_input'\n",
    "    profile_output = output_path + file_prefix + '_profile_output' \n",
    "    \n",
    "    # pre-process to get rid of bad orfs\n",
    "    cds_file = glob.glob(input_path + file_prefix + \"*_cds_from_genomic.fna.gz\")[0]\n",
    "    print(cds_file)\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    try:\n",
    "        #ret = subprocess.check_output([\"perl\", \"keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "        ret = subprocess.check_output([\"perl\", \"multi_keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "        print(\"done with multi\")\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing multi_keep_only_valid_orfs for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    #TODO later make sequence upstream of genes handle multiple chromosomes!!!!    \n",
    "    genome_matches = glob.glob(input_path + file_prefix + \"*_genomic.fna.gz\")\n",
    "    full_genome_file = \"\"\n",
    "    for name in genome_matches:\n",
    "        if (\"from_genomic\" not in name):\n",
    "            full_genome_file = name\n",
    "    \n",
    "    print(full_genome_file)\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"multi_get_sequence_upstream_of_gene.pl\", processed_cds_file,\\\n",
    "                                      full_genome_file, rbs_input], stderr=subprocess.STDOUT)\n",
    "        print(\"done with get sequence upstream\")\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != ''):\n",
    "            print(ret)\n",
    "            return(\"\") # problem with input\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing multi_get_sequence_upstream_of_gene.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    \n",
    "    # run RBSidentify\n",
    "    p = Popen(['tompa_test/RBSidentify/RBSidentify'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input= (rbs_input + '\\n' + rbs_output + '\\nn\\n7\\n1\\nn\\n').encode())[0]\n",
    "    \n",
    "    print(\"done with RBSidentify\")\n",
    "    # check that the output file was created\n",
    "    names= glob.glob(rbs_output)\n",
    "    if (len(names) == 0):\n",
    "        return(\"\") # problem with input\n",
    "     #   print(\"Error, more than 1 file with prefix:\", first)\n",
    "    \n",
    "    # process output of RBSidentify \n",
    "    p = Popen(\"tail -n +27 \" + rbs_output  + \"|  awk '{print $NF,$0}' | sort -nr | cut -f2- -d' ' >\"  + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "\n",
    "    # run profile\n",
    "    p = Popen(['tompa_test/profile/profile'], stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    stdout = p.communicate(input=(profile_input + '\\n' + rbs_input + '\\n'  + profile_output + '\\n20\\n7\\n1\\nn\\n').encode())[0]\n",
    "\n",
    "    print(\"done with profile\")\n",
    "    lookup = { 0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "    found_matrix = False \n",
    "    found_start = False\n",
    "    found_end = False\n",
    "    \n",
    "                    \n",
    "    SDs = []\n",
    "    SD = \"\"\n",
    "    with open(profile_output, \"r+\") as f:\n",
    "        last_threshold = 0\n",
    "        for line in f:\n",
    "            if (line.startswith(\"% Weight matrix composed\")):\n",
    "                found_matrix = True\n",
    "            elif (line.startswith(\"     A\")):\n",
    "                pass\n",
    "            elif (found_matrix):\n",
    "                #print(\"found_matrix\")\n",
    "                numbers = line.strip().split(\"\\t\")\n",
    "                if (len(numbers) != 6) :\n",
    "                    break # continue # bug in output\n",
    "                    #print(\"if_1\")\n",
    "                if (not found_start and float(numbers[5]) >= 0.4): \n",
    "                    found_start = True\n",
    "                    found_end = False\n",
    "                    SD = \"\"\n",
    "                    #print(\"if_2\")\n",
    "                if (found_start and (float(numbers[5]) < 0.4) and (last_threshold < 0.4)):\n",
    "                    found_end = True\n",
    "                    found_start = False\n",
    "                    SDs.append(SD[:-1])\n",
    "                    #print(\"if_3\")\n",
    "                if (found_start and not found_end):\n",
    "                    max_num = float(numbers[0])\n",
    "                    max_i = 0\n",
    "                    for i in range(1,4):\n",
    "                        if (float(numbers[i]) > max_num):\n",
    "                            max_num = float(numbers[i])\n",
    "                            max_i = i\n",
    "                    SD = SD + lookup[max_i]\n",
    "                    #print(\"if_4\")\n",
    "                last_threshold = float(numbers[5])\n",
    "    maxSD = \"\"\n",
    "    for SD in SDs:\n",
    "        if(len(SD) > len(maxSD)):\n",
    "            maxSD = SD              \n",
    "                    \n",
    "    # clean up the files\n",
    "    #p = Popen(\"rm \" + processed_cds_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + rbs_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + rbs_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + profile_input, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + profile_output, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    return(maxSD)\n",
    "\n",
    "\n",
    "\n",
    "# expected file\n",
    "#csv_with_file_prefixes = \"\"\n",
    "#csv_with_tompa_sd_added = ''\n",
    "csv_with_file_prefixes = \"Bruce_July5_orgs.csv\"\n",
    "csv_with_tompa_sd_added = 'Bruce_July5_orgs_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"166_bad_tompa_Bacteria_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = '166_bad_tompa_Bacteria_DataFilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Our_NOT_IN_DIWAN_SD_Bacteria_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Our_NOT_IN_DIWAN_SD_Bacteria_DataFilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Diwan_MissingArchaea_FilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Diwan_MissingArchaea_FilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Diwan_MissingBacteria_FilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Diwan_MissingBacteria_FilesInfo_withTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Ruhul_Annotate_153_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Ruhul_Annotate_153_DataFiles_WithTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"DiwanDataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'DiwanDataFiles_WithTompa_v4.csv'\n",
    "#csv_with_file_prefixes = \"Our_SD_Bacteria_DataFilesInfo.csv\"\n",
    "#csv_with_tompa_sd_added = 'Our_SD_Bacteria_DataFilesInfo_with_Tompa.csv'\n",
    "#input_file_path = \"/scratch4/moamin/ncbi_archaea_db/GbArchaea_FNA/\"\n",
    "#input_file_path = \"/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "input_file_path = \"/home/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "output_file_path = \"./\"\n",
    "#file_prefix = \"Sulfolobus_acidocaldarius\"\n",
    "\n",
    "count = 0\n",
    "with open(csv_with_file_prefixes, 'r') as in_file:\n",
    "    with open(csv_with_tompa_sd_added, 'w') as out_file:\n",
    "        csvReader = csv.reader(in_file, delimiter=',')\n",
    "        csvWriter = csv.writer(out_file, delimiter=',')\n",
    "        for row in csvReader:\n",
    "            if (row[0] == \"name\"):\n",
    "                #row.append(\"tompa_sd\")\n",
    "                csvWriter.writerow(row)\n",
    "                continue\n",
    "            count += 1\n",
    "            #if (count > 2):\n",
    "             #   break\n",
    "            file_prefix = row[1]\n",
    "            #if (file_prefix != \"GCA_000282115.1_Chryseobacterium.strCF314_v1.0\"):\n",
    "             #   continue\n",
    "            SD = run_Tompa_pipeline(file_prefix, input_file_path, output_file_path)\n",
    "            print(SD)\n",
    "            if (len(SD) < 4): \n",
    "                SD = \"------\"\n",
    "            if (len(SD) > 6):\n",
    "                SD = SD[:6]\n",
    "            #print(row[0])\n",
    "            #print(SD)\n",
    "            #print(row[3])\n",
    "            #row.append(SD)\n",
    "            #if (SD == row[3]):\n",
    "             #   csvWriter.writerow(row)\n",
    "            #else:\n",
    "             #   new_row = row[:3]\n",
    "            #  new_row.append(SD)\n",
    "               # csvWriter.writerow(new_row)\n",
    "            new_row = row[:2]\n",
    "            new_row.append(SD)\n",
    "            csvWriter.writerow(new_row)\n",
    "            \n",
    "# TODO: provide path to the installed tompa algo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello\n"
     ]
    }
   ],
   "source": [
    "print(\"hello\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.055\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats\n",
    "obj = scipy.stats.combine_pvalues([0.042, 0.233])\n",
    "print(str(format(float(obj[1]), '.3f')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
