{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "121   GCA_000009725.1_ASM972v1\n",
      "/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000946735.1_Vibrio_cholerae_G_298_Guinea_cds_from_genomic.fna.gz  is missing, not calculating.\n",
      "122   GCA_000946735.1_Vibrio_cholerae_G_298_Guinea\n",
      "123   GCA_000014685.1_ASM1468v1\n",
      "124   GCA_000007585.1_ASM758v1\n",
      "125   GCA_000014765.1_ASM1476v1\n",
      "126   GCA_000196135.1_ASM19613v1\n",
      "127   GCA_000011945.1_ASM1194v1\n",
      "128   GCA_000012805.1_ASM1280v1\n",
      "129   GCA_000196175.1_ASM19617v1\n",
      "130   GCA_000009145.1_ASM914v1\n",
      "131   GCA_000010245.1_ASM1024v1\n",
      "132   GCA_000011605.1_ASM1160v1\n",
      "133   GCA_000013405.1_ASM1340v1\n",
      "134   GCA_000007245.1_ASM724v1\n",
      "135   GCA_000009105.1_ASM910v1\n",
      "136   GCA_000009785.1_ASM978v1\n",
      "137   GCA_000006945.2_ASM694v2\n",
      "138   GCA_000011885.1_ASM1188v1\n",
      "139   GCA_000006845.1_ASM684v1\n",
      "140   GCA_000008605.1_ASM860v1\n",
      "141   GCA_000195815.1_ASM19581v1\n",
      "142   GCA_000196335.1_ASM19633v1\n",
      "143   GCA_000011345.1_ASM1134v1\n",
      "144   GCA_000196355.1_ASM19635v1\n",
      "145   GCA_000012985.1_ASM1298v1\n",
      "146   GCA_000012505.1_ASM1250v1\n",
      "147   GCA_000009365.1_ASM936v1\n",
      "148   GCA_000010085.1_ASM1008v1\n",
      "149   GCA_000012885.1_ASM1288v1\n",
      "150   GCA_000196115.1_ASM19611v1\n",
      "151   GCA_000010065.1_ASM1006v1\n",
      "152   GCA_000013705.1_ASM1370v1\n",
      "153   GCA_000013105.1_ASM1310v1\n",
      "154   GCA_000006985.1_ASM698v1\n",
      "155   GCA_000223195.1_ASM22319v1\n",
      "156   GCA_000008145.1_ASM814v1\n",
      "157   GCA_000012485.1_ASM1248v1\n",
      "158   GCA_000195855.1_ASM19585v1\n",
      "159   GCA_000012665.1_ASM1266v1\n",
      "160   GCA_000014045.1_ASM1404v1\n",
      "161   GCA_000014285.1_ASM1428v1\n",
      "162   GCA_000012425.1_ASM1242v1\n",
      "163   GCA_000012245.1_ASM1224v1\n",
      "164   GCA_000092025.1_ASM9202v1\n",
      "165   GCA_000012925.1_ASM1292v1\n",
      "193   GCA_000055785.1_ASM5578v1\n",
      "194   GCA_000026105.1_ASM2610v1\n",
      "195   GCA_000011965.2_ASM1196v2\n",
      "196   GCA_000203875.1_ASM20387v1\n",
      "197   GCA_000007165.1_ASM716v1\n",
      "198   GCA_000007705.1_ASM770v1\n",
      "199   GCA_000013685.1_ASM1368v1\n",
      "200   GCA_000007145.1_ASM714v1\n",
      "201   GCA_000009985.1_ASM998v1\n",
      "202   GCA_000025965.1_ASM2596v1\n",
      "203   GCA_000013325.1_ASM1332v1\n",
      "204   GCA_000013085.1_ASM1308v1\n",
      "205   GCA_000013985.1_ASM1398v1\n",
      "206   GCA_000195835.1_ASM19583v1\n",
      "207   GCA_000277735.2_ASM27773v2\n",
      "208   GCA_000012745.1_ASM1274v1\n",
      "209   GCA_000196275.1_ASM19627v1\n",
      "/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000414275.1_Wild_Type_454AllContigs_cds_from_genomic.fna.gz  is missing, not calculating.\n",
      "210   GCA_000414275.1_Wild_Type_454AllContigs\n",
      "211   GCA_000014085.1_ASM1408v1\n",
      "212   GCA_001638825.1_ASM163882v1\n",
      "213   GCA_000009125.1_ASM912v1\n",
      "214   GCA_000006905.1_ASM690v1\n",
      "/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000152285.1_ASM15228v1_cds_from_genomic.fna.gz  is missing, not calculating.\n",
      "215   GCA_000152285.1_ASM15228v1\n",
      "216   GCA_000014565.1_ASM1456v1\n",
      "217   GCA_000014785.1_ASM1478v1\n",
      "218   GCA_000012405.1_ASM1240v1\n",
      "219   GCA_000011545.1_ASM1154v1\n",
      "220   GCA_000007665.1_ASM766v1\n",
      "221   GCA_000195715.1_ASM19571v1\n",
      "222   GCA_000317935.1_ASM31793v1\n",
      "223   GCA_000195675.1_ASM19567v1\n",
      "224   GCA_000011705.1_ASM1170v1\n",
      "225   GCA_000014165.1_ASM1416v1\n",
      "226   GCA_000009905.1_ASM990v1\n",
      "227   GCA_000012685.1_ASM1268v1\n",
      "/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/GCA_000269625.2_PB_Rhod_Spha_2_4_1_V1_cds_from_genomic.fna.gz  is missing, not calculating.\n",
      "228   GCA_000269625.2_PB_Rhod_Spha_2_4_1_V1\n",
      "229   GCA_000007865.1_ASM786v1\n",
      "230   GCA_000008125.1_ASM812v1\n",
      "231   GCA_000013345.1_ASM1334v1\n",
      "232   GCA_000014185.1_ASM1418v1\n",
      "233   GCA_000009805.1_ASM980v1\n"
     ]
    }
   ],
   "source": [
    "#TODO: to reproduce paper organisms results, only keep ATG in keep_only_valid_orfs\n",
    "\n",
    "# expected perl: v5\n",
    "# expected input files:\n",
    "# *._cds_from_genomic.fna.gz\n",
    "# keep_only_valid_orfs.pl alisa_efficient_codon_pair_score_hexamers_paper.pl expected_codon_freq.pl\n",
    "# get_atcg_freq.pl paper_get_5or_6_match_to_sd_pairs.pl paper_get_sd_p_value.pl\n",
    "# codon_pair.txt codons.txt syn_codons.txt\n",
    "import subprocess\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import csv\n",
    "import scipy.stats\n",
    "from os.path import join\n",
    "import glob\n",
    "\n",
    "def run_table_generate_pipeline(file_prefix, input_path, output_path, SD): \n",
    "    \n",
    "    # pre-process to get rid of bad orfs\n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    \n",
    "    # return empty if cds file is missing\n",
    "    names= glob.glob(cds_file)\n",
    "    if (len(names) < 1):\n",
    "        print(cds_file, \" is missing, not calculating.\")\n",
    "        return([\"-1\",\"-1\",\"-1\",\"-1\"])\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing keep_only_valid_orfs for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # get the codon pair scores of the cds\n",
    "    codon_pair_scores_file = output_path + file_prefix + \"_pairwise_syn_shuffle_frm123.csv\"\n",
    "    codon_pair_scores_helper_file = output_path + file_prefix + \"_obs_exp_frm123.cts\"\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"alisa_efficient_codon_pair_score_hexamers_paper.pl\", \"-s\", \\\n",
    "                                       processed_cds_file, \"-o\", file_prefix, \"-z\", \"2\"], stderr=subprocess.STDOUT)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # get the codon bias of the cds\n",
    "    codon_bias_file = output_path + file_prefix + \".codon_bias\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"cat \" + processed_cds_file + \" | perl expected_codon_freq.pl >\" + codon_bias_file, \\\n",
    "                              shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing expected_codon_freq.pl for \", cds_file)  \n",
    "    except:\n",
    "        print(\"Problem executing expected_codon_freq.pl for \", cds_file)\n",
    "    \n",
    "    # collect the paper stats\n",
    "    stats = []\n",
    "    \n",
    "    # calculate the depletion score\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_atcg_freq.pl\", \"-i\", processed_cds_file, \"-s\", SD])\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        try:\n",
    "            stats.append(str(format(float(ret), '.3f')))\n",
    "        except:\n",
    "            print(\"Problem: get_atcg_freq.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_atcg_freq.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "      \n",
    "    # calculate the codon pair p-value\n",
    "    codon_pair_scores_only_matches_file = output_path + file_prefix + \".codon_pairs_SD_match\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"perl paper_get_5or_6_match_to_sd_pairs.pl -s \" + SD + \" -c \" + \\\n",
    "                                codon_pair_scores_file + \">\" + codon_pair_scores_only_matches_file, \\\n",
    "                                shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file) \n",
    "    except:\n",
    "        print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file)\n",
    "    # do the 1 -tailed test p-value calculation (previously done by hand in Excel)\n",
    "    # read in the array of all codon pair scores minus the ones with stop codons\n",
    "    all_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        codon1 = row[0][0:3]\n",
    "        codon2 = row[0][3:6]\n",
    "        if ((codon1 in (\"TAG\", \"TGA\", \"TAA\")) or (codon2 in (\"TAG\", \"TGA\", \"TAA\"))):\n",
    "            continue\n",
    "        all_f1_list.append(float(row[1]))\n",
    "    #print(all_f1_list)\n",
    "    #print(len(all_f1_list))\n",
    "    matches_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_only_matches_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        matches_f1_list.append(float(row[1]))\n",
    "    #print(matches_f1_list)\n",
    "    t_test_stats = scipy.stats.ttest_ind(all_f1_list, matches_f1_list, equal_var = False) # 2-tailed test, get 1 tail by / 2\n",
    "    codon_pair_p_value = float(t_test_stats.pvalue/2)\n",
    "    stats.append(str(format(codon_pair_p_value, '.3f')))\n",
    "    \n",
    "    # calculate motif depletion p-value\n",
    "    motif_depletion_p_value = 0.0\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"paper_get_sd_p_value.pl\", \"-s\", \\\n",
    "                                       SD, \"-f\", codon_bias_file], stderr=subprocess.STDOUT)\n",
    "        motif_depletion_p_value = float(ret.decode(\"utf-8\"))\n",
    "        try:\n",
    "            stats.append(str(format(float(motif_depletion_p_value), '.3f')))\n",
    "        except:\n",
    "            print(\"Problem: paper_get_sd_p_value.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # calculate the combined p-value\n",
    "    obj = scipy.stats.combine_pvalues([codon_pair_p_value, motif_depletion_p_value])\n",
    "    stats.append(str(format(float(obj[1]), '.3f')))\n",
    "    \n",
    "    # output the final file\n",
    "    #with open(join(output_file_path, file_prefix + \".table3_values\"), 'w') as fout:\n",
    "     #   fout.write(stats)\n",
    "    \n",
    "    # cleanup: remove created temporary files\n",
    "    #TODO optionally: rm processed_cds_file, codon_pair_scores_file, codon_pair_scores_helper_file codon_bias_file \\\n",
    "    # codon_pair_scores_only_matches_file\n",
    "    # clean up the files\n",
    "    p = Popen(\"rm \" + processed_cds_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_pair_scores_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_pair_scores_helper_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_bias_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_pair_scores_only_matches_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    \n",
    "    return (stats)\n",
    "    \n",
    "\n",
    "# expected file\n",
    "csv_with_tompa_sd_added = 'DiwanDataFiles_WithTompa.csv'\n",
    "csv_with_table_stats_added = 'DiwanDataFiles_WithAllStats.csv'\n",
    "input_file_path = \"/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "output_file_path = \"./\"\n",
    "\n",
    "\n",
    "count = 0\n",
    "with open(csv_with_tompa_sd_added, 'r') as in_file:\n",
    "    with open(csv_with_table_stats_added, 'a') as out_file: # change to 'a' to append when re-starting\n",
    "        csvReader = csv.reader(in_file, delimiter=',')\n",
    "        csvWriter = csv.writer(out_file, delimiter=',')\n",
    "        for row in csvReader:\n",
    "            if (row[0] == \"name\"):\n",
    "                row = row + [\"depletion_score\", \"codon pair p_value\", \"codon usage p_value\", \"combined p_value\"]\n",
    "                csvWriter.writerow(row)\n",
    "                continue\n",
    "            count += 1\n",
    "            #if (count > 2):\n",
    "            #    break\n",
    "            if (count <= 120):\n",
    "                continue\n",
    "            #if (count > 4):\n",
    "             #   break\n",
    "            file_prefix = row[1]\n",
    "            #if (file_prefix != \"GCA_000946735.1_Vibrio_cholerae_G_298_Guinea\"):\n",
    "             #   continue\n",
    "            SD = \"AGGAGG\"\n",
    "            stats = run_table_generate_pipeline(file_prefix, input_file_path, output_file_path, SD)\n",
    "            row = row + stats\n",
    "            print(\",\".join(row))\n",
    "            csvWriter.writerow(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sulfolobus_acidocaldarius,-0.036,0.043,0.427,0.091\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#TODO:  after testing with sulfolobus, uncomment GTG in keep only valid orfs\n",
    "import subprocess\n",
    "import csv\n",
    "import scipy\n",
    "from os.path import join\n",
    "\n",
    "def run_pipeline2(file_prefix, input_path, output_path, SD): \n",
    "    \n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    codon_pair_scores_file = output_path + file_prefix + \"_pairwise_syn_shuffle_frm123.csv\"\n",
    "    codon_pair_scores_helper_file = output_path + file_prefix + \"_obs_exp_frm123.cts\"\n",
    "    \n",
    "    # get the codon bias of the cds\n",
    "    codon_bias_file = output_path + file_prefix + \".codon_bias\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"cat \" + processed_cds_file + \" | perl expected_codon_freq.pl >\" + codon_bias_file, \\\n",
    "                              shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing expected_codon_freq.pl for \", cds_file)  \n",
    "    except:\n",
    "        print(\"Problem executing expected_codon_freq.pl for \", cds_file)\n",
    "                                      \n",
    "    # collect the paper stats\n",
    "    stats = file_prefix + \",\"\n",
    "    \n",
    "    # calculate the depletion score\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_atcg_freq.pl\", \"-i\", processed_cds_file, \"-s\", SD])\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        try:\n",
    "            stats += format(float(ret), '.3f') + \",\"\n",
    "        except:\n",
    "            print(\"Problem: get_atcg_freq.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_atcg_freq.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "      \n",
    "    # calculate the codon pair p-value\n",
    "    codon_pair_scores_only_matches_file = output_path + file_prefix + \".codon_pairs_SD_match\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"perl paper_get_5or_6_match_to_sd_pairs.pl -s \" + SD + \" -c \" + \\\n",
    "                                codon_pair_scores_file + \">\" + codon_pair_scores_only_matches_file, \\\n",
    "                                shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file) \n",
    "    except:\n",
    "        print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file)\n",
    "    # do the 1 -tailed test p-value calculation (previously done by hand in Excel)\n",
    "    # read in the array of all codon pair scores minus the ones with stop codons\n",
    "    all_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        codon1 = row[0][0:3]\n",
    "        codon2 = row[0][3:6]\n",
    "        if ((codon1 in (\"TAG\", \"TGA\", \"TAA\")) or (codon2 in (\"TAG\", \"TGA\", \"TAA\"))):\n",
    "            continue\n",
    "        all_f1_list.append(float(row[1]))\n",
    "    #print(all_f1_list)\n",
    "    #print(len(all_f1_list))\n",
    "    matches_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_only_matches_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        matches_f1_list.append(float(row[1]))\n",
    "    #print(matches_f1_list)\n",
    "    t_test_stats = scipy.stats.ttest_ind(all_f1_list, matches_f1_list, equal_var = False) # 2-tailed test, get 1 tail by / 2\n",
    "    codon_pair_p_value = float(t_test_stats.pvalue/2)\n",
    "    stats += format(codon_pair_p_value, '.3f') + \",\"\n",
    "    \n",
    "    # calculate motif depletion p-value\n",
    "    motif_depletion_p_value = 0.0\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"paper_get_sd_p_value.pl\", \"-s\", \\\n",
    "                                       SD, \"-f\", codon_bias_file], stderr=subprocess.STDOUT)\n",
    "        motif_depletion_p_value = float(ret.decode(\"utf-8\"))\n",
    "        try:\n",
    "            stats += format(float(motif_depletion_p_value), '.3f') + \",\"\n",
    "        except:\n",
    "            print(\"Problem: paper_get_sd_p_value.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # calculate the combined p-value\n",
    "    obj = scipy.stats.combine_pvalues([codon_pair_p_value, motif_depletion_p_value])\n",
    "    stats += format(float(obj[1]), '.3f')\n",
    "    stats += \"\\n\"\n",
    "    \n",
    "    # output the final file\n",
    "    with open(join(output_file_path, file_prefix + \".table3_values\"), 'w') as fout:\n",
    "        fout.write(stats)\n",
    "    \n",
    "    print(stats)\n",
    "    \n",
    "    # cleanup: remove created temporary files\n",
    "    #TODO optionally: rm processed_cds_file, codon_pair_scores_file, codon_pair_scores_helper_file codon_bias_file \\\n",
    "    # codon_pair_scores_only_matches_file\n",
    "    \n",
    "\n",
    "input_file_path = \"./\"\n",
    "output_file_path = \"./\"\n",
    "file_prefix = \"Sulfolobus_acidocaldarius\"\n",
    "SD = \"GGTGA\"\n",
    "run_pipeline2(file_prefix, input_file_path, output_file_path, SD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
