{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "For Ruhul - Annotate\n",
    "         ruhul_re-annotation_all_organism_information.csv -> Ruhul_Annotate_153_DataFilesInfo.csv\n",
    "        Ruhul_Annotate_153_DataFilesInfo.csv -> Ruhul_Annotate_153_DataFiles_WithTompa_v4.csv\n",
    "        \n",
    "        AutomateTable1Calculations_standalone_only_tompa_ruhul_v4_*.py -> merge 3 data files\n",
    "        Ruhul_Annotate_153_DataFiles_WithAllStats_only_tompa_v4_*.csv\n",
    "        Ruhul_Annotate_153_DataFiles_WithAllStats_only_tompa_v4_153.csv rounded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "So far:  \n",
    "    Diwan:  284 organisms;  for now got 236:  \n",
    "            DiwanSupplementOriginal.csv -> DiwanDataFilesInfo.csv (DataPreProcessing)\n",
    "            DiwanDataFilesInfo.csv -> DiwanDataFiles_WithTompa_v4.csv (AutomateTompaSearch)\n",
    "    AGGAGG only:  DiwanDataFiles_WithAll_Stats_236_ONLY_AGGAGG.csv  (AutomateTable1Calculations_standalone.py)\n",
    "    Tompa only: AutomateTable1Calculations_standalone_only_tompa_v4_*.py - merge 4 data files\n",
    "        combine the AGGAGG and the 4 tompa files, fix lines where \"same as AGGAGG\", round up all zeros to 0.001\n",
    "        DiwanDataFiles_WithAllStats_236_AGGAGG_and_Tompa_v4_rounded.csv\n",
    "        post-processing step to get DiwanDataFiles_WithAllStats_236_AGGAGG_and_Tompa_v4_rounded_annotated.csv\n",
    "        \n",
    "        now, get the rest:  Diwan_MissingArchaea_FilesInfo.csv 29 & Diwan_MissingBacteria_FilesInfo.csv 19\n",
    "            Diwan_MissingBacteria_FilesInfo_withTompa_v4.csv Diwan_MissingArchaea_FilesInfo_withTompa_v4.csv\n",
    "            AGGAGG only: \n",
    "                [1]   Running                 nohup python3 -u AutomateTable1Calculations_standalone_diwan_missing19_bacteria.py > diwan_missing19_bacteria_aggagg.log &\n",
    "[2]   Running                 nohup python3 -u AutomateTable1Calculations_standalone_diwan_missing29_archaea.py > diwan_missing20_archaea_aggagg.log &\n",
    "[3]-  Running                 nohup python3 -u AutomateTable1Calculations_standalone_only_tompa_v4_missing19_bacteria.py > diwan_missing19_bacteria_tompa.log &\n",
    "[4]+  Running                 nohup python3 -u AutomateTable1Calculations_standalone_only_tompa_v4_missing29_archaea.py > diwan_missing29_archaea_tompa.log &\n",
    "\n",
    "combine the all missing files into Diwan_Missing_48_allstats.csv, also rounded when necessary\n",
    "post-processing step to get Diwan_Missing_48_allstats_rounded_annotated.csv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGAG\n",
      "CCGGTT\n",
      "GGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "GGAG\n",
      "GAGAA\n",
      "GGTGG\n",
      "GAGGA\n",
      "GGAGG\n",
      "CGTGAT\n",
      "ATGACG\n",
      "GAGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "GGAG\n",
      "AAGGA\n",
      "AGGAG\n",
      "GCAG\n",
      "AGGA\n",
      "AAGGA\n",
      "AGGAG\n",
      "AGGAGA\n",
      "AAGG\n",
      "AGGAG\n",
      "GGAGG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGA\n",
      "AGGAG\n",
      "AGAGG\n",
      "AGGAG\n",
      "AAGGA\n",
      "AAGGA\n",
      "AGGA\n",
      "GAGGAG\n",
      "CTTTG\n",
      "AAGGA\n",
      "AGGAG\n",
      "AGGA\n",
      "CTTTGC\n",
      "GGAGG\n",
      "AGGAG\n",
      "GAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGA\n",
      "AAGGA\n",
      "AGGAG\n",
      "GGAGGT\n",
      "AAGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "GTCTG\n",
      "AAGGA\n",
      "AAGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "CCAA\n",
      "AGGAG\n",
      "AGGAA\n",
      "AGGAG\n",
      "AGGA\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "CAGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGA\n",
      "AGGAG\n",
      "GGTAAT\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGA\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AAGGA\n",
      "AAGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "AGGAG\n",
      "GAGGAG\n",
      "AGGAG\n",
      "Will run 149\n"
     ]
    }
   ],
   "source": [
    "# nohup python3 -u AutomateTable1Calculations_standalone_v2.py > check_progress.log &\n",
    "#TODO: to reproduce paper organisms results, only keep ATG in keep_only_valid_orfs\n",
    "\n",
    "# This is Version 2:  do multiple sequences\n",
    "\n",
    "# expected perl: v5\n",
    "# expected input files:\n",
    "# *._cds_from_genomic.fna.gz\n",
    "# keep_only_valid_orfs.pl alisa_efficient_codon_pair_score_hexamers_paper.pl expected_codon_freq.pl\n",
    "# get_atcg_freq.pl paper_get_5or_6_match_to_sd_pairs.pl paper_get_sd_p_value.pl\n",
    "# codon_pair.txt codons.txt syn_codons.txt\n",
    "import subprocess\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import csv\n",
    "import scipy.stats\n",
    "from os.path import join\n",
    "import glob\n",
    "\n",
    "def generate_non_SD_pipeline_files(file_prefix, input_path, output_path, SD):\n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    codon_pair_scores_file = output_path + file_prefix + \"_pairwise_syn_shuffle_frm123.csv\"\n",
    "    codon_pair_scores_helper_file = output_path + file_prefix + \"_obs_exp_frm123.cts\"\n",
    "    codon_bias_file = output_path + file_prefix + \".codon_bias\"\n",
    "    codon_pair_scores_only_matches_file = output_path + file_prefix + \".codon_pairs_SD_match\"\n",
    "    \n",
    "    # return empty if cds file is missing\n",
    "    names= glob.glob(cds_file)\n",
    "    if (len(names) < 1):\n",
    "        print(cds_file, \" is missing, not calculating.\")\n",
    "        return(False)\n",
    "    \n",
    "    # pre-process to get rid of bad orfs\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing keep_only_valid_orfs for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "        return(False)\n",
    "        \n",
    "    # get the codon pair scores of the cds\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"alisa_efficient_codon_pair_score_hexamers_paper.pl\", \"-s\", \\\n",
    "                                       processed_cds_file, \"-o\", file_prefix, \"-z\", \"2\"], stderr=subprocess.STDOUT)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "        return(False)\n",
    "    \n",
    "    # get the codon bias of the cds\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"cat \" + processed_cds_file + \" | perl expected_codon_freq.pl >\" + codon_bias_file, \\\n",
    "                              shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing expected_codon_freq.pl for \", cds_file)\n",
    "            return(False)\n",
    "    except:\n",
    "        print(\"Problem executing expected_codon_freq.pl for \", cds_file)\n",
    "        return(False)\n",
    "    \n",
    "    return(True)\n",
    "\n",
    "def generate_SD_pipeline_file(file_prefix, input_path, output_path, SD): \n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    codon_pair_scores_file = output_path + file_prefix + \"_pairwise_syn_shuffle_frm123.csv\"\n",
    "    codon_pair_scores_helper_file = output_path + file_prefix + \"_obs_exp_frm123.cts\"\n",
    "    codon_bias_file = output_path + file_prefix + \".codon_bias\"\n",
    "    codon_pair_scores_only_matches_file = output_path + file_prefix + \".codon_pairs_SD_match\"\n",
    "    \n",
    "    # return empty if cds file is missing\n",
    "    names= glob.glob(cds_file)\n",
    "    if (len(names) < 1):\n",
    "        print(cds_file, \" is missing, not calculating.\")\n",
    "        return([\"-1\",\"-1\",\"-1\",\"-1\"])     \n",
    "    \n",
    "    # collect the paper stats\n",
    "    stats = []\n",
    "    \n",
    "    # calculate the depletion score\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_atcg_freq.pl\", \"-i\", processed_cds_file, \"-s\", SD])\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        try:\n",
    "            stats.append(str(format(float(ret), '.3f')))\n",
    "        except:\n",
    "            print(\"Problem: get_atcg_freq.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_atcg_freq.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "      \n",
    "    # calculate the codon pair p-value\n",
    "    patched_SD = SD\n",
    "    if (len(SD) == 4):\n",
    "        if (SD[3] == \"G\"):\n",
    "            patched_SD = SD + \"A\"\n",
    "        else:\n",
    "            patched_SD = SD = \"G\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"perl paper_get_5or_6_match_to_sd_pairs.pl -s \" + patched_SD + \" -c \" + \\\n",
    "                                codon_pair_scores_file + \">\" + codon_pair_scores_only_matches_file, \\\n",
    "                                shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file) \n",
    "    except:\n",
    "        print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file)\n",
    "    # do the 1 -tailed test p-value calculation (previously done by hand in Excel)\n",
    "    # read in the array of all codon pair scores minus the ones with stop codons\n",
    "    all_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        codon1 = row[0][0:3]\n",
    "        codon2 = row[0][3:6]\n",
    "        if ((codon1 in (\"TAG\", \"TGA\", \"TAA\")) or (codon2 in (\"TAG\", \"TGA\", \"TAA\"))):\n",
    "            continue\n",
    "        all_f1_list.append(float(row[1]))\n",
    "    #print(all_f1_list)\n",
    "    #print(len(all_f1_list))\n",
    "    matches_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_only_matches_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        matches_f1_list.append(float(row[1]))\n",
    "    #print(matches_f1_list)\n",
    "    t_test_stats = scipy.stats.ttest_ind(all_f1_list, matches_f1_list, equal_var = False) # 2-tailed test, get 1 tail by / 2\n",
    "    codon_pair_p_value = float(t_test_stats.pvalue/2)\n",
    "    stats.append(str(format(codon_pair_p_value, '.3f')))\n",
    "    \n",
    "    # calculate motif depletion p-value\n",
    "    motif_depletion_p_value = 0.0\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"paper_get_sd_p_value.pl\", \"-s\", \\\n",
    "                                       SD, \"-f\", codon_bias_file], stderr=subprocess.STDOUT)\n",
    "        motif_depletion_p_value = float(ret.decode(\"utf-8\"))\n",
    "        try:\n",
    "            stats.append(str(format(float(motif_depletion_p_value), '.3f')))\n",
    "        except:\n",
    "            print(\"Problem: paper_get_sd_p_value.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # calculate the combined p-value\n",
    "    obj = scipy.stats.combine_pvalues([codon_pair_p_value, motif_depletion_p_value])\n",
    "    stats.append(str(format(float(obj[1]), '.3f')))\n",
    "    \n",
    "    return (stats)\n",
    "\n",
    "def clean_up_pipeline_files(file_prefix, input_path, output_path, SD): \n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    codon_pair_scores_file = output_path + file_prefix + \"_pairwise_syn_shuffle_frm123.csv\"\n",
    "    codon_pair_scores_helper_file = output_path + file_prefix + \"_obs_exp_frm123.cts\"\n",
    "    codon_bias_file = output_path + file_prefix + \".codon_bias\"\n",
    "    codon_pair_scores_only_matches_file = output_path + file_prefix + \".codon_pairs_SD_match\"\n",
    "    # cleanup: remove created temporary files\n",
    "    p = Popen(\"rm \" + processed_cds_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_pair_scores_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_pair_scores_helper_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_bias_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "    p = Popen(\"rm \" + codon_pair_scores_only_matches_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    p.wait()\n",
    "\n",
    "# expected file\n",
    "csv_with_tompa_sd_added = 'DiwanDataFiles_WithTompa_v2.csv'\n",
    "csv_with_table_stats_added = 'DiwanDataFiles_WithAllStats_v2.csv'\n",
    "input_file_path = \"/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "output_file_path = \"./\"\n",
    "\n",
    "\n",
    "count = 0\n",
    "will_run = 0\n",
    "with open(csv_with_tompa_sd_added, 'r') as in_file:\n",
    "    with open(csv_with_table_stats_added, 'a') as out_file: # change to 'a' to append when re-starting\n",
    "        csvReader = csv.reader(in_file, delimiter=',')\n",
    "        csvWriter = csv.writer(out_file, delimiter=',')\n",
    "        for row in csvReader:\n",
    "            if (row[0] == \"name\"):\n",
    "                row = row + [\"depletion_score\", \"codon pair p_value\", \"codon usage p_value\", \"combined p_value\"]\n",
    "                csvWriter.writerow(row)\n",
    "                continue\n",
    "            count += 1\n",
    "            #if (count > 2):\n",
    "            #    break\n",
    "            #if (count <= 120):\n",
    "             #   continue\n",
    "            #if (count > 2):\n",
    "             #   break\n",
    "            file_prefix = row[1]\n",
    "            done_SD = \"AGGAGG\"\n",
    "            #if (row[3] != done_SD):\n",
    "             #   will_run +=1\n",
    "              #  print(row[3])\n",
    "            if ((row[4] != done_SD) and (\"-\" not in row[4])):\n",
    "                will_run +=1\n",
    "                print(row[4])\n",
    "            #if (not generate_non_SD_pipeline_files(file_prefix, input_file_path, output_file_path, SD)):\n",
    "             #   stats = [\"-1\",\"-1\",\"-1\",\"-1\"]\n",
    "            #else:\n",
    "             #   stats = generate_SD_pipeline_file(file_prefix, input_file_path, output_file_path, SD)\n",
    "                #clean_up_pipeline_files(file_prefix, input_file_path, output_file_path, SD)\n",
    "            row = row + stats\n",
    "            #print(\",\".join(row))\n",
    "            csvWriter.writerow(row)\n",
    "print(\"Will run \" + str(will_run))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3721\n",
      "[-0.0953101798043249, 0.0, 0.108124857431976, 0.159848700941896, 0.356674943938732, 0.142668784873961, -0.500775287912489, -0.22314355131421]\n",
      "Wigglesworthia glossinidia endosymbiont of Glossina brevipalpis,GCA_000008885.1_ASM888v1,-------------,AGGAGG,GGAG,1.215,0.236,0.766,0.489\n"
     ]
    }
   ],
   "source": [
    "#  nohup python3 -u AutomateTable1Calculations_standalone.py > check_progress.log &\n",
    "#TODO: to reproduce paper organisms results, only keep ATG in keep_only_valid_orfs\n",
    "\n",
    "# expected perl: v5\n",
    "# expected input files:\n",
    "# *._cds_from_genomic.fna.gz\n",
    "# keep_only_valid_orfs.pl alisa_efficient_codon_pair_score_hexamers_paper.pl expected_codon_freq.pl\n",
    "# get_atcg_freq.pl paper_get_5or_6_match_to_sd_pairs.pl paper_get_sd_p_value.pl\n",
    "# codon_pair.txt codons.txt syn_codons.txt\n",
    "import subprocess\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import csv\n",
    "import scipy.stats\n",
    "from os.path import join\n",
    "import glob\n",
    "\n",
    "def run_table_generate_pipeline(file_prefix, input_path, output_path, SD): \n",
    "    \n",
    "    # pre-process to get rid of bad orfs\n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    \n",
    "    # return empty if cds file is missing\n",
    "    names= glob.glob(cds_file)\n",
    "    if (len(names) < 1):\n",
    "        print(cds_file, \" is missing, not calculating.\")\n",
    "        return([\"-1\",\"-1\",\"-1\",\"-1\"])\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"keep_only_valid_orfs.pl\", cds_file, processed_cds_file], stderr=subprocess.STDOUT)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing keep_only_valid_orfs for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # get the codon pair scores of the cds\n",
    "    codon_pair_scores_file = output_path + file_prefix + \"_pairwise_syn_shuffle_frm123.csv\"\n",
    "    codon_pair_scores_helper_file = output_path + file_prefix + \"_obs_exp_frm123.cts\"\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"alisa_efficient_codon_pair_score_hexamers_paper.pl\", \"-s\", \\\n",
    "                                       processed_cds_file, \"-o\", file_prefix, \"-z\", \"2\"], stderr=subprocess.STDOUT)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # get the codon bias of the cds\n",
    "    codon_bias_file = output_path + file_prefix + \".codon_bias\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"cat \" + processed_cds_file + \" | perl expected_codon_freq.pl >\" + codon_bias_file, \\\n",
    "                              shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing expected_codon_freq.pl for \", cds_file)  \n",
    "    except:\n",
    "        print(\"Problem executing expected_codon_freq.pl for \", cds_file)\n",
    "    \n",
    "    # collect the paper stats\n",
    "    stats = []\n",
    "    \n",
    "    # calculate the depletion score\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_atcg_freq.pl\", \"-i\", processed_cds_file, \"-s\", SD])\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        try:\n",
    "            stats.append(str(format(float(ret), '.3f')))\n",
    "        except:\n",
    "            print(\"Problem: get_atcg_freq.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_atcg_freq.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "      \n",
    "    # calculate the codon pair p-value\n",
    "    # patch:\n",
    "    patched_SD = SD\n",
    "    if (len(SD) == 4):\n",
    "        if (SD[3] == \"G\"):\n",
    "            patched_SD = SD + \"A\"\n",
    "        else:\n",
    "            patched_SD = SD = \"G\"\n",
    "    codon_pair_scores_only_matches_file = output_path + file_prefix + \".codon_pairs_SD_match\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"perl paper_get_5or_6_match_to_sd_pairs.pl -s \" + patched_SD + \" -c \" + \\\n",
    "                                codon_pair_scores_file + \">\" + codon_pair_scores_only_matches_file, \\\n",
    "                                shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file) \n",
    "    except:\n",
    "        print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file)\n",
    "    # do the 1 -tailed test p-value calculation (previously done by hand in Excel)\n",
    "    # read in the array of all codon pair scores minus the ones with stop codons\n",
    "    all_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        codon1 = row[0][0:3]\n",
    "        codon2 = row[0][3:6]\n",
    "        if ((codon1 in (\"TAG\", \"TGA\", \"TAA\")) or (codon2 in (\"TAG\", \"TGA\", \"TAA\"))):\n",
    "            continue\n",
    "        all_f1_list.append(float(row[1]))\n",
    "    #print(all_f1_list)\n",
    "    print(len(all_f1_list))\n",
    "    matches_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_only_matches_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        matches_f1_list.append(float(row[1]))\n",
    "    print(matches_f1_list)\n",
    "    t_test_stats = scipy.stats.ttest_ind(all_f1_list, matches_f1_list, equal_var = False) # 2-tailed test, get 1 tail by / 2\n",
    "    codon_pair_p_value = float(t_test_stats.pvalue/2)\n",
    "    stats.append(str(format(codon_pair_p_value, '.3f')))\n",
    "    \n",
    "    # calculate motif depletion p-value\n",
    "    motif_depletion_p_value = 0.0\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"paper_get_sd_p_value.pl\", \"-s\", \\\n",
    "                                       SD, \"-f\", codon_bias_file], stderr=subprocess.STDOUT)\n",
    "        motif_depletion_p_value = float(ret.decode(\"utf-8\"))\n",
    "        try:\n",
    "            stats.append(str(format(float(motif_depletion_p_value), '.3f')))\n",
    "        except:\n",
    "            print(\"Problem: paper_get_sd_p_value.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # calculate the combined p-value\n",
    "    obj = scipy.stats.combine_pvalues([codon_pair_p_value, motif_depletion_p_value])\n",
    "    stats.append(str(format(float(obj[1]), '.3f')))\n",
    "    \n",
    "    # output the final file\n",
    "    #with open(join(output_file_path, file_prefix + \".table3_values\"), 'w') as fout:\n",
    "     #   fout.write(stats)\n",
    "    \n",
    "    # cleanup: remove created temporary files\n",
    "    #TODO optionally: rm processed_cds_file, codon_pair_scores_file, codon_pair_scores_helper_file codon_bias_file \\\n",
    "    # codon_pair_scores_only_matches_file\n",
    "    # clean up the files\n",
    "    #p = Popen(\"rm \" + processed_cds_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + codon_pair_scores_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + codon_pair_scores_helper_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + codon_bias_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    #p = Popen(\"rm \" + codon_pair_scores_only_matches_file, shell=True, stdout=PIPE, stdin=PIPE, stderr=STDOUT)\n",
    "    #p.wait()\n",
    "    \n",
    "    return (stats)\n",
    "    \n",
    "\n",
    "# expected file\n",
    "csv_with_tompa_sd_added = 'DiwanDataFiles_WithTompa.csv'\n",
    "csv_with_table_stats_added = 'test.csv'\n",
    "input_file_path = \"/scratch4/moamin/ncbi_bacteria_db/GbBac_FNA/\"\n",
    "output_file_path = \"./\"\n",
    "\n",
    "\n",
    "count = 0\n",
    "with open(csv_with_tompa_sd_added, 'r') as in_file:\n",
    "    with open(csv_with_table_stats_added, 'a') as out_file: # change to 'a' to append when re-starting\n",
    "        csvReader = csv.reader(in_file, delimiter=',')\n",
    "        csvWriter = csv.writer(out_file, delimiter=',')\n",
    "        for row in csvReader:\n",
    "            if (row[0] == \"name\"):\n",
    "                row = row + [\"depletion_score\", \"codon pair p_value\", \"codon usage p_value\", \"combined p_value\"]\n",
    "                csvWriter.writerow(row)\n",
    "                continue\n",
    "            count += 1\n",
    "            #if (count > 2):\n",
    "            #    break\n",
    "            #if (count <= 120):\n",
    "             #   continue\n",
    "            #if (count > 4):\n",
    "             #   break\n",
    "            file_prefix = row[1]\n",
    "            if (file_prefix != \"GCA_000008885.1_ASM888v1\"):\n",
    "                 continue\n",
    "            SD = \"GGAG\"\n",
    "            stats = run_table_generate_pipeline(file_prefix, input_file_path, output_file_path, SD)\n",
    "            row = row + stats\n",
    "            print(\",\".join(row))\n",
    "            csvWriter.writerow(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sulfolobus_acidocaldarius,-0.036,0.043,0.427,0.091\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#TODO:  after testing with sulfolobus, uncomment GTG in keep only valid orfs\n",
    "import subprocess\n",
    "import csv\n",
    "import scipy\n",
    "from os.path import join\n",
    "\n",
    "def run_pipeline2(file_prefix, input_path, output_path, SD): \n",
    "    \n",
    "    cds_file = input_path + file_prefix + \"_cds_from_genomic.fna.gz\"\n",
    "    processed_cds_file = output_path + file_prefix + \"_cds_processed.fa\"\n",
    "    codon_pair_scores_file = output_path + file_prefix + \"_pairwise_syn_shuffle_frm123.csv\"\n",
    "    codon_pair_scores_helper_file = output_path + file_prefix + \"_obs_exp_frm123.cts\"\n",
    "    \n",
    "    # get the codon bias of the cds\n",
    "    codon_bias_file = output_path + file_prefix + \".codon_bias\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"cat \" + processed_cds_file + \" | perl expected_codon_freq.pl >\" + codon_bias_file, \\\n",
    "                              shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing expected_codon_freq.pl for \", cds_file)  \n",
    "    except:\n",
    "        print(\"Problem executing expected_codon_freq.pl for \", cds_file)\n",
    "                                      \n",
    "    # collect the paper stats\n",
    "    stats = file_prefix + \",\"\n",
    "    \n",
    "    # calculate the depletion score\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"get_atcg_freq.pl\", \"-i\", processed_cds_file, \"-s\", SD])\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        try:\n",
    "            stats += format(float(ret), '.3f') + \",\"\n",
    "        except:\n",
    "            print(\"Problem: get_atcg_freq.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing get_atcg_freq.pl for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "      \n",
    "    # calculate the codon pair p-value\n",
    "    codon_pair_scores_only_matches_file = output_path + file_prefix + \".codon_pairs_SD_match\"\n",
    "    try:\n",
    "        ps = subprocess.Popen(\"perl paper_get_5or_6_match_to_sd_pairs.pl -s \" + SD + \" -c \" + \\\n",
    "                                codon_pair_scores_file + \">\" + codon_pair_scores_only_matches_file, \\\n",
    "                                shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)\n",
    "        ret = ps.communicate()[0]\n",
    "        ret = ret.decode(\"utf-8\")\n",
    "        if (ret != \"\"):\n",
    "            print(\"_\", ret, \"_\")\n",
    "            print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file) \n",
    "    except:\n",
    "        print(\"Problem executing paper_get_5or_6_match_to_sd_pairs.pl for \", cds_file)\n",
    "    # do the 1 -tailed test p-value calculation (previously done by hand in Excel)\n",
    "    # read in the array of all codon pair scores minus the ones with stop codons\n",
    "    all_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        codon1 = row[0][0:3]\n",
    "        codon2 = row[0][3:6]\n",
    "        if ((codon1 in (\"TAG\", \"TGA\", \"TAA\")) or (codon2 in (\"TAG\", \"TGA\", \"TAA\"))):\n",
    "            continue\n",
    "        all_f1_list.append(float(row[1]))\n",
    "    #print(all_f1_list)\n",
    "    #print(len(all_f1_list))\n",
    "    matches_f1_list = []\n",
    "    csvfile = open(codon_pair_scores_only_matches_file)\n",
    "    reader = csv.reader(csvfile, delimiter=',')\n",
    "    for row in reader:\n",
    "        matches_f1_list.append(float(row[1]))\n",
    "    #print(matches_f1_list)\n",
    "    t_test_stats = scipy.stats.ttest_ind(all_f1_list, matches_f1_list, equal_var = False) # 2-tailed test, get 1 tail by / 2\n",
    "    codon_pair_p_value = float(t_test_stats.pvalue/2)\n",
    "    stats += format(codon_pair_p_value, '.3f') + \",\"\n",
    "    \n",
    "    # calculate motif depletion p-value\n",
    "    motif_depletion_p_value = 0.0\n",
    "    try:\n",
    "        ret = subprocess.check_output([\"perl\", \"paper_get_sd_p_value.pl\", \"-s\", \\\n",
    "                                       SD, \"-f\", codon_bias_file], stderr=subprocess.STDOUT)\n",
    "        motif_depletion_p_value = float(ret.decode(\"utf-8\"))\n",
    "        try:\n",
    "            stats += format(float(motif_depletion_p_value), '.3f') + \",\"\n",
    "        except:\n",
    "            print(\"Problem: paper_get_sd_p_value.pl did not return a float for \", cds_file)\n",
    "    except subprocess.CalledProcessError as exc:\n",
    "        print(\"Problem executing alisa_efficient_codon_pair_score_hexamers for \", cds_file, \"; \", exc.returncode, exc.output)\n",
    "    \n",
    "    # calculate the combined p-value\n",
    "    obj = scipy.stats.combine_pvalues([codon_pair_p_value, motif_depletion_p_value])\n",
    "    stats += format(float(obj[1]), '.3f')\n",
    "    stats += \"\\n\"\n",
    "    \n",
    "    # output the final file\n",
    "    with open(join(output_file_path, file_prefix + \".table3_values\"), 'w') as fout:\n",
    "        fout.write(stats)\n",
    "    \n",
    "    print(stats)\n",
    "    \n",
    "    # cleanup: remove created temporary files\n",
    "    #TODO optionally: rm processed_cds_file, codon_pair_scores_file, codon_pair_scores_helper_file codon_bias_file \\\n",
    "    # codon_pair_scores_only_matches_file\n",
    "    \n",
    "\n",
    "input_file_path = \"./\"\n",
    "output_file_path = \"./\"\n",
    "file_prefix = \"Sulfolobus_acidocaldarius\"\n",
    "SD = \"GGTGA\"\n",
    "run_pipeline2(file_prefix, input_file_path, output_file_path, SD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
